diff --git a/thunar/thunar-preferences.c b/thunar/thunar-preferences.c
index f6276a20..25593812 100644
--- a/thunar/thunar-preferences.c
+++ b/thunar/thunar-preferences.c
@@ -563,7 +563,7 @@ thunar_preferences_class_init (ThunarPreferencesClass *klass)
       g_param_spec_boolean ("misc-full-path-in-title",
                             "MiscFullPathInTitle",
                             NULL,
-                            FALSE,
+                            TRUE,
                             EXO_PARAM_READWRITE);
 
   /**
