diff --git a/thunar/thunar-chooser-dialog.c b/thunar/thunar-chooser-dialog.c
index 91fe07c7..efb1f703 100644
--- a/thunar/thunar-chooser-dialog.c
+++ b/thunar/thunar-chooser-dialog.c
@@ -1258,11 +1258,13 @@ thunar_chooser_dialog_set_open (ThunarChooserDialog *dialog,
 
 /**
  * thunar_show_chooser_dialog:
- * @parent : the #GtkWidget or the #GdkScreen on which to open the
- *           dialog. May also be %NULL in which case the default
- *           #GdkScreen will be used.
- * @file   : the #ThunarFile for which an application should be chosen.
- * @open   : whether to also open the @file.
+ * @parent                      : the #GtkWidget or the #GdkScreen on which to open the
+ *                                dialog. May also be %NULL in which case the default
+ *                                #GdkScreen will be used.
+ * @file                        : the #ThunarFile for which an application should be chosen.
+ * @open                        : whether to also open the @file.
+ * @preselect_default_checkbox  : Check the checkbox by default and
+ *                                set the title "Set Default Application" of chooser dialog box.
  *
  * Convenience function to display a #ThunarChooserDialog with the
  * given parameters.
@@ -1275,7 +1277,8 @@ thunar_chooser_dialog_set_open (ThunarChooserDialog *dialog,
 void
 thunar_show_chooser_dialog (gpointer    parent,
                             ThunarFile *file,
-                            gboolean    open)
+                            gboolean    open,
+                            gboolean    preselect_default_checkbox)
 {
   ThunarApplication *application;
   GdkScreen         *screen;
@@ -1310,6 +1313,13 @@ thunar_show_chooser_dialog (gpointer    parent,
                          "screen", screen,
                          NULL);
 
+  /* change title of chooser dialog box and checked the checkbox when it is used for select default application */
+  if (preselect_default_checkbox)
+  {
+    gtk_window_set_title (GTK_WINDOW (dialog), _("Set Default Application"));
+    gtk_toggle_button_set_active (GTK_TOGGLE_BUTTON (THUNAR_CHOOSER_DIALOG (dialog)->default_button), preselect_default_checkbox);
+  }
+
   /* check if we have a toplevel window */
   if (G_LIKELY (window != NULL && gtk_widget_get_toplevel (window)))
     {
diff --git a/thunar/thunar-chooser-dialog.h b/thunar/thunar-chooser-dialog.h
index 149c9d31..0137eb2a 100644
--- a/thunar/thunar-chooser-dialog.h
+++ b/thunar/thunar-chooser-dialog.h
@@ -38,7 +38,8 @@ GType       thunar_chooser_dialog_get_type  (void) G_GNUC_CONST;
 
 void        thunar_show_chooser_dialog      (gpointer             parent,
                                              ThunarFile          *file,
-                                             gboolean             open);
+                                             gboolean             open,
+                                             gboolean             preselect_default_checkbox);
 
 G_END_DECLS;
 
diff --git a/thunar/thunar-dbus-service.c b/thunar/thunar-dbus-service.c
index 2d27642d..78064885 100644
--- a/thunar/thunar-dbus-service.c
+++ b/thunar/thunar-dbus-service.c
@@ -451,7 +451,7 @@ thunar_dbus_service_display_chooser_dialog (ThunarDBusFileManager  *object,
 
   /* popup the chooser dialog */
   /* TODO use the startup id! */
-  thunar_show_chooser_dialog (screen, file, open);
+  thunar_show_chooser_dialog (screen, file, open, FALSE);
 
   /* cleanup */
   g_object_unref (G_OBJECT (screen));
diff --git a/thunar/thunar-file.c b/thunar/thunar-file.c
index a5423010..69bf75c2 100644
--- a/thunar/thunar-file.c
+++ b/thunar/thunar-file.c
@@ -1828,7 +1828,7 @@ thunar_file_launch (ThunarFile  *file,
    * type yet */
   if (G_UNLIKELY (app_info == NULL))
     {
-      thunar_show_chooser_dialog (parent, file, TRUE);
+      thunar_show_chooser_dialog (parent, file, TRUE, FALSE);
       return TRUE;
     }
 
@@ -1839,7 +1839,7 @@ thunar_file_launch (ThunarFile  *file,
       || g_strcmp0 (g_app_info_get_name (app_info), "exo-file-manager") == 0)
     {
       g_object_unref (G_OBJECT (app_info));
-      thunar_show_chooser_dialog (parent, file, TRUE);
+      thunar_show_chooser_dialog (parent, file, TRUE, FALSE);
       return TRUE;
     }
 
diff --git a/thunar/thunar-launcher.c b/thunar/thunar-launcher.c
index a7b8cb6c..58130947 100644
--- a/thunar/thunar-launcher.c
+++ b/thunar/thunar-launcher.c
@@ -173,6 +173,7 @@ static void                    thunar_launcher_action_open                (Thuna
 static void                    thunar_launcher_action_open_in_new_tabs    (ThunarLauncher                 *launcher);
 static void                    thunar_launcher_action_open_in_new_windows (ThunarLauncher                 *launcher);
 static void                    thunar_launcher_action_open_with_other     (ThunarLauncher                 *launcher);
+static void                    thunar_launcher_action_set_default_app     (ThunarLauncher                 *launcher);
 static void                    thunar_launcher_action_sendto_desktop      (ThunarLauncher                 *launcher);
 static void                    thunar_launcher_action_properties          (ThunarLauncher                 *launcher);
 static void                    thunar_launcher_action_sendto_device       (ThunarLauncher                 *launcher,
@@ -267,6 +268,7 @@ static XfceGtkActionEntry thunar_launcher_action_entries[] =
     { THUNAR_LAUNCHER_ACTION_OPEN_IN_TAB,      "<Actions>/ThunarLauncher/open-in-new-tab",         "<Primary><shift>P", XFCE_GTK_MENU_ITEM,       NULL,                                   NULL,                                                                                            NULL,                   G_CALLBACK (thunar_launcher_action_open_in_new_tabs),    },
     { THUNAR_LAUNCHER_ACTION_OPEN_IN_WINDOW,   "<Actions>/ThunarLauncher/open-in-new-window",      "<Primary><shift>O", XFCE_GTK_MENU_ITEM,       NULL,                                   NULL,                                                                                            NULL,                   G_CALLBACK (thunar_launcher_action_open_in_new_windows), },
     { THUNAR_LAUNCHER_ACTION_OPEN_WITH_OTHER,  "<Actions>/ThunarLauncher/open-with-other",         "",                  XFCE_GTK_MENU_ITEM,       N_ ("Open With Other _Application..."), N_ ("Choose another application with which to open the selected file"),                          NULL,                   G_CALLBACK (thunar_launcher_action_open_with_other),     },
+    { THUNAR_LAUNCHER_ACTION_SET_DEFAULT_APP,  "<Actions>/ThunarStandardView/set-default-app",     "",                  XFCE_GTK_MENU_ITEM,       N_ ("Set _Default Application"),        N_ ("Choose an application which should be used by default to open the selected file"),          NULL,                   G_CALLBACK (thunar_launcher_action_set_default_app),     },
 
     /* For backward compatibility the old accel paths are re-used. Currently not possible to automatically migrate to new accel paths. */
     /* Waiting for https://gitlab.gnome.org/GNOME/gtk/issues/2375 to be able to fix that */
@@ -857,7 +859,7 @@ thunar_launcher_open_files (ThunarLauncher *launcher,
       else
         {
           /* display a chooser dialog for the file and stop */
-          thunar_show_chooser_dialog (launcher->widget, lp->data, TRUE);
+          thunar_show_chooser_dialog (launcher->widget, lp->data, TRUE, FALSE);
           break;
         }
     }
@@ -1372,7 +1374,24 @@ thunar_launcher_action_open_with_other (ThunarLauncher *launcher)
   _thunar_return_if_fail (THUNAR_IS_LAUNCHER (launcher));
 
   if (launcher->n_files_to_process == 1)
-    thunar_show_chooser_dialog (launcher->widget, launcher->files_to_process->data, TRUE);
+    thunar_show_chooser_dialog (launcher->widget, launcher->files_to_process->data, TRUE, FALSE);
+}
+
+
+
+/**
+ * thunar_launcher_action_set_default_app:
+ * @launcher : a #ThunarLauncher instance
+ *
+ * Choose an application which should be used by default to open the selected file
+ **/
+static void
+thunar_launcher_action_set_default_app (ThunarLauncher *launcher)
+{
+  _thunar_return_if_fail (THUNAR_IS_LAUNCHER (launcher));
+
+  if (launcher->n_files_to_process == 1)
+    thunar_show_chooser_dialog (launcher->widget, launcher->files_to_process->data, TRUE, TRUE);
 }
 
 
@@ -1482,6 +1501,11 @@ thunar_launcher_append_menu_item (ThunarLauncher       *launcher,
         return xfce_gtk_menu_item_new (action_entry->menu_item_label_text, action_entry->menu_item_tooltip_text,
                                        action_entry->accel_path, action_entry->callback, G_OBJECT (launcher), menu);
 
+      case THUNAR_LAUNCHER_ACTION_SET_DEFAULT_APP:
+        if (launcher->n_files_to_process != 1)
+          return NULL;
+        return xfce_gtk_menu_item_new_from_action_entry (action_entry, G_OBJECT (launcher), GTK_MENU_SHELL (menu));
+
       case THUNAR_LAUNCHER_ACTION_SENDTO_MENU:
         if (launcher->files_are_selected == FALSE)
           return NULL;
@@ -2917,6 +2941,8 @@ thunar_launcher_build_application_submenu (ThunarLauncher *launcher,
     {
       xfce_gtk_menu_append_seperator (GTK_MENU_SHELL (submenu));
       thunar_launcher_append_menu_item (launcher, GTK_MENU_SHELL (submenu), THUNAR_LAUNCHER_ACTION_OPEN_WITH_OTHER, FALSE);
+      xfce_gtk_menu_append_seperator (GTK_MENU_SHELL (submenu));
+      thunar_launcher_append_menu_item (launcher, GTK_MENU_SHELL (submenu), THUNAR_LAUNCHER_ACTION_SET_DEFAULT_APP, FALSE);
     }
 
   return submenu;
@@ -3015,7 +3041,11 @@ thunar_launcher_append_open_section (ThunarLauncher *launcher,
   else
     {
       if (launcher->n_files_to_process == 1)
+      {
         thunar_launcher_append_menu_item (launcher, GTK_MENU_SHELL (menu), THUNAR_LAUNCHER_ACTION_OPEN_WITH_OTHER, FALSE);
+        xfce_gtk_menu_append_seperator (GTK_MENU_SHELL (menu));
+        thunar_launcher_append_menu_item (launcher, GTK_MENU_SHELL (menu), THUNAR_LAUNCHER_ACTION_SET_DEFAULT_APP, FALSE);
+      }
     }
 
   g_list_free_full (applications, g_object_unref);
diff --git a/thunar/thunar-launcher.h b/thunar/thunar-launcher.h
index 91b73bd2..1075e3b7 100644
--- a/thunar/thunar-launcher.h
+++ b/thunar/thunar-launcher.h
@@ -44,6 +44,7 @@ typedef enum
   THUNAR_LAUNCHER_ACTION_OPEN_IN_TAB,
   THUNAR_LAUNCHER_ACTION_OPEN_IN_WINDOW,
   THUNAR_LAUNCHER_ACTION_OPEN_WITH_OTHER,
+  THUNAR_LAUNCHER_ACTION_SET_DEFAULT_APP,
   THUNAR_LAUNCHER_ACTION_SENDTO_MENU,
   THUNAR_LAUNCHER_ACTION_SENDTO_SHORTCUTS,
   THUNAR_LAUNCHER_ACTION_SENDTO_DESKTOP,
