diff --git a/thunar/thunar-component.c b/thunar/thunar-component.c
index eaed115d..9f6a9d3d 100644
--- a/thunar/thunar-component.c
+++ b/thunar/thunar-component.c
@@ -140,7 +140,7 @@ thunar_component_restore_selection (ThunarComponent *component)
 
   _thunar_return_if_fail (THUNAR_IS_COMPONENT (component));
 
-  selected_files = thunar_g_file_list_copy (thunar_component_get_selected_files (component));
+  selected_files = thunar_g_list_copy_deep (thunar_component_get_selected_files (component));
   thunar_component_set_selected_files (component, selected_files);
   thunar_g_file_list_free (selected_files);
 }
diff --git a/thunar/thunar-gio-extensions.c b/thunar/thunar-gio-extensions.c
index 0544593e..a6034bc1 100644
--- a/thunar/thunar-gio-extensions.c
+++ b/thunar/thunar-gio-extensions.c
@@ -497,7 +497,7 @@ thunar_g_file_list_get_type (void)
   if (G_UNLIKELY (type == G_TYPE_INVALID))
     {
       type = g_boxed_type_register_static (I_("ThunarGFileList"),
-                                           (GBoxedCopyFunc) thunar_g_file_list_copy,
+                                           (GBoxedCopyFunc) thunar_g_list_copy_deep,
                                            (GBoxedFreeFunc) thunar_g_file_list_free);
     }
 
diff --git a/thunar/thunar-gio-extensions.h b/thunar/thunar-gio-extensions.h
index 9a68e59c..7bc9e972 100644
--- a/thunar/thunar-gio-extensions.h
+++ b/thunar/thunar-gio-extensions.h
@@ -79,7 +79,7 @@ GList    *thunar_g_file_list_get_parents            (GList             *list);
 /* deep copy jobs for GLists */
 #define   thunar_g_list_append_deep(list,object)    g_list_append (list, g_object_ref (G_OBJECT (object)))
 #define   thunar_g_list_prepend_deep(list,object)   g_list_prepend (list, g_object_ref (G_OBJECT (object)))
-#define   thunar_g_file_list_copy                   thunarx_file_info_list_copy
+#define   thunar_g_list_copy_deep                   thunarx_file_info_list_copy
 #define   thunar_g_file_list_free                   thunarx_file_info_list_free
 
 gboolean  thunar_g_app_info_launch                  (GAppInfo          *info,
diff --git a/thunar/thunar-io-jobs.c b/thunar/thunar-io-jobs.c
index 9c5635ef..de9b5505 100644
--- a/thunar/thunar-io-jobs.c
+++ b/thunar/thunar-io-jobs.c
@@ -961,7 +961,7 @@ _thunar_io_jobs_chown (ThunarJob  *job,
   if (recursive)
     file_list = _tij_collect_nofollow (job, file_list, FALSE, &err);
   else
-    file_list = thunar_g_file_list_copy (file_list);
+    file_list = thunar_g_list_copy_deep (file_list);
 
   if (err != NULL)
     {
@@ -1104,7 +1104,7 @@ _thunar_io_jobs_chmod (ThunarJob  *job,
   if (recursive)
     file_list = _tij_collect_nofollow (job, file_list, FALSE, &err);
   else
-    file_list = thunar_g_file_list_copy (file_list);
+    file_list = thunar_g_list_copy_deep (file_list);
 
   if (err != NULL)
     {
diff --git a/thunar/thunar-launcher.c b/thunar/thunar-launcher.c
index ebde68d8..13a9f2a5 100644
--- a/thunar/thunar-launcher.c
+++ b/thunar/thunar-launcher.c
@@ -636,7 +636,7 @@ thunar_launcher_set_selected_files (ThunarComponent *component,
 
   /* if nothing is selected, the current directory is the folder to use for all menus */
   if (launcher->files_are_selected)
-    launcher->files_to_process = thunar_g_file_list_copy (selected_files);
+    launcher->files_to_process = thunar_g_list_copy_deep (selected_files);
   else
     launcher->files_to_process = g_list_append (launcher->files_to_process, launcher->current_directory);
 
@@ -845,7 +845,7 @@ thunar_launcher_open_files (ThunarLauncher *launcher,
           if (G_LIKELY (file_list != NULL))
             {
               /* take a copy of the list as the old one will be dropped by the insert */
-              file_list = thunar_g_file_list_copy (file_list);
+              file_list = thunar_g_list_copy_deep (file_list);
             }
 
           /* append our new URI to the list */
@@ -1241,7 +1241,7 @@ thunar_launcher_poke_data_new (GList                          *files_to_poke,
   ThunarLauncherPokeData *data;
 
   data = g_slice_new0 (ThunarLauncherPokeData);
-  data->files_to_poke = thunar_g_file_list_copy (files_to_poke);
+  data->files_to_poke = thunar_g_list_copy_deep (files_to_poke);
   data->files_poked = NULL;
   data->application_to_use = application_to_use;
 
diff --git a/thunar/thunar-shortcuts-pane.c b/thunar/thunar-shortcuts-pane.c
index df0ef648..e63992d9 100644
--- a/thunar/thunar-shortcuts-pane.c
+++ b/thunar/thunar-shortcuts-pane.c
@@ -337,7 +337,7 @@ thunar_shortcuts_pane_set_selected_files (ThunarComponent *component,
   thunar_g_file_list_free (shortcuts_pane->selected_files);
 
   /* ...and take a copy of the newly selected thunar files */
-  shortcuts_pane->selected_files = thunar_g_file_list_copy (selected_files);
+  shortcuts_pane->selected_files = thunar_g_list_copy_deep (selected_files);
 
   /* notify listeners */
   g_object_notify (G_OBJECT (shortcuts_pane), "selected-files");
diff --git a/thunar/thunar-standard-view.c b/thunar/thunar-standard-view.c
index 464818a5..dfa5e99b 100644
--- a/thunar/thunar-standard-view.c
+++ b/thunar/thunar-standard-view.c
@@ -1173,7 +1173,7 @@ thunar_standard_view_set_selected_files_component (ThunarComponent *component,
   if (thunar_view_get_loading (THUNAR_VIEW (standard_view)))
     {
       /* remember a copy of the list for later */
-      standard_view->priv->selected_files = thunar_g_file_list_copy (selected_files);
+      standard_view->priv->selected_files = thunar_g_list_copy_deep (selected_files);
     }
   else
     {
@@ -2329,7 +2329,7 @@ thunar_standard_view_new_files (ThunarStandardView *standard_view,
   if (G_UNLIKELY (standard_view->loading))
     {
       /* schedule the "new-files" paths for later processing */
-      standard_view->priv->new_files_path_list = thunar_g_file_list_copy (path_list);
+      standard_view->priv->new_files_path_list = thunar_g_list_copy_deep (path_list);
     }
   else if (G_LIKELY (path_list != NULL))
     {
@@ -2364,7 +2364,7 @@ thunar_standard_view_new_files (ThunarStandardView *standard_view,
           /* thunar files are not created yet, try again later because we know
            * some of them belong in this directory, so eventually they
            * will get a ThunarFile */
-          standard_view->priv->new_files_path_list = thunar_g_file_list_copy (path_list);
+          standard_view->priv->new_files_path_list = thunar_g_list_copy_deep (path_list);
         }
     }
 
diff --git a/thunar/thunar-window.c b/thunar/thunar-window.c
index f74d0b96..1a9ee987 100644
--- a/thunar/thunar-window.c
+++ b/thunar/thunar-window.c
@@ -3070,7 +3070,7 @@ thunar_window_replace_view (ThunarWindow *window,
         g_object_ref (current_directory);
 
       /* remember the file selection from the old view */
-      selected_thunar_files = thunar_g_file_list_copy (thunar_component_get_selected_files (THUNAR_COMPONENT (view)));
+      selected_thunar_files = thunar_g_list_copy_deep (thunar_component_get_selected_files (THUNAR_COMPONENT (view)));
 
       /* save the history of the current view */
       history = NULL;
