diff --git a/thunar/thunar-application.c b/thunar/thunar-application.c
index 688d149e..60441cab 100644
--- a/thunar/thunar-application.c
+++ b/thunar/thunar-application.c
@@ -398,7 +398,7 @@ thunar_application_shutdown (GApplication *gapp)
   ThunarApplication *application = THUNAR_APPLICATION (gapp);
 
   /* unqueue all files waiting to be processed */
-  thunar_g_file_list_free (application->files_to_launch);
+  thunar_g_list_free_full (application->files_to_launch);
 
   /* save the current accel map */
   if (G_UNLIKELY (application->accel_map_save_id != 0))
@@ -761,7 +761,7 @@ thunar_application_collect_and_launch (ThunarApplication *application,
     }
 
   /* release the target path list */
-  thunar_g_file_list_free (target_file_list);
+  thunar_g_list_free_full (target_file_list);
 }
 
 
@@ -1445,7 +1445,7 @@ thunar_application_bulk_rename (ThunarApplication *application,
 
   /* cleanup */
   g_object_unref (G_OBJECT (current_directory));
-  thunar_g_file_list_free (file_list);
+  thunar_g_list_free_full (file_list);
 
   return result;
 }
@@ -1507,7 +1507,7 @@ thunar_application_process_files_finish (ThunarBrowser *browser,
         }
 
       /* stop processing files */
-      thunar_g_file_list_free (application->files_to_launch);
+      thunar_g_list_free_full (application->files_to_launch);
       application->files_to_launch = NULL;
     }
   else
@@ -1641,7 +1641,7 @@ thunar_application_process_filenames (ThunarApplication *application,
                        _("Failed to open \"%s\": %s"), filenames[n], derror->message);
           g_error_free (derror);
 
-          thunar_g_file_list_free (file_list);
+          thunar_g_list_free_full (file_list);
 
           return FALSE;
         }
@@ -2188,7 +2188,7 @@ thunar_application_unlink_files (ThunarApplication *application,
     }
 
   /* release the path list */
-  thunar_g_file_list_free (path_list);
+  thunar_g_list_free_full (path_list);
 }
 
 
@@ -2439,8 +2439,8 @@ thunar_application_restore_files (ThunarApplication *application,
     }
 
   /* free path lists */
-  thunar_g_file_list_free (source_path_list);
-  thunar_g_file_list_free (target_path_list);
+  thunar_g_list_free_full (source_path_list);
+  thunar_g_list_free_full (target_path_list);
 }
 
 
diff --git a/thunar/thunar-clipboard-manager.c b/thunar/thunar-clipboard-manager.c
index 9092ef61..7928ca52 100644
--- a/thunar/thunar-clipboard-manager.c
+++ b/thunar/thunar-clipboard-manager.c
@@ -332,7 +332,7 @@ thunar_clipboard_manager_contents_received (GtkClipboard     *clipboard,
       else
         thunar_application_move_into (application, request->widget, file_list, request->target_file, request->new_files_closure);
       g_object_unref (G_OBJECT (application));
-      thunar_g_file_list_free (file_list);
+      thunar_g_list_free_full (file_list);
 
       /* clear the clipboard if it contained "cutted data"
        * (gtk_clipboard_clear takes care of not clearing
@@ -487,7 +487,7 @@ thunar_clipboard_manager_get_callback (GtkClipboard     *clipboard,
     }
 
   /* cleanup */
-  thunar_g_file_list_free (file_list);
+  thunar_g_list_free_full (file_list);
 }
 
 
diff --git a/thunar/thunar-component.c b/thunar/thunar-component.c
index 9f6a9d3d..efaa54ee 100644
--- a/thunar/thunar-component.c
+++ b/thunar/thunar-component.c
@@ -142,5 +142,5 @@ thunar_component_restore_selection (ThunarComponent *component)
 
   selected_files = thunar_g_list_copy_deep (thunar_component_get_selected_files (component));
   thunar_component_set_selected_files (component, selected_files);
-  thunar_g_file_list_free (selected_files);
+  thunar_g_list_free_full (selected_files);
 }
diff --git a/thunar/thunar-dbus-service.c b/thunar/thunar-dbus-service.c
index 298e2403..2ac4ff61 100644
--- a/thunar/thunar-dbus-service.c
+++ b/thunar/thunar-dbus-service.c
@@ -870,7 +870,7 @@ thunar_dbus_service_move_to_trash (ThunarDBusTrash        *object,
         }
 
       /* cleanup */
-      thunar_g_file_list_free (file_list);
+      thunar_g_list_free_full (file_list);
       g_object_unref (screen);
     }
 
@@ -1253,8 +1253,8 @@ thunar_dbus_service_transfer_files (ThunarDBusTransferMode transfer_mode,
         }
 
       /* free the file lists */
-      thunar_g_file_list_free (source_file_list);
-      thunar_g_file_list_free (target_file_list);
+      thunar_g_list_free_full (source_file_list);
+      thunar_g_list_free_full (target_file_list);
 
       /* release the screen */
       g_object_unref (screen);
@@ -1467,7 +1467,7 @@ thunar_dbus_service_unlink_files (ThunarDBusFileManager  *object,
         }
 
       /* cleanup */
-      thunar_g_file_list_free (file_list);
+      thunar_g_list_free_full (file_list);
       g_object_unref (screen);
     }
 
diff --git a/thunar/thunar-file.c b/thunar/thunar-file.c
index 7bd9a4c3..756b488d 100644
--- a/thunar/thunar-file.c
+++ b/thunar/thunar-file.c
@@ -4452,7 +4452,7 @@ thunar_file_list_get_applications (GList *file_list)
  * the #ThunarFile<!---->s contained within @file_list.
  *
  * The caller is responsible to free the returned list using
- * thunar_g_file_list_free() when no longer needed.
+ * thunar_g_list_free_full() when no longer needed.
  *
  * Return value: the list of #GFile<!---->s for @file_list.
  **/
diff --git a/thunar/thunar-folder.c b/thunar/thunar-folder.c
index 6e7af454..d686ec67 100644
--- a/thunar/thunar-folder.c
+++ b/thunar/thunar-folder.c
@@ -335,10 +335,10 @@ thunar_folder_finalize (GObject *object)
     g_source_remove (folder->content_type_idle_id);
 
   /* release references to the new files */
-  thunar_g_file_list_free (folder->new_files);
+  thunar_g_list_free_full (folder->new_files);
 
   /* release references to the current files */
-  thunar_g_file_list_free (folder->files);
+  thunar_g_list_free_full (folder->files);
 
   (*G_OBJECT_CLASS (thunar_folder_parent_class)->finalize) (object);
 }
@@ -559,11 +559,11 @@ thunar_folder_finished (ExoJob       *job,
           g_signal_emit (G_OBJECT (folder), folder_signals[FILES_REMOVED], 0, files);
 
           /* release the removed files list */
-          thunar_g_file_list_free (files);
+          thunar_g_list_free_full (files);
         }
 
       /* drop the temporary new_files list */
-      thunar_g_file_list_free (folder->new_files);
+      thunar_g_list_free_full (folder->new_files);
       folder->new_files = NULL;
     }
   else
@@ -1026,7 +1026,7 @@ thunar_folder_reload (ThunarFolder *folder,
     }
 
   /* reset the new_files list */
-  thunar_g_file_list_free (folder->new_files);
+  thunar_g_list_free_full (folder->new_files);
   folder->new_files = NULL;
 
   /* start a new job */
diff --git a/thunar/thunar-gio-extensions.c b/thunar/thunar-gio-extensions.c
index a6034bc1..f25f4500 100644
--- a/thunar/thunar-gio-extensions.c
+++ b/thunar/thunar-gio-extensions.c
@@ -498,7 +498,7 @@ thunar_g_file_list_get_type (void)
     {
       type = g_boxed_type_register_static (I_("ThunarGFileList"),
                                            (GBoxedCopyFunc) thunar_g_list_copy_deep,
-                                           (GBoxedFreeFunc) thunar_g_file_list_free);
+                                           (GBoxedFreeFunc) thunar_g_list_free_full);
     }
 
   return type;
diff --git a/thunar/thunar-gio-extensions.h b/thunar/thunar-gio-extensions.h
index 7bc9e972..7b6944cc 100644
--- a/thunar/thunar-gio-extensions.h
+++ b/thunar/thunar-gio-extensions.h
@@ -80,7 +80,7 @@ GList    *thunar_g_file_list_get_parents            (GList             *list);
 #define   thunar_g_list_append_deep(list,object)    g_list_append (list, g_object_ref (G_OBJECT (object)))
 #define   thunar_g_list_prepend_deep(list,object)   g_list_prepend (list, g_object_ref (G_OBJECT (object)))
 #define   thunar_g_list_copy_deep                   thunarx_file_info_list_copy
-#define   thunar_g_file_list_free                   thunarx_file_info_list_free
+#define   thunar_g_list_free_full                   thunarx_file_info_list_free
 
 gboolean  thunar_g_app_info_launch                  (GAppInfo          *info,
                                                      GFile             *working_directory,
diff --git a/thunar/thunar-io-jobs.c b/thunar/thunar-io-jobs.c
index de9b5505..5a69b489 100644
--- a/thunar/thunar-io-jobs.c
+++ b/thunar/thunar-io-jobs.c
@@ -78,7 +78,7 @@ _tij_collect_nofollow (ThunarJob *job,
         g_propagate_error (error, err);
 
       /* release the collected files */
-      thunar_g_file_list_free (file_list);
+      thunar_g_list_free_full (file_list);
 
       return NULL;
     }
@@ -469,7 +469,7 @@ _thunar_io_jobs_unlink (ThunarJob  *job,
       else
         g_propagate_error (error, err);
 
-      thunar_g_file_list_free (file_list);
+      thunar_g_list_free_full (file_list);
       return FALSE;
     }
 
@@ -551,7 +551,7 @@ again:
   g_object_unref (thumbnail_cache);
 
   /* release the file list */
-  thunar_g_file_list_free (file_list);
+  thunar_g_list_free_full (file_list);
 
   if (exo_job_set_error_if_cancelled (EXO_JOB (job), error))
     return FALSE;
@@ -804,14 +804,14 @@ _thunar_io_jobs_link (ThunarJob  *job,
 
   if (err != NULL)
     {
-      thunar_g_file_list_free (new_files_list);
+      thunar_g_list_free_full (new_files_list);
       g_propagate_error (error, err);
       return FALSE;
     }
   else
     {
       thunar_job_new_files (THUNAR_JOB (job), new_files_list);
-      thunar_g_file_list_free (new_files_list);
+      thunar_g_list_free_full (new_files_list);
       return TRUE;
     }
 }
@@ -1033,7 +1033,7 @@ retry_chown:
     }
 
   /* release the file list */
-  thunar_g_file_list_free (file_list);
+  thunar_g_list_free_full (file_list);
 
   if (err != NULL)
     {
@@ -1185,7 +1185,7 @@ retry_chown:
     }
 
   /* release the file list */
-  thunar_g_file_list_free (file_list);
+  thunar_g_list_free_full (file_list);
 
   if (err != NULL)
     {
@@ -1273,7 +1273,7 @@ _thunar_io_jobs_ls (ThunarJob  *job,
         {
           /* none of the handlers took over the file list, so it's up to us
            * to destroy it */
-          thunar_g_file_list_free (file_list);
+          thunar_g_list_free_full (file_list);
         }
     }
 
diff --git a/thunar/thunar-io-scan-directory.c b/thunar/thunar-io-scan-directory.c
index 307564b0..473cbe35 100644
--- a/thunar/thunar-io-scan-directory.c
+++ b/thunar/thunar-io-scan-directory.c
@@ -168,13 +168,13 @@ thunar_io_scan_directory (ThunarJob          *job,
   if (G_UNLIKELY (err != NULL))
     {
       g_propagate_error (error, err);
-      thunar_g_file_list_free (files);
+      thunar_g_list_free_full (files);
       return NULL;
     }
   else if (job != NULL && exo_job_set_error_if_cancelled (EXO_JOB (job), &err))
     {
       g_propagate_error (error, err);
-      thunar_g_file_list_free (files);
+      thunar_g_list_free_full (files);
       return NULL;
     }
 
diff --git a/thunar/thunar-launcher.c b/thunar/thunar-launcher.c
index 13a9f2a5..5ea0c498 100644
--- a/thunar/thunar-launcher.c
+++ b/thunar/thunar-launcher.c
@@ -470,7 +470,7 @@ thunar_launcher_dispose (GObject *object)
   thunar_launcher_set_widget (THUNAR_LAUNCHER (launcher), NULL);
 
   /* disconnect from the currently selected files */
-  thunar_g_file_list_free (launcher->files_to_process);
+  thunar_g_list_free_full (launcher->files_to_process);
   launcher->files_to_process = NULL;
 
   /* unref parent, if any */
@@ -610,7 +610,7 @@ thunar_launcher_set_selected_files (ThunarComponent *component,
 
   /* disconnect from the previous files to process */
   if (launcher->files_to_process != NULL)
-    thunar_g_file_list_free (launcher->files_to_process);
+    thunar_g_list_free_full (launcher->files_to_process);
   launcher->files_to_process = NULL;
 
   /* notify listeners */
@@ -817,7 +817,7 @@ thunar_launcher_open_files (ThunarLauncher *launcher,
   applications = g_hash_table_new_full (thunar_launcher_g_app_info_hash,
                                         (GEqualFunc) g_app_info_equal,
                                         (GDestroyNotify) g_object_unref,
-                                        (GDestroyNotify) thunar_g_file_list_free);
+                                        (GDestroyNotify) thunar_g_list_free_full);
 
   for (lp = files; lp != NULL; lp = lp->next)
     {
@@ -1266,8 +1266,8 @@ thunar_launcher_poke_data_free (ThunarLauncherPokeData *data)
 {
   _thunar_return_if_fail (data != NULL);
 
-  thunar_g_file_list_free (data->files_to_poke);
-  thunar_g_file_list_free (data->files_poked);
+  thunar_g_list_free_full (data->files_to_poke);
+  thunar_g_list_free_full (data->files_poked);
 
   if (data->location_to_poke != NULL)
     g_object_unref (data->location_to_poke);
@@ -1763,7 +1763,7 @@ thunar_launcher_action_sendto_desktop (ThunarLauncher *launcher)
 
   /* cleanup */
   g_object_unref (desktop_file);
-  thunar_g_file_list_free (files);
+  thunar_g_list_free_full (files);
 }
 
 
@@ -1798,7 +1798,7 @@ thunar_launcher_sendto_device (ThunarLauncher *launcher,
     }
 
   /* cleanup */
-  thunar_g_file_list_free (files);
+  thunar_g_list_free_full (files);
 }
 
 
@@ -2120,7 +2120,7 @@ thunar_launcher_action_duplicate (ThunarLauncher *launcher)
       g_object_unref (G_OBJECT (application));
 
       /* clean up */
-      thunar_g_file_list_free (files_to_process);
+      thunar_g_list_free_full (files_to_process);
     }
 }
 
@@ -2689,7 +2689,7 @@ thunar_launcher_create_document_submenu_new (ThunarLauncher *launcher)
   else
     {
       thunar_launcher_create_document_submenu_templates (launcher, submenu, files);
-      thunar_g_file_list_free (files);
+      thunar_g_list_free_full (files);
     }
 
   xfce_gtk_menu_append_seperator (GTK_MENU_SHELL (submenu));
diff --git a/thunar/thunar-location-button.c b/thunar/thunar-location-button.c
index 126ba936..79790314 100644
--- a/thunar/thunar-location-button.c
+++ b/thunar/thunar-location-button.c
@@ -276,7 +276,7 @@ thunar_location_button_finalize (GObject *object)
   ThunarLocationButton *location_button = THUNAR_LOCATION_BUTTON (object);
 
   /* release the drop path list (just in case the drag-leave wasn't fired before) */
-  thunar_g_file_list_free (location_button->drop_file_list);
+  thunar_g_list_free_full (location_button->drop_file_list);
 
   /* be sure to cancel any pending enter timeout */
   if (G_UNLIKELY (location_button->enter_timeout_id != 0))
@@ -721,7 +721,7 @@ thunar_location_button_drag_leave (GtkWidget            *button,
   /* reset the "drop data ready" status and free the path list */
   if (G_LIKELY (location_button->drop_data_ready))
     {
-      thunar_g_file_list_free (location_button->drop_file_list);
+      thunar_g_list_free_full (location_button->drop_file_list);
       location_button->drop_data_ready = FALSE;
       location_button->drop_file_list = NULL;
     }
diff --git a/thunar/thunar-shortcuts-pane.c b/thunar/thunar-shortcuts-pane.c
index e63992d9..34de1902 100644
--- a/thunar/thunar-shortcuts-pane.c
+++ b/thunar/thunar-shortcuts-pane.c
@@ -334,7 +334,7 @@ thunar_shortcuts_pane_set_selected_files (ThunarComponent *component,
   ThunarShortcutsPane *shortcuts_pane = THUNAR_SHORTCUTS_PANE (component);
 
   /* disconnect from the previously selected thunar files... */
-  thunar_g_file_list_free (shortcuts_pane->selected_files);
+  thunar_g_list_free_full (shortcuts_pane->selected_files);
 
   /* ...and take a copy of the newly selected thunar files */
   shortcuts_pane->selected_files = thunar_g_list_copy_deep (selected_files);
diff --git a/thunar/thunar-shortcuts-view.c b/thunar/thunar-shortcuts-view.c
index a76d3ee9..584e2077 100644
--- a/thunar/thunar-shortcuts-view.c
+++ b/thunar/thunar-shortcuts-view.c
@@ -425,7 +425,7 @@ thunar_shortcuts_view_finalize (GObject *object)
   ThunarShortcutsView *view = THUNAR_SHORTCUTS_VIEW (object);
 
   /* release drop path list (if drag_leave wasn't called) */
-  thunar_g_file_list_free (view->drop_file_list);
+  thunar_g_list_free_full (view->drop_file_list);
 
   /* release the provider factory */
   g_object_unref (G_OBJECT (view->provider_factory));
@@ -990,7 +990,7 @@ thunar_shortcuts_view_drag_leave (GtkWidget      *widget,
   /* reset the "drop data ready" status and free the URI list */
   if (G_LIKELY (view->drop_data_ready))
     {
-      thunar_g_file_list_free (view->drop_file_list);
+      thunar_g_list_free_full (view->drop_file_list);
       view->drop_data_ready = FALSE;
       view->drop_file_list = NULL;
     }
diff --git a/thunar/thunar-standard-view.c b/thunar/thunar-standard-view.c
index dfa5e99b..1c090e4c 100644
--- a/thunar/thunar-standard-view.c
+++ b/thunar/thunar-standard-view.c
@@ -887,13 +887,13 @@ thunar_standard_view_finalize (GObject *object)
     g_object_unref (G_OBJECT (standard_view->priv->scroll_to_file));
 
   /* release the selected_files list (if any) */
-  thunar_g_file_list_free (standard_view->priv->selected_files);
+  thunar_g_list_free_full (standard_view->priv->selected_files);
 
   /* release the drag path list (just in case the drag-end wasn't fired before) */
-  thunar_g_file_list_free (standard_view->priv->drag_g_file_list);
+  thunar_g_list_free_full (standard_view->priv->drag_g_file_list);
 
   /* release the drop path list (just in case the drag-leave wasn't fired before) */
-  thunar_g_file_list_free (standard_view->priv->drop_file_list);
+  thunar_g_list_free_full (standard_view->priv->drop_file_list);
 
   /* release the history */
   g_object_unref (standard_view->priv->history);
@@ -913,7 +913,7 @@ thunar_standard_view_finalize (GObject *object)
     }
 
   /* drop any remaining "new-files" paths */
-  thunar_g_file_list_free (standard_view->priv->new_files_path_list);
+  thunar_g_list_free_full (standard_view->priv->new_files_path_list);
 
   /* release our reference on the preferences */
   g_object_unref (G_OBJECT (standard_view->preferences));
@@ -1165,7 +1165,7 @@ thunar_standard_view_set_selected_files_component (ThunarComponent *component,
   /* release the previous selected files list (if any) */
   if (G_UNLIKELY (standard_view->priv->selected_files != NULL))
     {
-      thunar_g_file_list_free (standard_view->priv->selected_files);
+      thunar_g_list_free_full (standard_view->priv->selected_files);
       standard_view->priv->selected_files = NULL;
     }
 
@@ -1516,7 +1516,7 @@ thunar_standard_view_set_loading (ThunarStandardView *standard_view,
       thunar_standard_view_new_files (standard_view, new_files_path_list);
 
       /* cleanup */
-      thunar_g_file_list_free (new_files_path_list);
+      thunar_g_list_free_full (new_files_path_list);
     }
 
   /* check if we're done loading */
@@ -1530,7 +1530,7 @@ thunar_standard_view_set_loading (ThunarStandardView *standard_view,
       thunar_component_set_selected_files (THUNAR_COMPONENT (standard_view), selected_files);
 
       /* cleanup */
-      thunar_g_file_list_free (selected_files);
+      thunar_g_list_free_full (selected_files);
     }
 
   /* check if we're done loading and a thumbnail timeout or idle was requested */
@@ -2321,7 +2321,7 @@ thunar_standard_view_new_files (ThunarStandardView *standard_view,
   /* release the previous "new-files" paths (if any) */
   if (G_UNLIKELY (standard_view->priv->new_files_path_list != NULL))
     {
-      thunar_g_file_list_free (standard_view->priv->new_files_path_list);
+      thunar_g_list_free_full (standard_view->priv->new_files_path_list);
       standard_view->priv->new_files_path_list = NULL;
     }
 
@@ -2850,7 +2850,7 @@ thunar_standard_view_drag_leave (GtkWidget          *widget,
   /* reset the "drop data ready" status and free the URI list */
   if (G_LIKELY (standard_view->priv->drop_data_ready))
     {
-      thunar_g_file_list_free (standard_view->priv->drop_file_list);
+      thunar_g_list_free_full (standard_view->priv->drop_file_list);
       standard_view->priv->drop_file_list = NULL;
       standard_view->priv->drop_data_ready = FALSE;
     }
@@ -2959,7 +2959,7 @@ thunar_standard_view_drag_begin (GtkWidget          *view,
   gint        size;
 
   /* release the drag path list (just in case the drag-end wasn't fired before) */
-  thunar_g_file_list_free (standard_view->priv->drag_g_file_list);
+  thunar_g_list_free_full (standard_view->priv->drag_g_file_list);
 
   /* query the list of selected URIs */
   standard_view->priv->drag_g_file_list = thunar_file_list_to_thunar_g_file_list (standard_view->priv->selected_files);
@@ -3025,7 +3025,7 @@ thunar_standard_view_drag_end (GtkWidget          *view,
     g_source_remove (standard_view->priv->drag_scroll_timer_id);
 
   /* release the list of dragged URIs */
-  thunar_g_file_list_free (standard_view->priv->drag_g_file_list);
+  thunar_g_list_free_full (standard_view->priv->drag_g_file_list);
   standard_view->priv->drag_g_file_list = NULL;
 }
 
@@ -3746,7 +3746,7 @@ thunar_standard_view_selection_changed (ThunarStandardView *standard_view)
     }
 
   /* release the previously selected files */
-  thunar_g_file_list_free (standard_view->priv->selected_files);
+  thunar_g_list_free_full (standard_view->priv->selected_files);
 
   /* determine the new list of selected files (replacing GtkTreePath's with ThunarFile's) */
   selected_thunar_files = (*THUNAR_STANDARD_VIEW_GET_CLASS (standard_view)->get_selected_items) (standard_view);
diff --git a/thunar/thunar-transfer-job.c b/thunar/thunar-transfer-job.c
index 527f4d4f..783f2dc5 100644
--- a/thunar/thunar-transfer-job.c
+++ b/thunar/thunar-transfer-job.c
@@ -203,7 +203,7 @@ thunar_transfer_job_finalize (GObject *object)
   if (job->target_device_fs_id != NULL)
     g_free (job->target_device_fs_id);
 
-  thunar_g_file_list_free (job->target_file_list);
+  thunar_g_list_free_full (job->target_file_list);
 
   g_object_unref (job->preferences);
 
@@ -384,7 +384,7 @@ thunar_transfer_job_collect_node (ThunarTransferJob  *job,
         }
 
       /* release the child files */
-      thunar_g_file_list_free (file_list);
+      thunar_g_list_free_full (file_list);
     }
 
   /* release file info */
@@ -1598,7 +1598,7 @@ thunar_transfer_job_execute (ExoJob  *job,
   else
     {
       thunar_job_new_files (THUNAR_JOB (job), new_files_list);
-      thunar_g_file_list_free (new_files_list);
+      thunar_g_list_free_full (new_files_list);
       return TRUE;
     }
 }
diff --git a/thunar/thunar-tree-view.c b/thunar/thunar-tree-view.c
index f16d1aed..898b7ea7 100644
--- a/thunar/thunar-tree-view.c
+++ b/thunar/thunar-tree-view.c
@@ -418,7 +418,7 @@ thunar_tree_view_finalize (GObject *object)
   ThunarTreeView *view = THUNAR_TREE_VIEW (object);
 
   /* release drop path list (if drag_leave wasn't called) */
-  thunar_g_file_list_free (view->drop_file_list);
+  thunar_g_list_free_full (view->drop_file_list);
 
   /* release the provider factory */
   g_object_unref (G_OBJECT (view->provider_factory));
@@ -1051,7 +1051,7 @@ thunar_tree_view_drag_leave (GtkWidget      *widget,
   /* reset the "drop data ready" status and free the URI list */
   if (G_LIKELY (view->drop_data_ready))
     {
-      thunar_g_file_list_free (view->drop_file_list);
+      thunar_g_list_free_full (view->drop_file_list);
       view->drop_data_ready = FALSE;
       view->drop_file_list = NULL;
     }
diff --git a/thunar/thunar-window.c b/thunar/thunar-window.c
index 1a9ee987..36ac4195 100644
--- a/thunar/thunar-window.c
+++ b/thunar/thunar-window.c
@@ -3117,7 +3117,7 @@ thunar_window_replace_view (ThunarWindow *window,
 
   /* restore the file selection */
   thunar_component_set_selected_files (THUNAR_COMPONENT (new_view), selected_thunar_files);
-  thunar_g_file_list_free (selected_thunar_files);
+  thunar_g_list_free_full (selected_thunar_files);
 
   /* remember the last view type if this is the active view and directory specific settings are not enabled */
   if (is_current_view && !window->directory_specific_settings && gtk_widget_get_visible (GTK_WIDGET (window)) && view_type != G_TYPE_NONE)
