diff --git a/thunar/thunar-preferences.c b/thunar/thunar-preferences.c
index 25593812..26b8ef07 100644
--- a/thunar/thunar-preferences.c
+++ b/thunar/thunar-preferences.c
@@ -104,6 +104,7 @@ enum
   PROP_SHORTCUTS_ICON_SIZE,
   PROP_TREE_ICON_EMBLEMS,
   PROP_TREE_ICON_SIZE,
+  PROP_MISC_SWITCH_TO_NEW_TAB,
   N_PROPERTIES,
 };
 
@@ -870,6 +871,18 @@ thunar_preferences_class_init (ThunarPreferencesClass *klass)
                          THUNAR_ICON_SIZE_16,
                          EXO_PARAM_READWRITE);
 
+  /**
+ * ThunarPreferences:misc-switch-to-new-tab:
+ *
+ * Whether to switch to the new tab after opening a folder in a new tab.
+ **/
+  preferences_props[PROP_MISC_SWITCH_TO_NEW_TAB] =
+      g_param_spec_boolean ("misc-switch-to-new-tab",
+                            "SwitchToNewTab",
+                            NULL,
+                            TRUE,
+                            EXO_PARAM_READWRITE);
+
   /* install all properties */
   g_object_class_install_properties (gobject_class, N_PROPERTIES, preferences_props);
 }
diff --git a/thunar/thunar-window.c b/thunar/thunar-window.c
index 22526d77..e8e583b2 100644
--- a/thunar/thunar-window.c
+++ b/thunar/thunar-window.c
@@ -2259,6 +2259,7 @@ void thunar_window_notebook_open_new_tab (ThunarWindow *window,
   GtkWidget     *view;
   gint           page_num;
   GType          view_type;
+  gboolean       switch_to_new_tab;
 
   /* save the history of the current view */
   if (THUNAR_IS_STANDARD_VIEW (window->view))
@@ -2272,8 +2273,12 @@ void thunar_window_notebook_open_new_tab (ThunarWindow *window,
   view = thunar_window_notebook_insert_page (window, directory, view_type, page_num + 1, history);
 
   /* switch to the new view */
-  page_num = gtk_notebook_page_num (GTK_NOTEBOOK (window->notebook_selected), view);
-  gtk_notebook_set_current_page (GTK_NOTEBOOK (window->notebook_selected), page_num);
+  g_object_get (G_OBJECT (window->preferences), "misc-switch-to-new-tab", &switch_to_new_tab, NULL);
+  if (switch_to_new_tab == TRUE)
+    {
+      page_num = gtk_notebook_page_num (GTK_NOTEBOOK (window->notebook_selected), view);
+      gtk_notebook_set_current_page (GTK_NOTEBOOK (window->notebook_selected), page_num);
+    }
 
   /* take focus on the new view */
   gtk_widget_grab_focus (view);
