diff --git a/thunar/thunar-location-bar.c b/thunar/thunar-location-bar.c
index 8e728340..c069e02a 100644
--- a/thunar/thunar-location-bar.c
+++ b/thunar/thunar-location-bar.c
@@ -72,7 +72,6 @@ static void         thunar_location_bar_set_current_directory      (ThunarNaviga
                                                                     ThunarFile           *current_directory);
 static GtkWidget   *thunar_location_bar_install_widget             (ThunarLocationBar    *bar,
                                                                     GType                 type);
-static void         thunar_location_bar_reload_requested           (ThunarLocationBar    *bar);
 static void         thunar_location_bar_settings_changed           (ThunarLocationBar    *bar);
 static void         thunar_location_bar_on_enry_edit_done          (ThunarLocationEntry  *entry,
                                                                     ThunarLocationBar    *bar);
@@ -108,20 +107,6 @@ thunar_location_bar_class_init (ThunarLocationBarClass *klass)
 
   /* install signals */
 
-  /**
-   * ThunarLocationBar::reload-requested:
-   * @location_bar : a #ThunarLocationBar.
-   *
-   * Emitted by @location_bar whenever the user clicked a "reload" button
-   **/
-  g_signal_new ("reload-requested",
-                G_TYPE_FROM_CLASS (klass),
-                G_SIGNAL_RUN_LAST | G_SIGNAL_ACTION,
-                G_STRUCT_OFFSET (ThunarLocationBarClass, reload_requested),
-                NULL, NULL,
-                NULL,
-                G_TYPE_NONE, 0);
-
   /**
    * ThunarLocationBar::entry-done:
    * @location_bar : a #ThunarLocationBar.
@@ -252,14 +237,6 @@ thunar_location_bar_set_current_directory (ThunarNavigator      *navigator,
 
 
 
-static void
-thunar_location_bar_reload_requested (ThunarLocationBar *bar)
-{
-  g_signal_emit_by_name (bar, "reload-requested");
-}
-
-
-
 static GtkWidget *
 thunar_location_bar_install_widget (ThunarLocationBar    *bar,
                                     GType                 type)
@@ -276,7 +253,6 @@ thunar_location_bar_install_widget (ThunarLocationBar    *bar,
         {
           bar->locationEntry = gtk_widget_new (THUNAR_TYPE_LOCATION_ENTRY, "current-directory", NULL, NULL);
           g_object_ref (bar->locationEntry);
-          g_signal_connect_swapped (bar->locationEntry, "reload-requested", G_CALLBACK (thunar_location_bar_reload_requested), bar);
           g_signal_connect_swapped (bar->locationEntry, "change-directory", G_CALLBACK (thunar_navigator_change_directory), THUNAR_NAVIGATOR (bar));
           g_signal_connect_swapped (bar->locationEntry, "open-new-tab", G_CALLBACK (thunar_navigator_open_new_tab), THUNAR_NAVIGATOR (bar));
         }
diff --git a/thunar/thunar-location-entry.c b/thunar/thunar-location-entry.c
index bd9406ea..d9a695af 100644
--- a/thunar/thunar-location-entry.c
+++ b/thunar/thunar-location-entry.c
@@ -66,10 +66,6 @@ static gboolean    thunar_location_entry_button_press_event       (GtkWidget
                                                                    GdkEventButton           *event,
                                                                    ThunarLocationEntry      *location_entry);
 static gboolean    thunar_location_entry_reset                    (ThunarLocationEntry      *location_entry);
-static void        thunar_location_entry_reload                   (GtkEntry                 *entry,
-                                                                   GtkEntryIconPosition      icon_pos,
-                                                                   GdkEvent                 *event,
-                                                                   ThunarLocationEntry      *location_entry);
 static void        thunar_location_entry_emit_edit_done           (ThunarLocationEntry      *entry);
 
 
@@ -136,19 +132,7 @@ thunar_location_entry_class_init (ThunarLocationEntryClass *klass)
                 _thunar_marshal_BOOLEAN__VOID,
                 G_TYPE_BOOLEAN, 0);
 
-  /**
-   * ThunarLocationEntry::reload-requested:
-   * @location_entry : a #ThunarLocationEntry.
-   *
-   * Emitted by @location_entry whenever the user clicked a "reload" button
-   **/
-  g_signal_new ("reload-requested",
-                G_TYPE_FROM_CLASS (klass),
-                G_SIGNAL_RUN_LAST | G_SIGNAL_ACTION,
-                G_STRUCT_OFFSET (ThunarLocationEntryClass, reload_requested),
-                NULL, NULL,
-                NULL,
-                G_TYPE_NONE, 0);
+
 
   /**
    * ThunarLocationEntry::edit-done:
@@ -193,14 +177,6 @@ thunar_location_entry_init (ThunarLocationEntry *location_entry)
   gtk_box_pack_start (GTK_BOX (location_entry), location_entry->path_entry, TRUE, TRUE, 0);
   gtk_widget_show (location_entry->path_entry);
 
-  /* put reload button in entry */
-  gtk_entry_set_icon_from_icon_name (GTK_ENTRY (location_entry->path_entry),
-                                     GTK_ENTRY_ICON_SECONDARY, "view-refresh-symbolic");
-  gtk_entry_set_icon_tooltip_text (GTK_ENTRY (location_entry->path_entry),
-                                   GTK_ENTRY_ICON_SECONDARY, _("Reload the current folder"));
-  g_signal_connect (G_OBJECT (location_entry->path_entry), "icon-release",
-                    G_CALLBACK (thunar_location_entry_reload), location_entry);
-
   /* make sure the edit-done signal is emitted upon moving the focus somewhere else */
   g_signal_connect_swapped (location_entry->path_entry, "focus-out-event", G_CALLBACK (thunar_location_entry_emit_edit_done), location_entry);
 
@@ -454,22 +430,6 @@ thunar_location_entry_reset (ThunarLocationEntry *location_entry)
 
 
 
-static void
-thunar_location_entry_reload (GtkEntry            *entry,
-                              GtkEntryIconPosition icon_pos,
-                              GdkEvent            *event,
-                              ThunarLocationEntry *location_entry)
-{
-  _thunar_return_if_fail (THUNAR_IS_LOCATION_ENTRY (location_entry));
-
-  if (icon_pos == GTK_ENTRY_ICON_SECONDARY)
-    {
-      g_signal_emit_by_name (location_entry, "reload-requested");
-    }
-}
-
-
-
 static void
 thunar_location_entry_emit_edit_done (ThunarLocationEntry *entry)
 {
diff --git a/thunar/thunar-window.c b/thunar/thunar-window.c
index 95bf2e8b..89b2eb08 100644
--- a/thunar/thunar-window.c
+++ b/thunar/thunar-window.c
@@ -169,7 +169,6 @@ static void      thunar_window_paned_notebooks_indicate_focus (ThunarWindow
 static gboolean  thunar_window_split_view_is_active       (ThunarWindow           *window);
 
 static void      thunar_window_update_location_bar_visible(ThunarWindow           *window);
-static void      thunar_window_handle_reload_request      (ThunarWindow           *window);
 static void      thunar_window_install_sidepane           (ThunarWindow           *window,
                                                            GType                   type);
 static void      thunar_window_start_open_location        (ThunarWindow           *window,
@@ -361,6 +360,7 @@ struct _ThunarWindow
   GtkWidget              *location_toolbar_item_back;
   GtkWidget              *location_toolbar_item_forward;
   GtkWidget              *location_toolbar_item_parent;
+  GtkWidget              *location_toolbar_reload;
 
   ThunarLauncher         *launcher;
 
@@ -836,7 +836,6 @@ thunar_window_init (ThunarWindow *window)
   g_object_bind_property (G_OBJECT (window), "current-directory", G_OBJECT (window->location_bar), "current-directory", G_BINDING_SYNC_CREATE);
   g_signal_connect_swapped (G_OBJECT (window->location_bar), "change-directory", G_CALLBACK (thunar_window_set_current_directory), window);
   g_signal_connect_swapped (G_OBJECT (window->location_bar), "open-new-tab", G_CALLBACK (thunar_window_notebook_open_new_tab), window);
-  g_signal_connect_swapped (G_OBJECT (window->location_bar), "reload-requested", G_CALLBACK (thunar_window_handle_reload_request), window);
   g_signal_connect_swapped (G_OBJECT (window->location_bar), "entry-done", G_CALLBACK (thunar_window_update_location_bar_visible), window);
 
   /* setup the toolbar for the location bar */
@@ -869,6 +868,8 @@ thunar_window_init (ThunarWindow *window)
   /* add the location bar itself */
   gtk_container_add (GTK_CONTAINER (tool_item), window->location_bar);
 
+  window->location_toolbar_reload = xfce_gtk_tool_button_new_from_action_entry (get_action_entry (THUNAR_WINDOW_ACTION_RELOAD), G_OBJECT (window), GTK_TOOLBAR (window->location_toolbar));
+
   /* display the toolbar */
   gtk_widget_show_all (window->location_toolbar);
 
@@ -2387,17 +2388,6 @@ thunar_window_update_window_icon (ThunarWindow *window)
 
 
 
-static void
-thunar_window_handle_reload_request (ThunarWindow *window)
-{
-  gboolean result;
-
-  /* force the view to reload */
-  g_signal_emit (G_OBJECT (window), window_signals[RELOAD], 0, TRUE, &result);
-}
-
-
-
 static void
 thunar_window_install_sidepane (ThunarWindow *window,
                                 GType         type)
