diff --git a/thunar/thunar-window.c b/thunar/thunar-window.c
index f903a331..456d7d44 100644
--- a/thunar/thunar-window.c
+++ b/thunar/thunar-window.c
@@ -161,7 +161,7 @@ static void      thunar_window_notebook_select_current_page(ThunarWindow
 
 static GtkWidget*thunar_window_paned_notebooks_add        (ThunarWindow           *window);
 static void      thunar_window_paned_notebooks_switch     (ThunarWindow           *window);
-static gboolean  thunar_window_paned_notebooks_select     (GtkWidget              *notebook,
+static gboolean  thunar_window_paned_notebooks_select     (GtkWidget              *view,
                                                            GtkDirectionType       *direction,
                                                            ThunarWindow           *window);
 static void      thunar_window_paned_notebooks_indicate_focus (ThunarWindow       *window,
@@ -947,13 +947,13 @@ thunar_window_screen_changed (GtkWidget *widget,
 /**
  * thunar_window_select_files:
  * @window            : a #ThunarWindow instance.
- * @files_to_selected : a list of #GFile<!---->s
+ * @files_to_select   : a list of #GFile<!---->s
  *
  * Visually selects the files, given by the list
  **/
 void
 thunar_window_select_files (ThunarWindow *window,
-                            GList        *files_to_selected)
+                            GList        *files_to_select)
 {
   GList        *thunar_files = NULL;
   ThunarFolder *thunar_folder;
@@ -968,7 +968,7 @@ thunar_window_select_files (ThunarWindow *window,
       g_object_unref (thunar_folder);
     }
 
-  for (GList *lp = files_to_selected; lp != NULL; lp = lp->next)
+  for (GList *lp = files_to_select; lp != NULL; lp = lp->next)
     thunar_files = g_list_append (thunar_files, thunar_file_get (G_FILE (lp->data), NULL));
   thunar_view_set_selected_files (THUNAR_VIEW (window->view), thunar_files);
   g_list_free_full (thunar_files, g_object_unref);
diff --git a/thunar/thunar-window.h b/thunar/thunar-window.h
index e155a5a4..4b1237a7 100644
--- a/thunar/thunar-window.h
+++ b/thunar/thunar-window.h
@@ -130,8 +130,8 @@ void                      thunar_window_redirect_menu_tooltips_to_statusbar (Thu
                                                                              GtkMenu             *menu);
 const XfceGtkActionEntry* thunar_window_get_action_entry                    (ThunarWindow        *window,
                                                                              ThunarWindowAction   action);
- void                     thunar_window_select_files                        (ThunarWindow         *window,
-                                                                             GList                *path_list);
+ void                     thunar_window_select_files                        (ThunarWindow        *window,
+                                                                             GList               *files_to_select);
 G_END_DECLS;
 
 #endif /* !__THUNAR_WINDOW_H__ */
