diff --git a/NEWS b/NEWS
index 999e9785..ef10900d 100644
--- a/NEWS
+++ b/NEWS
@@ -1,3 +1,9 @@
+4.17.3
+======
+- Regression: Activating Desktop Icon does not Use Default Application 
+(Issue #575)
+
+
 4.17.2
 ======
 - Prevent crash if non-existing bookmark is right-clicked (Issue #552)
diff --git a/configure.ac.in b/configure.ac.in
index e23ea869..035c26d2 100644
--- a/configure.ac.in
+++ b/configure.ac.in
@@ -26,10 +26,10 @@ m4_define([thunarx_verinfo], [0:0:0])
 m4_define([thunarx_version_api], [3])
 m4_define([thunar_version_major], [4])
 m4_define([thunar_version_minor], [17])
-m4_define([thunar_version_micro], [2])
+m4_define([thunar_version_micro], [3])
 m4_define([thunar_version_nano], [])
 m4_define([thunar_version_build], [@REVISION@])
-m4_define([thunar_version_tag], [git])
+m4_define([thunar_version_tag], [])
 m4_define([thunar_version], [thunar_version_major().thunar_version_minor().thunar_version_micro()ifelse(thunar_version_nano(), [], [], [.thunar_version_nano()])ifelse(thunar_version_tag(), [git], [thunar_version_tag()-thunar_version_build()], [thunar_version_tag()])])
 
 dnl *******************************************
