diff --git a/thunar/thunar-history.c b/thunar/thunar-history.c
index 324235e5..aa81d931 100644
--- a/thunar/thunar-history.c
+++ b/thunar/thunar-history.c
@@ -729,7 +729,7 @@ thunar_history_peek_forward (ThunarHistory *history)
 /**
  * thunar_history_add:
  * @history : a #ThunarHistory
- * @history : a #ThunarFile
+ * @directory : a #ThunarFile
  *
  * Manually adds the passed directory to the history
  **/
