diff --git a/thunar/thunar-transfer-job.c b/thunar/thunar-transfer-job.c
index 2c832928..5baa48bf 100644
--- a/thunar/thunar-transfer-job.c
+++ b/thunar/thunar-transfer-job.c
@@ -411,10 +411,11 @@ ttj_copy_file (ThunarTransferJob *job,
                gboolean           merge_directories,
                GError           **error)
 {
-  GFileType source_type;
-  GFileType target_type;
-  gboolean  target_exists;
-  GError   *err = NULL;
+  GFileInfo *info;
+  GFileType  source_type;
+  GFileType  target_type;
+  gboolean   target_exists;
+  GError    *err = NULL;
 
   _thunar_return_val_if_fail (THUNAR_IS_TRANSFER_JOB (job), FALSE);
   _thunar_return_val_if_fail (G_IS_FILE (source_file), FALSE);
@@ -458,6 +459,26 @@ ttj_copy_file (ThunarTransferJob *job,
                exo_job_get_cancellable (EXO_JOB (job)),
                thunar_transfer_job_progress, job, &err);
 
+  /**
+   * MR !127 notes:
+   * (Discusssion: https://gitlab.xfce.org/xfce/thunar/-/merge_requests/127)
+   *
+   * Proper solution for copying lost metadata requires
+   * `g_file_build_attribute_list_for_copy`, which is
+   * added in Glib 2.68 (which is too recent a version
+   * to add for now) so we leave a comment here. Patch
+   * this when 2.68 is ready.
+   **/
+  if (G_UNLIKELY (err == NULL && !g_file_is_native (source_file)))
+    {
+      info = g_file_query_info (source_file, G_FILE_ATTRIBUTE_TIME_MODIFIED, G_FILE_QUERY_INFO_NONE,
+                                exo_job_get_cancellable (EXO_JOB (job)), &err);
+
+      g_file_set_attributes_from_info (target_file, info, G_FILE_QUERY_INFO_NONE,
+                                       exo_job_get_cancellable (EXO_JOB (job)), &err);
+      g_clear_object (&info);
+    }
+
   /* check if there were errors */
   if (G_UNLIKELY (err != NULL && err->domain == G_IO_ERROR))
     {
