diff --git a/thunar/thunar-window.c b/thunar/thunar-window.c
index d33aa77f..abe11fd4 100644
--- a/thunar/thunar-window.c
+++ b/thunar/thunar-window.c
@@ -2857,7 +2857,11 @@ static void
 thunar_window_action_close_window (ThunarWindow *window,
                                    GtkWidget    *menu_item)
 {
-  gtk_widget_destroy (GTK_WIDGET (window));
+  gboolean response;
+
+  response = thunar_window_delete (GTK_WIDGET (window), NULL, NULL);
+  if (response == FALSE)
+    gtk_widget_destroy (GTK_WIDGET (window));
 }
 
 
