diff --git a/thunar/thunar-location-entry.c b/thunar/thunar-location-entry.c
index bdda01ab..7a81cce3 100644
--- a/thunar/thunar-location-entry.c
+++ b/thunar/thunar-location-entry.c
@@ -479,10 +479,6 @@ thunar_location_entry_emit_edit_done (ThunarLocationEntry *entry)
 void
 thunar_location_entry_cancel_search (ThunarLocationEntry *entry)
 {
-  ThunarFolder *folder = thunar_folder_get_for_file (thunar_navigator_get_current_directory (THUNAR_NAVIGATOR (entry)));
-  thunar_folder_reload (folder, TRUE);
-  g_object_unref (folder);
-
   entry->is_searching = FALSE;
   entry->right_click_occurred = FALSE;
   thunar_location_entry_emit_edit_done (entry);
