diff --git a/thunar/thunar-util.c b/thunar/thunar-util.c
index c57c5e77..b9d2560c 100644
--- a/thunar/thunar-util.c
+++ b/thunar/thunar-util.c
@@ -710,7 +710,7 @@ thunar_util_next_new_file_name (ThunarFile            *dir,
           ThunarFile  *file = files->data;
           const gchar *name = thunar_file_get_display_name (file);
 
-          if (strcmp (new_name, name) == 0)
+          if (g_strcmp0 (new_name, name) == 0)
             {
               found_duplicate = TRUE;
               break;
