diff --git a/thunar/thunar-gtk-extensions.c b/thunar/thunar-gtk-extensions.c
index 6d2f4b0b..f1a18e65 100644
--- a/thunar/thunar-gtk-extensions.c
+++ b/thunar/thunar-gtk-extensions.c
@@ -125,8 +125,7 @@ thunar_gtk_menu_thunarx_menu_item_new (GObject      *thunarx_menu_item,
   g_free (accel_path);
   g_free (label_text);
   g_free (tooltip_text);
-  if (icon_name != NULL)
-    g_free (icon_name);
+  g_free (icon_name);
   if (icon != NULL)
     g_object_unref (icon);
 
diff --git a/thunar/thunar-transfer-job.c b/thunar/thunar-transfer-job.c
index ac70db44..f0e35d12 100644
--- a/thunar/thunar-transfer-job.c
+++ b/thunar/thunar-transfer-job.c
@@ -238,10 +238,8 @@ thunar_transfer_job_finalize (GObject *object)
 
   g_list_free_full (job->source_node_list, thunar_transfer_node_free);
 
-  if (job->source_device_fs_id != NULL)
-    g_free (job->source_device_fs_id);
-  if (job->target_device_fs_id != NULL)
-    g_free (job->target_device_fs_id);
+  g_free (job->source_device_fs_id);
+  g_free (job->target_device_fs_id);
 
   thunar_g_list_free_full (job->target_file_list);
 
