diff --git a/thunar/thunar-file.c b/thunar/thunar-file.c
index af26a643..4755565c 100644
--- a/thunar/thunar-file.c
+++ b/thunar/thunar-file.c
@@ -1048,11 +1048,11 @@ thunar_file_info_reload (ThunarFile   *file,
             }
           else
             {
-              /* drop any suffix (e.g. '.png') from themed icons */
+              /* drop freedesktop.org supported suffixes from themed icons, if any */
               if (!g_path_is_absolute (file->custom_icon_name))
                 {
                   p = strrchr (file->custom_icon_name, '.');
-                  if (p != NULL)
+                  if(g_strcmp0(p, ".png") == 0 || g_strcmp0(p, ".xpm") == 0 || g_strcmp0(p, ".svg") == 0)
                     *p = '\0';
                 }
             }
