diff --git a/thunar/thunar-list-model.c b/thunar/thunar-list-model.c
index a54a8cf4..1c330b46 100644
--- a/thunar/thunar-list-model.c
+++ b/thunar/thunar-list-model.c
@@ -337,7 +337,7 @@ thunar_list_model_class_init (ThunarListModelClass *klass)
       g_param_spec_string ("date-custom-style",
                            "DateCustomStyle",
                            NULL,
-                           "%Y-%mutex_files_to_add-%d %H:%M:%S",
+                           "%Y-%m-%d %H:%M:%S",
                            EXO_PARAM_READWRITE);
 
   /**
