diff --git a/configure.ac.in b/configure.ac.in
index 5e3cc6a7..9769e36d 100644
--- a/configure.ac.in
+++ b/configure.ac.in
@@ -151,7 +151,7 @@ XDT_CHECK_PACKAGE([GTHREAD], [gthread-2.0], [2.56.0])
 XDT_CHECK_PACKAGE([GMODULE], [gmodule-2.0], [2.56.0])
 XDT_CHECK_PACKAGE([GTK], [gtk+-3.0], [3.22.0])
 XDT_CHECK_PACKAGE([GDK_PIXBUF], [gdk-pixbuf-2.0], [2.14.0])
-XDT_CHECK_PACKAGE([LIBXFCE4UTIL], [libxfce4util-1.0], [4.17.1])
+XDT_CHECK_PACKAGE([LIBXFCE4UTIL], [libxfce4util-1.0], [4.17.2])
 XDT_CHECK_PACKAGE([LIBXFCE4UI], [libxfce4ui-2], [4.17.3])
 XDT_CHECK_PACKAGE([LIBXFCE4KBD_PRIVATE], [libxfce4kbd-private-3], [4.17.2])
 XDT_CHECK_PACKAGE([XFCONF], [libxfconf-0], [4.12.0])
diff --git a/plugins/thunar-sendto-email/Makefile.am b/plugins/thunar-sendto-email/Makefile.am
index 379a5428..7870ae80 100644
--- a/plugins/thunar-sendto-email/Makefile.am
+++ b/plugins/thunar-sendto-email/Makefile.am
@@ -18,10 +18,12 @@ thunar_sendto_email_SOURCES =						\
 
 thunar_sendto_email_CFLAGS =						\
 	$(EXO_CFLAGS)							\
+	$(LIBXFCE4UTIL_CFLAGS)						\
 	$(PLATFORM_CFLAGS)
 
 thunar_sendto_email_LDADD =						\
 	$(EXO_LIBS)							\
+	$(LIBXFCE4UTIL_LIBS)						\
 	$(PLATFORM_LDFLAGS)
 
 desktopdir = $(datadir)/Thunar/sendto
diff --git a/plugins/thunar-sendto-email/main.c b/plugins/thunar-sendto-email/main.c
index c2d1a511..82438e57 100644
--- a/plugins/thunar-sendto-email/main.c
+++ b/plugins/thunar-sendto-email/main.c
@@ -62,6 +62,8 @@
 
 #include <exo/exo.h>
 
+#include <libxfce4util/libxfce4util.h>
+
 
 
 typedef struct _TseData TseData;
diff --git a/plugins/thunar-uca/thunar-uca-model.c b/plugins/thunar-uca/thunar-uca-model.c
index 69091ebb..80dde184 100644
--- a/plugins/thunar-uca/thunar-uca-model.c
+++ b/plugins/thunar-uca/thunar-uca-model.c
@@ -49,8 +49,6 @@
 
 #include <gtk/gtk.h>
 
-#include <exo/exo.h>
-
 #include <libxfce4util/libxfce4util.h>
 
 #include <thunar-uca/thunar-uca-model.h>
@@ -877,7 +875,7 @@ end_element_handler (GMarkupParseContext *context,
                                    0, 0);
 
           /* check if a new id should've been generated */
-          if (exo_str_is_empty (parser->unique_id->str))
+          if (xfce_str_is_empty (parser->unique_id->str))
             parser->unique_id_generated = TRUE;
         }
       else
diff --git a/thunar/Makefile.am b/thunar/Makefile.am
index e2c554f0..6f9c5025 100644
--- a/thunar/Makefile.am
+++ b/thunar/Makefile.am
@@ -211,6 +211,7 @@ thunar_CFLAGS =								\
 	$(LIBNOTIFY_CFLAGS)						\
 	$(LIBSM_CFLAGS)							\
 	$(LIBXFCE4UI_CFLAGS)						\
+	$(LIBXFCE4UTIL_CFLAGS)						\
 	$(LIBXFCE4KBD_PRIVATE_CFLAGS) \
 	$(XFCONF_CFLAGS)						\
 	$(PANGO_CFLAGS)							\
@@ -230,6 +231,7 @@ thunar_LDADD =								\
 	$(LIBNOTIFY_LIBS)						\
 	$(LIBSM_LIBS)							\
 	$(LIBXFCE4UI_LIBS)						\
+	$(LIBXFCE4UTIL_LIBS)						\
 	$(LIBXFCE4KBD_PRIVATE_LIBS) \
 	$(XFCONF_LIBS)							\
 	$(PANGO_LIBS)
diff --git a/thunar/thunar-browser.c b/thunar/thunar-browser.c
index da2cfec4..db3520fe 100644
--- a/thunar/thunar-browser.c
+++ b/thunar/thunar-browser.c
@@ -24,6 +24,8 @@
 
 #include <glib-object.h>
 
+#include <libxfce4util/libxfce4util.h>
+
 #include <thunar/thunar-browser.h>
 #include <thunar/thunar-file.h>
 #include <thunar/thunar-private.h>
diff --git a/thunar/thunar-clipboard-manager.c b/thunar/thunar-clipboard-manager.c
index 7928ca52..cae6c4e1 100644
--- a/thunar/thunar-clipboard-manager.c
+++ b/thunar/thunar-clipboard-manager.c
@@ -30,6 +30,8 @@
 #include <string.h>
 #endif
 
+#include <libxfce4util/libxfce4util.h>
+
 #include <thunar/thunar-application.h>
 #include <thunar/thunar-clipboard-manager.h>
 #include <thunar/thunar-dialogs.h>
diff --git a/thunar/thunar-column-model.c b/thunar/thunar-column-model.c
index e00a5eda..92fcdeba 100644
--- a/thunar/thunar-column-model.c
+++ b/thunar/thunar-column-model.c
@@ -25,6 +25,8 @@
 #include <stdlib.h>
 #endif
 
+#include <libxfce4util/libxfce4util.h>
+
 #include <thunar/thunar-column-model.h>
 #include <thunar/thunar-preferences.h>
 #include <thunar/thunar-private.h>
diff --git a/thunar/thunar-dbus-service.c b/thunar/thunar-dbus-service.c
index b0468487..e516fef8 100644
--- a/thunar/thunar-dbus-service.c
+++ b/thunar/thunar-dbus-service.c
@@ -32,7 +32,7 @@
 
 #include <glib/gstdio.h>
 
-#include <exo/exo.h>
+#include <libxfce4util/libxfce4util.h>
 
 #include <thunar/thunar-application.h>
 #include <thunar/thunar-chooser-dialog.h>
@@ -1185,7 +1185,7 @@ thunar_dbus_service_transfer_files (ThunarDBusTransferMode transfer_mode,
   if (screen != NULL)
     {
       /* change the working directory if necessary */
-      if (!exo_str_is_empty (working_directory))
+      if (!xfce_str_is_empty (working_directory))
         old_working_dir = thunar_util_change_working_directory (working_directory);
 
       /* transform the source filenames into GFile objects */
@@ -1215,7 +1215,7 @@ thunar_dbus_service_transfer_files (ThunarDBusTransferMode transfer_mode,
         }
 
       /* switch back to the previous working directory */
-      if (!exo_str_is_empty (working_directory))
+      if (!xfce_str_is_empty (working_directory))
         {
           new_working_dir = thunar_util_change_working_directory (old_working_dir);
           g_free (old_working_dir);
@@ -1425,7 +1425,7 @@ thunar_dbus_service_unlink_files (ThunarDBusFileManager  *object,
   if (screen != NULL)
     {
       /* change the working directory if necessary */
-      if (!exo_str_is_empty (working_directory))
+      if (!xfce_str_is_empty (working_directory))
         old_working_dir = thunar_util_change_working_directory (working_directory);
 
       /* try to parse the specified filenames */
@@ -1450,7 +1450,7 @@ thunar_dbus_service_unlink_files (ThunarDBusFileManager  *object,
         }
 
       /* switch back to the previous working directory */
-      if (!exo_str_is_empty (working_directory))
+      if (!xfce_str_is_empty (working_directory))
         {
           new_working_dir = thunar_util_change_working_directory (old_working_dir);
           g_free (old_working_dir);
diff --git a/thunar/thunar-device-monitor.c b/thunar/thunar-device-monitor.c
index c518142b..82ba2549 100644
--- a/thunar/thunar-device-monitor.c
+++ b/thunar/thunar-device-monitor.c
@@ -27,6 +27,8 @@
 #include <gio/gunixmounts.h>
 #endif
 
+#include <libxfce4util/libxfce4util.h>
+
 #include <thunar/thunar-device-monitor.h>
 #include <thunar/thunar-private.h>
 #include <thunar/thunar-marshal.h>
diff --git a/thunar/thunar-dnd.c b/thunar/thunar-dnd.c
index 92d7276d..964d311f 100644
--- a/thunar/thunar-dnd.c
+++ b/thunar/thunar-dnd.c
@@ -23,6 +23,8 @@
 #include <config.h>
 #endif
 
+#include <libxfce4util/libxfce4util.h>
+
 #include <thunarx/thunarx.h>
 
 #include <thunar/thunar-application.h>
diff --git a/thunar/thunar-emblem-chooser.c b/thunar/thunar-emblem-chooser.c
index d93cc92c..64c0098f 100644
--- a/thunar/thunar-emblem-chooser.c
+++ b/thunar/thunar-emblem-chooser.c
@@ -28,6 +28,8 @@
 #include <string.h>
 #endif
 
+#include <libxfce4util/libxfce4util.h>
+
 #include <thunar/thunar-emblem-chooser.h>
 #include <thunar/thunar-gobject-extensions.h>
 #include <thunar/thunar-private.h>
diff --git a/thunar/thunar-enum-types.c b/thunar/thunar-enum-types.c
index 9e0ba1f4..be21cb33 100644
--- a/thunar/thunar-enum-types.c
+++ b/thunar/thunar-enum-types.c
@@ -22,7 +22,7 @@
 #include <config.h>
 #endif
 
-#include <exo/exo.h>
+#include <libxfce4util/libxfce4util.h>
 
 #include <thunar/thunar-enum-types.h>
 
diff --git a/thunar/thunar-file-monitor.c b/thunar/thunar-file-monitor.c
index 5949601b..eeb23a6f 100644
--- a/thunar/thunar-file-monitor.c
+++ b/thunar/thunar-file-monitor.c
@@ -21,6 +21,8 @@
 #include <config.h>
 #endif
 
+#include <libxfce4util/libxfce4util.h>
+
 #include <thunar/thunar-file-monitor.h>
 #include <thunar/thunar-private.h>
 
diff --git a/thunar/thunar-file.c b/thunar/thunar-file.c
index 7ec1cbbb..6ec3cea1 100644
--- a/thunar/thunar-file.c
+++ b/thunar/thunar-file.c
@@ -52,6 +52,7 @@
 
 #include <gio/gio.h>
 #include <libxfce4ui/libxfce4ui.h>
+#include <libxfce4util/libxfce4util.h>
 
 #include <thunarx/thunarx.h>
 
@@ -1039,7 +1040,7 @@ thunar_file_info_reload (ThunarFile   *file,
                                                           G_KEY_FILE_DESKTOP_KEY_ICON,
                                                           NULL);
 
-          if (G_UNLIKELY (exo_str_is_empty (file->custom_icon_name)))
+          if (G_UNLIKELY (xfce_str_is_empty (file->custom_icon_name)))
             {
               /* make sure we set null if the string is empty else the assertion in
                * thunar_icon_factory_lookup_icon() will fail */
@@ -3809,7 +3810,7 @@ static const gchar *
 thunar_file_get_icon_name_for_state (const gchar         *icon_name,
                                      ThunarFileIconState  icon_state)
 {
-  if (exo_str_is_empty (icon_name))
+  if (xfce_str_is_empty (icon_name))
     return NULL;
 
   /* check if we have an accept icon for the icon we found */
diff --git a/thunar/thunar-folder.c b/thunar/thunar-folder.c
index 72ede938..fb128f9f 100644
--- a/thunar/thunar-folder.c
+++ b/thunar/thunar-folder.c
@@ -22,6 +22,8 @@
 #include <config.h>
 #endif
 
+#include <libxfce4util/libxfce4util.h>
+
 #include <thunar/thunar-file-monitor.h>
 #include <thunar/thunar-folder.h>
 #include <thunar/thunar-gobject-extensions.h>
diff --git a/thunar/thunar-gio-extensions.c b/thunar/thunar-gio-extensions.c
index a126f4ff..ead8be6c 100644
--- a/thunar/thunar-gio-extensions.c
+++ b/thunar/thunar-gio-extensions.c
@@ -29,7 +29,6 @@
 #include <gio/gdesktopappinfo.h>
 #endif
 
-#include <exo/exo.h>
 #include <libxfce4util/libxfce4util.h>
 
 #include <thunar/thunar-file.h>
diff --git a/thunar/thunar-history.c b/thunar/thunar-history.c
index aa81d931..45cf658d 100644
--- a/thunar/thunar-history.c
+++ b/thunar/thunar-history.c
@@ -30,6 +30,7 @@
 #include <thunar/thunar-dialogs.h>
 
 #include <libxfce4ui/libxfce4ui.h>
+#include <libxfce4util/libxfce4util.h>
 
 
 
diff --git a/thunar/thunar-icon-factory.c b/thunar/thunar-icon-factory.c
index 020a9b19..7299ebaa 100644
--- a/thunar/thunar-icon-factory.c
+++ b/thunar/thunar-icon-factory.c
@@ -869,7 +869,7 @@ thunar_icon_factory_load_icon (ThunarIconFactory        *factory,
   /* cannot happen unless there's no XSETTINGS manager
    * for the default screen, but just in case...
    */
-  if (G_UNLIKELY (exo_str_is_empty (name)))
+  if (G_UNLIKELY (xfce_str_is_empty (name)))
     {
       /* check if the caller will happly accept the fallback icon */
       if (G_LIKELY (wants_default))
diff --git a/thunar/thunar-job.c b/thunar/thunar-job.c
index efc46024..95b455ff 100644
--- a/thunar/thunar-job.c
+++ b/thunar/thunar-job.c
@@ -31,6 +31,8 @@
 
 #include <exo/exo.h>
 
+#include <libxfce4util/libxfce4util.h>
+
 #include <thunar/thunar-enum-types.h>
 #include <thunar/thunar-job.h>
 #include <thunar/thunar-marshal.h>
diff --git a/thunar/thunar-launcher.c b/thunar/thunar-launcher.c
index 1b1c9470..4fec1042 100644
--- a/thunar/thunar-launcher.c
+++ b/thunar/thunar-launcher.c
@@ -54,6 +54,7 @@
 #include <thunar/thunar-window.h>
 
 #include <libxfce4ui/libxfce4ui.h>
+#include <libxfce4util/libxfce4util.h>
 
 
 
diff --git a/thunar/thunar-list-model.c b/thunar/thunar-list-model.c
index 1c330b46..3f6b8bb6 100644
--- a/thunar/thunar-list-model.c
+++ b/thunar/thunar-list-model.c
@@ -30,6 +30,8 @@
 #include <string.h>
 #endif
 
+#include <libxfce4util/libxfce4util.h>
+
 #include <thunar/thunar-application.h>
 #include <thunar/thunar-file-monitor.h>
 #include <thunar/thunar-gobject-extensions.h>
diff --git a/thunar/thunar-location-button.c b/thunar/thunar-location-button.c
index 79790314..ce805ea7 100644
--- a/thunar/thunar-location-button.c
+++ b/thunar/thunar-location-button.c
@@ -29,6 +29,8 @@
 #include <string.h>
 #endif
 
+#include <libxfce4util/libxfce4util.h>
+
 #include <thunar/thunar-application.h>
 #include <thunar/thunar-dnd.h>
 #include <thunar/thunar-gio-extensions.h>
diff --git a/thunar/thunar-location-entry.c b/thunar/thunar-location-entry.c
index 7a81cce3..ba3f03bb 100644
--- a/thunar/thunar-location-entry.c
+++ b/thunar/thunar-location-entry.c
@@ -25,6 +25,8 @@
 
 #include <gdk/gdkkeysyms.h>
 
+#include <libxfce4util/libxfce4util.h>
+
 #include <thunar/thunar-browser.h>
 #include <thunar/thunar-dialogs.h>
 #include <thunar/thunar-gobject-extensions.h>
diff --git a/thunar/thunar-navigator.c b/thunar/thunar-navigator.c
index 9ecf22c8..0f43a2ce 100644
--- a/thunar/thunar-navigator.c
+++ b/thunar/thunar-navigator.c
@@ -21,6 +21,8 @@
 #include <config.h>
 #endif
 
+#include <libxfce4util/libxfce4util.h>
+
 #include <thunar/thunar-navigator.h>
 #include <thunar/thunar-private.h>
 
diff --git a/thunar/thunar-notify.c b/thunar/thunar-notify.c
index f6e9e9b6..a125dedd 100644
--- a/thunar/thunar-notify.c
+++ b/thunar/thunar-notify.c
@@ -26,6 +26,8 @@
 #include <libnotify/notify.h>
 #endif
 
+#include <libxfce4util/libxfce4util.h>
+
 #include <thunar/thunar-notify.h>
 #include <thunar/thunar-device.h>
 #include <thunar/thunar-private.h>
diff --git a/thunar/thunar-properties-dialog.c b/thunar/thunar-properties-dialog.c
index 285d27e8..d9bbb1f7 100644
--- a/thunar/thunar-properties-dialog.c
+++ b/thunar/thunar-properties-dialog.c
@@ -35,6 +35,7 @@
 
 #include <exo/exo.h>
 #include <libxfce4ui/libxfce4ui.h>
+#include <libxfce4util/libxfce4util.h>
 
 #include <thunar/thunar-abstract-dialog.h>
 #include <thunar/thunar-application.h>
diff --git a/thunar/thunar-renamer-pair.c b/thunar/thunar-renamer-pair.c
index 23e77f44..d02ecdd7 100644
--- a/thunar/thunar-renamer-pair.c
+++ b/thunar/thunar-renamer-pair.c
@@ -21,6 +21,8 @@
 #include <config.h>
 #endif
 
+#include <libxfce4util/libxfce4util.h>
+
 #include <thunar/thunar-private.h>
 #include <thunar/thunar-renamer-pair.h>
 
diff --git a/thunar/thunar-shortcuts-view.c b/thunar/thunar-shortcuts-view.c
index 66953307..db3f125e 100644
--- a/thunar/thunar-shortcuts-view.c
+++ b/thunar/thunar-shortcuts-view.c
@@ -33,6 +33,8 @@
 
 #include <gdk/gdkkeysyms.h>
 
+#include <libxfce4util/libxfce4util.h>
+
 #include <thunar/thunar-navigator.h>
 #include <thunar/thunar-application.h>
 #include <thunar/thunar-browser.h>
diff --git a/thunar/thunar-side-pane.c b/thunar/thunar-side-pane.c
index ea896e31..7cb8ee95 100644
--- a/thunar/thunar-side-pane.c
+++ b/thunar/thunar-side-pane.c
@@ -21,6 +21,8 @@
 #include <config.h>
 #endif
 
+#include <libxfce4util/libxfce4util.h>
+
 #include <thunar/thunar-private.h>
 #include <thunar/thunar-side-pane.h>
 
diff --git a/thunar/thunar-standard-view.c b/thunar/thunar-standard-view.c
index 705af7f2..84f57e78 100644
--- a/thunar/thunar-standard-view.c
+++ b/thunar/thunar-standard-view.c
@@ -32,6 +32,8 @@
 
 #include <gdk/gdkkeysyms.h>
 
+#include <libxfce4util/libxfce4util.h>
+
 #include <thunar/thunar-application.h>
 #include <thunar/thunar-menu.h>
 #include <thunar/thunar-dialogs.h>
diff --git a/thunar/thunar-thumbnailer.c b/thunar/thunar-thumbnailer.c
index 5ef2c34d..591c61d6 100644
--- a/thunar/thunar-thumbnailer.c
+++ b/thunar/thunar-thumbnailer.c
@@ -23,6 +23,8 @@
 #include <config.h>
 #endif
 
+#include <libxfce4util/libxfce4util.h>
+
 #include <thunar/thunar-thumbnailer-proxy.h>
 #include <thunar/thunar-marshal.h>
 #include <thunar/thunar-preferences.h>
diff --git a/thunar/thunar-tree-view.c b/thunar/thunar-tree-view.c
index 69de731c..d468ffc7 100644
--- a/thunar/thunar-tree-view.c
+++ b/thunar/thunar-tree-view.c
@@ -28,6 +28,8 @@
 
 #include <gdk/gdkkeysyms.h>
 
+#include <libxfce4util/libxfce4util.h>
+
 #include <thunar/thunar-application.h>
 #include <thunar/thunar-clipboard-manager.h>
 #include <thunar/thunar-device.h>
diff --git a/thunar/thunar-user.c b/thunar/thunar-user.c
index 3fef90cf..3c29909d 100644
--- a/thunar/thunar-user.c
+++ b/thunar/thunar-user.c
@@ -50,6 +50,8 @@
 
 #include <exo/exo.h>
 
+#include <libxfce4util/libxfce4util.h>
+
 #include <thunar/thunar-user.h>
 #include <thunar/thunar-util.h>
 
@@ -290,7 +292,7 @@ thunar_user_load (ThunarUser *user)
           name[0] = g_ascii_toupper (name[0]);
 
           /* replace all occurances of '&' */
-          t = exo_str_replace (user->real_name, "&", name);
+          t = xfce_str_replace (user->real_name, "&", name);
           g_free (user->real_name);
           user->real_name = t;
 
diff --git a/thunar/thunar-view.c b/thunar/thunar-view.c
index 96593242..2f3f8bd0 100644
--- a/thunar/thunar-view.c
+++ b/thunar/thunar-view.c
@@ -21,6 +21,8 @@
 #include <config.h>
 #endif
 
+#include <libxfce4util/libxfce4util.h>
+
 #include <thunar/thunar-private.h>
 #include <thunar/thunar-view.h>
 
diff --git a/thunar/thunar-window.c b/thunar/thunar-window.c
index 67deb5c0..9a7ce952 100644
--- a/thunar/thunar-window.c
+++ b/thunar/thunar-window.c
@@ -32,6 +32,8 @@
 
 #include <gdk/gdkkeysyms.h>
 
+#include <libxfce4util/libxfce4util.h>
+
 #include <thunar/thunar-application.h>
 #include <thunar/thunar-browser.h>
 #include <thunar/thunar-clipboard-manager.h>
diff --git a/thunarx/Makefile.am b/thunarx/Makefile.am
index 234c8b10..8482b25d 100644
--- a/thunarx/Makefile.am
+++ b/thunarx/Makefile.am
@@ -53,6 +53,7 @@ libthunarx_3_la_CFLAGS =						\
 	$(GIO_CFLAGS)							\
 	$(GTK_CFLAGS)							\
 	$(GMODULE_CFLAGS)						\
+	$(LIBXFCE4UTIL_CFLAGS)						\
 	$(PLATFORM_CFLAGS)
 
 libthunarx_3_la_LDFLAGS =						\
@@ -66,7 +67,8 @@ libthunarx_3_la_LIBADD =						\
 	$(GLIB_LIBS)							\
 	$(GMODULE_LIBS)							\
 	$(GIO_LIBS)							\
-	$(GTK_LIBS)
+	$(GTK_LIBS)							\
+	$(LIBXFCE4UTIL_LIBS)
 
 pkgconfigdir = $(libdir)/pkgconfig
 pkgconfig_DATA = thunarx-3.pc
diff --git a/thunarx/thunarx-file-info.c b/thunarx/thunarx-file-info.c
index 4305f2ea..4061bdce 100644
--- a/thunarx/thunarx-file-info.c
+++ b/thunarx/thunarx-file-info.c
@@ -23,6 +23,8 @@
 #include <config.h>
 #endif
 
+#include <libxfce4util/libxfce4util.h>
+
 #include <thunarx/thunarx-file-info.h>
 #include <thunarx/thunarx-private.h>
 
diff --git a/thunarx/thunarx-menu-provider.c b/thunarx/thunarx-menu-provider.c
index 6fa48a1f..1123ab74 100644
--- a/thunarx/thunarx-menu-provider.c
+++ b/thunarx/thunarx-menu-provider.c
@@ -22,6 +22,8 @@
 #include <config.h>
 #endif
 
+#include <libxfce4util/libxfce4util.h>
+
 #include <thunarx/thunarx-menu-provider.h>
 #include <thunarx/thunarx-private.h>
 
diff --git a/thunarx/thunarx-preferences-provider.c b/thunarx/thunarx-preferences-provider.c
index 13f5075d..1a00454b 100644
--- a/thunarx/thunarx-preferences-provider.c
+++ b/thunarx/thunarx-preferences-provider.c
@@ -22,6 +22,8 @@
 #include <config.h>
 #endif
 
+#include <libxfce4util/libxfce4util.h>
+
 #include <thunarx/thunarx-preferences-provider.h>
 #include <thunarx/thunarx-private.h>
 
diff --git a/thunarx/thunarx-private.h b/thunarx/thunarx-private.h
index d0658cc5..a8ab3b06 100644
--- a/thunarx/thunarx-private.h
+++ b/thunarx/thunarx-private.h
@@ -39,8 +39,6 @@ G_GNUC_END_IGNORE_DEPRECATIONS
 
 G_BEGIN_DECLS
 
-#define I_(string) (g_intern_static_string ((string)))
-
 G_GNUC_INTERNAL
 void   thunarx_object_list_take_reference (GList      *object_list,
                                            gpointer    target);
diff --git a/thunarx/thunarx-property-page-provider.c b/thunarx/thunarx-property-page-provider.c
index 8b623ee7..7dd86f80 100644
--- a/thunarx/thunarx-property-page-provider.c
+++ b/thunarx/thunarx-property-page-provider.c
@@ -22,6 +22,8 @@
 #include <config.h>
 #endif
 
+#include <libxfce4util/libxfce4util.h>
+
 #include <thunarx/thunarx-private.h>
 #include <thunarx/thunarx-property-page-provider.h>
 
diff --git a/thunarx/thunarx-provider-plugin.c b/thunarx/thunarx-provider-plugin.c
index d39b2d67..1e0c3d71 100644
--- a/thunarx/thunarx-provider-plugin.c
+++ b/thunarx/thunarx-provider-plugin.c
@@ -24,6 +24,8 @@
 
 #include <glib/gi18n-lib.h>
 
+#include <libxfce4util/libxfce4util.h>
+
 #include <thunarx/thunarx-private.h>
 #include <thunarx/thunarx-provider-plugin.h>
 
diff --git a/thunarx/thunarx-renamer-provider.c b/thunarx/thunarx-renamer-provider.c
index aed1771c..ddd265ca 100644
--- a/thunarx/thunarx-renamer-provider.c
+++ b/thunarx/thunarx-renamer-provider.c
@@ -22,6 +22,8 @@
 #include <config.h>
 #endif
 
+#include <libxfce4util/libxfce4util.h>
+
 #include <thunarx/thunarx-renamer-provider.h>
 #include <thunarx/thunarx-private.h>
 
diff --git a/thunarx/thunarx-renamer.c b/thunarx/thunarx-renamer.c
index 34094e0e..53ae5d69 100644
--- a/thunarx/thunarx-renamer.c
+++ b/thunarx/thunarx-renamer.c
@@ -24,6 +24,8 @@
 
 #include <glib/gi18n-lib.h>
 
+#include <libxfce4util/libxfce4util.h>
+
 #include <thunarx/thunarx-renamer.h>
 #include <thunarx/thunarx-private.h>
 
