diff --git a/thunar/thunar-toolbar-editor.c b/thunar/thunar-toolbar-editor.c
index 7112a099..d1bb139f 100644
--- a/thunar/thunar-toolbar-editor.c
+++ b/thunar/thunar-toolbar-editor.c
@@ -140,6 +140,7 @@ thunar_toolbar_editor_init (ThunarToolbarEditor *toolbar_editor)
   grid = gtk_grid_new ();
   gtk_grid_set_column_spacing (GTK_GRID (grid), 6);
   gtk_grid_set_row_spacing (GTK_GRID (grid), 6);
+  gtk_grid_set_row_homogeneous (GTK_GRID (grid), TRUE);
   gtk_container_set_border_width (GTK_CONTAINER (grid), 12);
   gtk_container_add (GTK_CONTAINER (frame), grid);
   gtk_widget_show (grid);
@@ -227,14 +228,6 @@ thunar_toolbar_editor_init (ThunarToolbarEditor *toolbar_editor)
   /* next row */
   row++;
 
-  /* create the horiz separator */
-  separator = gtk_separator_new (GTK_ORIENTATION_HORIZONTAL);
-  gtk_grid_attach (GTK_GRID (grid), separator, 1, row, 1, 1);
-  gtk_widget_show (separator);
-
-  /* next row */
-  row++;
-
   /* create the "Use Default" button */
   button = gtk_button_new_with_mnemonic (_("De_fault Order"));
   g_signal_connect_swapped (G_OBJECT (button), "clicked", G_CALLBACK (thunar_toolbar_editor_use_defaults), toolbar_editor);
