diff --git a/thunar/thunar-toolbar-editor.c b/thunar/thunar-toolbar-editor.c
index 1fc76d74..d660695f 100644
--- a/thunar/thunar-toolbar-editor.c
+++ b/thunar/thunar-toolbar-editor.c
@@ -103,7 +103,6 @@ static void
 thunar_toolbar_editor_init (ThunarToolbarEditor *toolbar_editor)
 {
   GtkCellRenderer        *renderer;
-  GtkWidget              *separator;
   GtkWidget              *button;
   GtkWidget              *frame;
   GtkWidget              *image;
