diff --git a/configure.ac.in b/configure.ac.in
index 90c9bb4e..09efbfe6 100644
--- a/configure.ac.in
+++ b/configure.ac.in
@@ -106,6 +106,16 @@ AC_ARG_WITH([helper-path-prefix],
             [HELPER_PATH_PREFIX="$libdir"])
 AC_SUBST([HELPER_PATH_PREFIX])
 
+dnl *******************************************************
+dnl *** Optional support for custom thunarx directories ***
+dnl *******************************************************
+AC_ARG_WITH([custom-thunarx-dirs-enabled],
+            [AS_HELP_STRING([--with-custom-thunarx-dirs-enabled],
+                            [Enables the posibillity to set custom thunarx directories by using the environment variable THUNARX_DIRS])],
+            [CUSTOM_THUNARX_DIRS_ENABLED="TRUE"],
+            [CUSTOM_THUNARX_DIRS_ENABLED="FALSE"])
+AC_SUBST([CUSTOM_THUNARX_DIRS_ENABLED])
+
 dnl ***********************************************
 dnl *** Work-around system-specific limitations ***
 dnl ***********************************************
diff --git a/thunarx/Makefile.am b/thunarx/Makefile.am
index 8482b25d..8af6a80a 100644
--- a/thunarx/Makefile.am
+++ b/thunarx/Makefile.am
@@ -3,6 +3,7 @@ AM_CPPFLAGS =								\
 	-DG_LOG_DOMAIN=\"thunarx\"					\
 	-DTHUNARX_COMPILATION						\
 	-DTHUNARX_DIRECTORY=\"$(libdir)/thunarx-$(THUNARX_VERSION_API)\"	\
+	-DTHUNARX_ENABLE_CUSTOM_DIRS=\"$(CUSTOM_THUNARX_DIRS_ENABLED)\"	\
 	$(PLATFORM_CPPFLAGS)
 
 libthunarx_headers =							\
diff --git a/thunarx/thunarx-provider-factory.c b/thunarx/thunarx-provider-factory.c
index c110b56d..5d378bbc 100644
--- a/thunarx/thunarx-provider-factory.c
+++ b/thunarx/thunarx-provider-factory.c
@@ -157,12 +157,15 @@ thunarx_provider_factory_create_modules (ThunarxProviderFactory *factory)
   const gchar           *name;
   GList                 *lp;
   GDir                  *dp;
-  gchar                 *dirs_string;
+  gchar                 *dirs_string = NULL;
   gchar                **dirs;
 
-  dirs_string = (gchar *) g_getenv ("THUNARX_DIRS");
-  if (!dirs_string)
+  if (g_strcmp0 (THUNARX_ENABLE_CUSTOM_DIRS, "TRUE") == 0)
+    dirs_string = (gchar *) g_getenv ("THUNARX_DIRS");
+
+  if (dirs_string == NULL)
     dirs_string = THUNARX_DIRECTORY;
+
   dirs = g_strsplit (dirs_string, G_SEARCHPATH_SEPARATOR_S, 0);
 
   for (int i = 0; dirs[i] != NULL; i++)
diff --git a/thunarx/thunarx-provider-module.c b/thunarx/thunarx-provider-module.c
index 27066233..9ffba66f 100644
--- a/thunarx/thunarx-provider-module.c
+++ b/thunarx/thunarx-provider-module.c
@@ -175,13 +175,16 @@ thunarx_provider_module_load (GTypeModule *type_module)
 {
   ThunarxProviderModule *module = THUNARX_PROVIDER_MODULE (type_module);
   gchar                 *path;
-  gchar                 *dirs_string;
+  gchar                 *dirs_string = NULL;
   gchar                **dirs;
   gboolean               found;
 
-  dirs_string = (gchar *) g_getenv ("THUNARX_DIRS");
-  if (!dirs_string)
+  if (g_strcmp0 (THUNARX_ENABLE_CUSTOM_DIRS, "TRUE") == 0)
+    dirs_string = (gchar *) g_getenv ("THUNARX_DIRS");
+
+  if (dirs_string == NULL)
     dirs_string = THUNARX_DIRECTORY;
+
   dirs = g_strsplit (dirs_string, G_SEARCHPATH_SEPARATOR_S, 0);
 
   found = FALSE;
