diff --git a/thunar/thunar-file.c b/thunar/thunar-file.c
index 9a483afb..9e2c850d 100644
--- a/thunar/thunar-file.c
+++ b/thunar/thunar-file.c
@@ -1634,7 +1634,8 @@ thunar_file_execute (ThunarFile  *file,
       type = g_key_file_get_string (key_file, G_KEY_FILE_DESKTOP_GROUP, G_KEY_FILE_DESKTOP_KEY_TYPE, NULL);
       if (G_LIKELY (g_strcmp0 (type, "Application") == 0))
         {
-          exec = g_key_file_get_string (key_file, G_KEY_FILE_DESKTOP_GROUP, G_KEY_FILE_DESKTOP_KEY_EXEC, NULL);
+          /* load exec property but ignore escape sequences */
+          exec = g_key_file_get_value (key_file, G_KEY_FILE_DESKTOP_GROUP, G_KEY_FILE_DESKTOP_KEY_EXEC, NULL);
           if (G_LIKELY (exec != NULL))
             {
               /* if the .desktop file is not secure, ask user what to do */
