diff --git a/thunar/thunar-application.c b/thunar/thunar-application.c
index 3c2012be..1faffc8a 100644
--- a/thunar/thunar-application.c
+++ b/thunar/thunar-application.c
@@ -652,8 +652,8 @@ thunar_application_load_css (void)
   css_provider = gtk_css_provider_new ();
 
   gtk_css_provider_load_from_data (css_provider,
-    /* for the pathbar-buttons any margin looks ugly*/
-    ".path-bar-button { margin-right: 0; }"
+    /* for the location-buttons any margin looks ugly */
+    ".location-button { margin-right: 0; }"
     /* remove extra border between side pane and view */
     ".shortcuts-pane { border-right-width: 0px; }"
     /* add missing top border to side pane */
diff --git a/thunar/thunar-location-button.c b/thunar/thunar-location-button.c
index ce805ea7..b83c951e 100644
--- a/thunar/thunar-location-button.c
+++ b/thunar/thunar-location-button.c
@@ -267,6 +267,8 @@ thunar_location_button_init (ThunarLocationButton *button)
   /* but don't show it, as it is only a fake to retrieve the bold size */
 
   /* add widget to css class which matches all buttons in the path-bar */
+  /* keep as well the old name 'path-bar-button'. It might be used by themes */
+  gtk_style_context_add_class (gtk_widget_get_style_context (GTK_WIDGET (button)), "location-button");
   gtk_style_context_add_class (gtk_widget_get_style_context (GTK_WIDGET (button)), "path-bar-button");
 }
 
diff --git a/thunar/thunar-location-buttons.c b/thunar/thunar-location-buttons.c
index 59206ec6..e0070c30 100644
--- a/thunar/thunar-location-buttons.c
+++ b/thunar/thunar-location-buttons.c
@@ -279,6 +279,13 @@ thunar_location_buttons_init (ThunarLocationButtons *buttons)
                                "path-bar");
 
   /* add sub-buttons to css class which matches all buttons in the path-bar */
+  /* keep as well the old name 'path-bar-button'. It might be used by themes */
+  gtk_style_context_add_class (gtk_widget_get_style_context (GTK_WIDGET (buttons->left_slider)),
+                               "location-button");
+  gtk_style_context_add_class (gtk_widget_get_style_context (GTK_WIDGET (buttons->right_slider)),
+                               "location-button");
+  gtk_style_context_add_class (gtk_widget_get_style_context (GTK_WIDGET (buttons->filler_widget)),
+                               "location-button");
   gtk_style_context_add_class (gtk_widget_get_style_context (GTK_WIDGET (buttons->left_slider)),
                                "path-bar-button");
   gtk_style_context_add_class (gtk_widget_get_style_context (GTK_WIDGET (buttons->right_slider)),
@@ -1225,7 +1232,7 @@ thunar_location_buttons_context_menu (ThunarLocationButton  *button,
     }
   else
     {
-      /* context menu on some other pathbar element */
+      /* context menu on some other location button */
       thunar_menu_add_sections (context_menu, THUNAR_MENU_SECTION_OPEN
                                             | THUNAR_MENU_SECTION_SENDTO
                                             | THUNAR_MENU_SECTION_COPY_PASTE
diff --git a/thunar/thunar-window.c b/thunar/thunar-window.c
index e067ae6e..18682e1c 100644
--- a/thunar/thunar-window.c
+++ b/thunar/thunar-window.c
@@ -205,8 +205,8 @@ static gboolean  thunar_window_action_reload              (ThunarWindow
 static gboolean  thunar_window_action_toggle_split_view   (ThunarWindow           *window);
 static gboolean  thunar_window_action_switch_next_tab     (ThunarWindow           *window);
 static gboolean  thunar_window_action_switch_previous_tab (ThunarWindow           *window);
-static gboolean  thunar_window_action_pathbar_changed     (ThunarWindow           *window);
-static gboolean  thunar_window_action_toolbar_changed     (ThunarWindow           *window);
+static gboolean  thunar_window_action_locationbar_entry   (ThunarWindow           *window);
+static gboolean  thunar_window_action_locationbar_buttons (ThunarWindow           *window);
 static gboolean  thunar_window_action_shortcuts_changed   (ThunarWindow           *window);
 static gboolean  thunar_window_action_tree_changed        (ThunarWindow           *window);
 static gboolean  thunar_window_action_statusbar_changed   (ThunarWindow           *window);
@@ -462,8 +462,8 @@ static XfceGtkActionEntry thunar_window_action_entries[] =
     { THUNAR_WINDOW_ACTION_RELOAD_ALT,                     "<Actions>/ThunarWindow/reload-alt",                      "F5",                   XFCE_GTK_IMAGE_MENU_ITEM, NULL,                          NULL,                                                                                NULL,                      G_CALLBACK (thunar_window_action_reload),             },
     { THUNAR_WINDOW_ACTION_VIEW_SPLIT,                     "<Actions>/ThunarWindow/toggle-split-view",               "F3",                   XFCE_GTK_CHECK_MENU_ITEM, N_ ("Spl_it View"),            N_ ("Open/Close Split View"),                                                        NULL,                      G_CALLBACK (thunar_window_action_toggle_split_view),  },
     { THUNAR_WINDOW_ACTION_VIEW_LOCATION_SELECTOR_MENU,    "<Actions>/ThunarWindow/view-location-selector-menu",     "",                     XFCE_GTK_MENU_ITEM,       N_ ("_Location Selector"),     NULL,                                                                                NULL,                      NULL,                                                 },
-    { THUNAR_WINDOW_ACTION_VIEW_LOCATION_SELECTOR_PATHBAR, "<Actions>/ThunarWindow/view-location-selector-pathbar",  "",                     XFCE_GTK_CHECK_MENU_ITEM, N_ ("_Pathbar Style"),         N_ ("Modern approach with buttons that correspond to folders"),                      NULL,                      G_CALLBACK (thunar_window_action_pathbar_changed),    },
-    { THUNAR_WINDOW_ACTION_VIEW_LOCATION_SELECTOR_TOOLBAR, "<Actions>/ThunarWindow/view-location-selector-toolbar",  "",                     XFCE_GTK_CHECK_MENU_ITEM, N_ ("_Toolbar Style"),         N_ ("Traditional approach with location bar and navigation buttons"),                NULL,                      G_CALLBACK (thunar_window_action_toolbar_changed),    },
+    { THUNAR_WINDOW_ACTION_VIEW_LOCATION_SELECTOR_ENTRY,   "<Actions>/ThunarWindow/view-location-selector-entry",    "",                     XFCE_GTK_CHECK_MENU_ITEM, N_ ("_Entry Style"),           N_ ("Traditional entry showing the current path"),                                   NULL,                      G_CALLBACK (thunar_window_action_locationbar_entry),  },
+    { THUNAR_WINDOW_ACTION_VIEW_LOCATION_SELECTOR_BUTTONS, "<Actions>/ThunarWindow/view-location-selector-buttons",  "",                     XFCE_GTK_CHECK_MENU_ITEM, N_ ("_Buttons Style"),         N_ ("Modern approach with buttons that correspond to folders"),                      NULL,                      G_CALLBACK (thunar_window_action_locationbar_buttons),},
     { THUNAR_WINDOW_ACTION_VIEW_SIDE_PANE_MENU,            "<Actions>/ThunarWindow/view-side-pane-menu",             "",                     XFCE_GTK_MENU_ITEM,       N_ ("_Side Pane"),             NULL,                                                                                NULL,                      NULL,                                                 },
     { THUNAR_WINDOW_ACTION_VIEW_SIDE_PANE_SHORTCUTS,       "<Actions>/ThunarWindow/view-side-pane-shortcuts",        "<Primary>b",           XFCE_GTK_CHECK_MENU_ITEM, N_ ("_Shortcuts"),             N_ ("Toggles the visibility of the shortcuts pane"),                                 NULL,                      G_CALLBACK (thunar_window_action_shortcuts_changed),  },
     { THUNAR_WINDOW_ACTION_VIEW_SIDE_PANE_TREE,            "<Actions>/ThunarWindow/view-side-pane-tree",             "<Primary>e",           XFCE_GTK_CHECK_MENU_ITEM, N_ ("_Tree"),                  N_ ("Toggles the visibility of the tree pane"),                                      NULL,                      G_CALLBACK (thunar_window_action_tree_changed),       },
@@ -1250,10 +1250,10 @@ thunar_window_update_view_menu (ThunarWindow *window,
   sub_items =  gtk_menu_new();
   gtk_menu_set_accel_group (GTK_MENU (sub_items), window->accel_group);
   g_object_get (window->preferences, "last-location-bar", &last_location_bar, NULL);
-  xfce_gtk_toggle_menu_item_new_from_action_entry (get_action_entry (THUNAR_WINDOW_ACTION_VIEW_LOCATION_SELECTOR_PATHBAR), G_OBJECT (window),
-                                                   (g_strcmp0 (last_location_bar, g_type_name (THUNAR_TYPE_LOCATION_ENTRY)) == 0), GTK_MENU_SHELL (sub_items));
-  xfce_gtk_toggle_menu_item_new_from_action_entry (get_action_entry (THUNAR_WINDOW_ACTION_VIEW_LOCATION_SELECTOR_TOOLBAR), G_OBJECT (window),
+  xfce_gtk_toggle_menu_item_new_from_action_entry (get_action_entry (THUNAR_WINDOW_ACTION_VIEW_LOCATION_SELECTOR_BUTTONS), G_OBJECT (window),
                                                    (g_strcmp0 (last_location_bar, g_type_name (THUNAR_TYPE_LOCATION_BUTTONS)) == 0), GTK_MENU_SHELL (sub_items));
+  xfce_gtk_toggle_menu_item_new_from_action_entry (get_action_entry (THUNAR_WINDOW_ACTION_VIEW_LOCATION_SELECTOR_ENTRY), G_OBJECT (window),
+                                                   (g_strcmp0 (last_location_bar, g_type_name (THUNAR_TYPE_LOCATION_ENTRY)) == 0), GTK_MENU_SHELL (sub_items));
   g_free (last_location_bar);
   gtk_menu_item_set_submenu (GTK_MENU_ITEM (item), GTK_WIDGET (sub_items));
   item = xfce_gtk_menu_item_new_from_action_entry (get_action_entry (THUNAR_WINDOW_ACTION_VIEW_SIDE_PANE_MENU), G_OBJECT (window), GTK_MENU_SHELL (menu));
@@ -3464,18 +3464,18 @@ thunar_window_action_toggle_split_view (ThunarWindow *window)
 
 
 static gboolean
-thunar_window_action_pathbar_changed (ThunarWindow *window)
+thunar_window_action_locationbar_entry (ThunarWindow *window)
 {
   gchar    *last_location_bar;
-  gboolean  pathbar_checked;
+  gboolean  entry_checked;
 
   _thunar_return_val_if_fail (THUNAR_IS_WINDOW (window), FALSE);
 
   g_object_get (window->preferences, "last-location-bar", &last_location_bar, NULL);
-  pathbar_checked = (g_strcmp0 (last_location_bar, g_type_name (THUNAR_TYPE_LOCATION_ENTRY)) == 0);
+  entry_checked = (g_strcmp0 (last_location_bar, g_type_name (THUNAR_TYPE_LOCATION_ENTRY)) == 0);
   g_free (last_location_bar);
 
-  if (pathbar_checked)
+  if (entry_checked)
     g_object_set (window->preferences, "last-location-bar", g_type_name (G_TYPE_NONE), NULL);
   else
     g_object_set (window->preferences, "last-location-bar", g_type_name (THUNAR_TYPE_LOCATION_ENTRY), NULL);
@@ -3487,18 +3487,18 @@ thunar_window_action_pathbar_changed (ThunarWindow *window)
 
 
 static gboolean
-thunar_window_action_toolbar_changed (ThunarWindow *window)
+thunar_window_action_locationbar_buttons (ThunarWindow *window)
 {
   gchar    *last_location_bar;
-  gboolean  toolbar_checked;
+  gboolean  buttons_checked;
 
   _thunar_return_val_if_fail (THUNAR_IS_WINDOW (window), FALSE);
 
   g_object_get (window->preferences, "last-location-bar", &last_location_bar, NULL);
-  toolbar_checked = (g_strcmp0 (last_location_bar, g_type_name (THUNAR_TYPE_LOCATION_BUTTONS)) == 0);
+  buttons_checked = (g_strcmp0 (last_location_bar, g_type_name (THUNAR_TYPE_LOCATION_BUTTONS)) == 0);
   g_free (last_location_bar);
 
-  if (toolbar_checked)
+  if (buttons_checked)
     g_object_set (window->preferences, "last-location-bar", g_type_name (G_TYPE_NONE), NULL);
   else
     g_object_set (window->preferences, "last-location-bar", g_type_name (THUNAR_TYPE_LOCATION_BUTTONS), NULL);
diff --git a/thunar/thunar-window.h b/thunar/thunar-window.h
index 0a728727..20789cc8 100644
--- a/thunar/thunar-window.h
+++ b/thunar/thunar-window.h
@@ -55,8 +55,8 @@ typedef enum
   THUNAR_WINDOW_ACTION_RELOAD_ALT,
   THUNAR_WINDOW_ACTION_VIEW_SPLIT,
   THUNAR_WINDOW_ACTION_VIEW_LOCATION_SELECTOR_MENU,
-  THUNAR_WINDOW_ACTION_VIEW_LOCATION_SELECTOR_PATHBAR,
-  THUNAR_WINDOW_ACTION_VIEW_LOCATION_SELECTOR_TOOLBAR,
+  THUNAR_WINDOW_ACTION_VIEW_LOCATION_SELECTOR_ENTRY,
+  THUNAR_WINDOW_ACTION_VIEW_LOCATION_SELECTOR_BUTTONS,
   THUNAR_WINDOW_ACTION_VIEW_SIDE_PANE_MENU,
   THUNAR_WINDOW_ACTION_VIEW_SIDE_PANE_SHORTCUTS,
   THUNAR_WINDOW_ACTION_VIEW_SIDE_PANE_TREE,
