diff --git a/thunar/thunar-preferences-dialog.c b/thunar/thunar-preferences-dialog.c
index 967e5288..ae63505c 100644
--- a/thunar/thunar-preferences-dialog.c
+++ b/thunar/thunar-preferences-dialog.c
@@ -280,7 +280,6 @@ thunar_preferences_dialog_init (ThunarPreferencesDialog *dialog)
   GtkWidget      *vbox;
   GtkWidget      *infobar;
   GEnumClass     *type;
-  gchar          *path;
   gchar          *date;
   gint            row = 0;
 
@@ -1267,9 +1266,6 @@ thunar_preferences_dialog_init (ThunarPreferencesDialog *dialog)
   gtk_container_add (GTK_CONTAINER (frame), grid);
   gtk_widget_show (grid);
 
-  /* check if "thunar-volman" is found */
-  path = g_find_program_in_path ("thunar-volman");
-
   /* add check button to enable/disable auto mounting */
   button = gtk_check_button_new_with_mnemonic (_("Enable _Volume Management"));
   g_object_bind_property (G_OBJECT (dialog->preferences),
@@ -1346,9 +1342,6 @@ thunar_preferences_dialog_init (ThunarPreferencesDialog *dialog)
                                     );
   gtk_container_add (GTK_CONTAINER (frame), grid);
   gtk_widget_show (grid);
-
-  /* cleanup */
-  g_free (path);
 }
 
 
