diff --git a/thunar/thunar-details-view.c b/thunar/thunar-details-view.c
index d2cd93b3..6c0ab807 100644
--- a/thunar/thunar-details-view.c
+++ b/thunar/thunar-details-view.c
@@ -791,8 +791,8 @@ thunar_details_view_button_press_event (GtkTreeView       *tree_view,
                   gtk_tree_selection_select_path (selection, path);
                 }
 
-              /* show the context menu */
-              thunar_standard_view_context_menu (THUNAR_STANDARD_VIEW (details_view));
+              /* queue the menu popup */
+              thunar_standard_view_queue_popup (THUNAR_STANDARD_VIEW (details_view), event);
             }
           gtk_tree_path_free (path);
         }
