diff --git a/thunar/thunar-standard-view.c b/thunar/thunar-standard-view.c
index 4c259e89..d4b6155e 100644
--- a/thunar/thunar-standard-view.c
+++ b/thunar/thunar-standard-view.c
@@ -4374,6 +4374,7 @@ thunar_standard_view_connect_accelerators (ThunarStandardView *standard_view)
   if (standard_view->accel_group == NULL)
     return;
 
+  /* can cause emissions of "accel-map::changed" which in turns calls thunar_window_reconnect_accelerators */
   xfce_gtk_accel_map_add_entries (thunar_standard_view_action_entries, G_N_ELEMENTS (thunar_standard_view_action_entries));
   xfce_gtk_accel_group_connect_action_entries (standard_view->accel_group,
                                                thunar_standard_view_action_entries,
diff --git a/thunar/thunar-window.c b/thunar/thunar-window.c
index 8b3c4187..95da11e9 100644
--- a/thunar/thunar-window.c
+++ b/thunar/thunar-window.c
@@ -1992,7 +1992,6 @@ thunar_window_notebook_switch_page (GtkWidget    *notebook,
   /* Use accelerators only on the current active tab */
   if (window->view != NULL)
     g_object_set (G_OBJECT (window->view), "accel-group", NULL, NULL);
-  g_object_set (G_OBJECT (page), "accel-group", window->accel_group, NULL);
 
   if (G_LIKELY (window->view != NULL))
     {
@@ -2047,6 +2046,10 @@ thunar_window_notebook_switch_page (GtkWidget    *notebook,
   window->view = page;
   window->view_type = G_TYPE_FROM_INSTANCE (page);
 
+  /* connect accelerators for the view, we need to do this after window->view has been set,
+   * see thunar_window_reconnect_accelerators and thunar_standard_view_connect_accelerators */
+  g_object_set (G_OBJECT (page), "accel-group", window->accel_group, NULL);
+
   g_signal_connect_swapped (G_OBJECT (window->view), "notify::selected-files",
                             G_CALLBACK (thunar_window_trash_selection_updated), window);
 
