diff --git a/configure.ac.in b/configure.ac.in
index 05ae312c..342152c9 100644
--- a/configure.ac.in
+++ b/configure.ac.in
@@ -183,7 +183,7 @@ AC_DEFINE(GLIB_VERSION_MAX_ALLOWED, GLIB_VERSION_2_66, [Prevent post 2.66 APIs])
 dnl ******************************
 dnl *** GObject Instrospection ***
 dnl ******************************
-GOBJECT_INTROSPECTION_CHECK([1.30.0])
+GOBJECT_INTROSPECTION_CHECK([1.66.0])
 
 dnl ********************************************
 dnl *** Check for session management support ***
