diff --git a/thunar/thunar-application.c b/thunar/thunar-application.c
index 0a54204e..a972e8e4 100644
--- a/thunar/thunar-application.c
+++ b/thunar/thunar-application.c
@@ -1409,6 +1409,8 @@ thunar_application_open_window (ThunarApplication *application,
   GtkWidget *window;
   gchar     *role;
   gboolean   open_new_window_as_tab;
+  gboolean   misc_open_new_windows_in_split_view;
+  gboolean   restore_tabs;
 
   _thunar_return_val_if_fail (THUNAR_IS_APPLICATION (application), NULL);
   _thunar_return_val_if_fail (directory == NULL || THUNAR_IS_FILE (directory), NULL);
@@ -1469,6 +1471,13 @@ thunar_application_open_window (ThunarApplication *application,
   if (directory != NULL)
     thunar_window_set_current_directory (THUNAR_WINDOW (window), directory);
 
+  /* enable split view, if preferred */
+  g_object_get (G_OBJECT (application->preferences),
+                "misc-open-new-windows-in-split-view", &misc_open_new_windows_in_split_view,
+                "last-restore-tabs", &restore_tabs, NULL);
+  if (misc_open_new_windows_in_split_view && !restore_tabs)
+    thunar_window_notebook_toggle_split_view (THUNAR_WINDOW (window));
+
   return window;
 }
 
diff --git a/thunar/thunar-preferences.c b/thunar/thunar-preferences.c
index 226c6063..2c9e9665 100644
--- a/thunar/thunar-preferences.c
+++ b/thunar/thunar-preferences.c
@@ -120,6 +120,7 @@ enum
   PROP_TREE_ICON_SIZE,
   PROP_MISC_SWITCH_TO_NEW_TAB,
   PROP_MISC_VERTICAL_SPLIT_PANE,
+  PROP_MISC_OPEN_NEW_WINDOWS_IN_SPLIT_VIEW,
   PROP_MISC_COMPACT_VIEW_MAX_CHARS,
   PROP_MISC_HIGHLIGHTING_ENABLED,
   N_PROPERTIES,
@@ -1093,6 +1094,18 @@ thunar_preferences_class_init (ThunarPreferencesClass *klass)
                           FALSE,
                           EXO_PARAM_READWRITE);
 
+  /**
+   * ThunarPreferences:misc-open-new-windows-in-split-view
+   *
+   * If true, all thunar windows will have split view enabled.
+   **/
+  preferences_props[PROP_MISC_OPEN_NEW_WINDOWS_IN_SPLIT_VIEW] =
+    g_param_spec_boolean ("misc-open-new-windows-in-split-view",
+                          "MiscOpenNewWindowsInSplitView",
+                          NULL,
+                          FALSE,
+                          EXO_PARAM_READWRITE);
+
   /**
    * ThunarPreferences:compact-view-max-chars
    *
diff --git a/thunar/thunar-window.c b/thunar/thunar-window.c
index f03b3d84..04533c60 100644
--- a/thunar/thunar-window.c
+++ b/thunar/thunar-window.c
@@ -3242,6 +3242,7 @@ thunar_window_action_open_new_window (ThunarWindow *window,
   ThunarHistory     *history;
   ThunarWindow      *new_window;
   ThunarFile        *start_file;
+  gboolean           misc_open_new_windows_in_split_view;
 
   /* popup a new window */
   application = thunar_application_get ();
@@ -3249,6 +3250,10 @@ thunar_window_action_open_new_window (ThunarWindow *window,
                                                               gtk_widget_get_screen (GTK_WIDGET (window)), NULL, TRUE));
   g_object_unref (G_OBJECT (application));
 
+  g_object_get (G_OBJECT (window->preferences), "misc-open-new-windows-in-split-view", &misc_open_new_windows_in_split_view, NULL);
+  if (misc_open_new_windows_in_split_view)
+    thunar_window_notebook_toggle_split_view (new_window);
+
   /* if we have no origin view we are done */
   if (window->view == NULL)
     return TRUE;
