diff --git a/thunar/thunar-dbus-service-infos.xml b/thunar/thunar-dbus-service-infos.xml
index 572447f5..1e08f7c6 100644
--- a/thunar/thunar-dbus-service-infos.xml
+++ b/thunar/thunar-dbus-service-infos.xml
@@ -46,12 +46,39 @@
       the file after choosing an application.
     -->
     <method name="DisplayChooserDialog">
+      <annotation name="org.freedesktop.DBus.Deprecated" value="Deprecated. Use DisplayApplicationChooserDialog instead"/>
       <arg direction="in" name="uri" type="s" />
       <arg direction="in" name="open" type="b" />
       <arg direction="in" name="display" type="s" />
       <arg direction="in" name="startup_id" type="s" />
     </method>
 
+    <!--
+      DisplayApplicationChooserDialog (uri : STRING, open : BOOLEAN, display : STRING, startup_id : STRING) : VOID
+
+      uri                        : either a file:-URI or an absolute path.
+      open                       : TRUE to let the user open the file identified by the
+                                   uri after choosing an application, FALSE to only update
+                                   the mime database.
+      preselect_default_checkbox : TRUE in order to preselect the 'use as default application' checkbox
+      display                    : the screen on which to display the chooser dialog or
+                                   "" to use the default screen of the file manager.
+      startup_id                 : the DESKTOP_STARTUP_ID environment variable for properly
+                                   handling startup notification and focus stealing.
+
+      Displays the chooser dialog, which lets the user select an application
+      to handle the file identified by the uri (and all other files of the
+      same mime type). If open is TRUE the user will also be able to open
+      the file after choosing an application.
+    -->
+    <method name="DisplayApplicationChooserDialog">
+      <arg direction="in" name="uri" type="s" />
+      <arg direction="in" name="open" type="b" />
+      <arg direction="in" name="preselect_default_checkbox" type="b" />
+      <arg direction="in" name="display" type="s" />
+      <arg direction="in" name="startup_id" type="s" />
+    </method>
+
 
     <!--
       DisplayFolder (uri : STRING, display : STRING, startup_id : STRING) : VOID
diff --git a/thunar/thunar-dbus-service.c b/thunar/thunar-dbus-service.c
index c26f0f81..73462f90 100644
--- a/thunar/thunar-dbus-service.c
+++ b/thunar/thunar-dbus-service.c
@@ -86,6 +86,14 @@ static gboolean thunar_dbus_service_display_chooser_dialog      (ThunarDBusFileM
                                                                  const gchar            *display,
                                                                  const gchar            *startup_id,
                                                                  ThunarDBusService      *dbus_service);
+static gboolean thunar_dbus_service_display_app_chooser_dialog  (ThunarDBusFileManager  *object,
+                                                                 GDBusMethodInvocation  *invocation,
+                                                                 const gchar            *uri,
+                                                                 gboolean                open,
+                                                                 gboolean                preselect_default_checkbox,
+                                                                 const gchar            *display,
+                                                                 const gchar            *startup_id,
+                                                                 ThunarDBusService      *dbus_service);
 static gboolean thunar_dbus_service_display_folder              (ThunarDBusFileManager  *object,
                                                                  GDBusMethodInvocation  *invocation,
                                                                  const gchar            *uri,
@@ -305,6 +313,7 @@ thunar_dbus_service_init (ThunarDBusService *dbus_service)
 
   connect_signals_multiple (dbus_service->file_manager, dbus_service,
                             "handle-display-chooser-dialog", thunar_dbus_service_display_chooser_dialog,
+                            "handle-display-application-chooser-dialog", thunar_dbus_service_display_app_chooser_dialog,
                             "handle-display-folder", thunar_dbus_service_display_folder,
                             "handle-display-folder-and-select", thunar_dbus_service_display_folder_and_select,
                             "handle-display-file-properties", thunar_dbus_service_display_file_properties,
@@ -432,6 +441,9 @@ thunar_dbus_service_trash_bin_changed (ThunarDBusService *dbus_service,
 
 
 
+/**
+ * TODO: Method to be removed after thunar 4.18 got released (dbus call is marked as deprecated). Use thunar_dbus_service_display_app_chooser_dialog instead.
+ */
 static gboolean
 thunar_dbus_service_display_chooser_dialog (ThunarDBusFileManager  *object,
                                             GDBusMethodInvocation  *invocation,
@@ -468,6 +480,43 @@ out:
 
 
 
+static gboolean
+thunar_dbus_service_display_app_chooser_dialog (ThunarDBusFileManager  *object,
+                                                GDBusMethodInvocation  *invocation,
+                                                const gchar            *uri,
+                                                gboolean                open,
+                                                gboolean                preselect_default_checkbox,
+                                                const gchar            *display,
+                                                const gchar            *startup_id,
+                                                ThunarDBusService      *dbus_service)
+{
+  ThunarFile *file;
+  GdkScreen  *screen;
+  GError     *error = NULL;
+
+  /* parse uri and display parameters */
+  if (!thunar_dbus_service_parse_uri_and_display (dbus_service, uri, display, &file, &screen, &error))
+    goto out;
+
+  /* popup the chooser dialog */
+  /* TODO use the startup id! */
+  thunar_show_chooser_dialog (screen, file, open, preselect_default_checkbox);
+
+  /* cleanup */
+  g_object_unref (G_OBJECT (screen));
+  g_object_unref (G_OBJECT (file));
+
+out:
+  if (error)
+    g_dbus_method_invocation_take_error (invocation, error);
+  else
+    thunar_dbus_file_manager_complete_display_chooser_dialog (object, invocation);
+
+  return TRUE;
+}
+
+
+
 static gboolean
 thunar_dbus_service_display_folder (ThunarDBusFileManager  *object,
                                     GDBusMethodInvocation  *invocation,
