diff --git a/thunar/thunar-action-manager.c b/thunar/thunar-action-manager.c
index 5ee1362b..2a04a1e3 100644
--- a/thunar/thunar-action-manager.c
+++ b/thunar/thunar-action-manager.c
@@ -41,7 +41,6 @@
 #include <thunar/thunar-gtk-extensions.h>
 #include <thunar/thunar-icon-factory.h>
 #include <thunar/thunar-io-scan-directory.h>
-#include <thunar/thunar-job-operation.h>
 #include <thunar/thunar-preferences.h>
 #include <thunar/thunar-private.h>
 #include <thunar/thunar-properties-dialog.h>
@@ -205,8 +204,6 @@ static gboolean                thunar_action_manager_action_create_document
 static GtkWidget              *thunar_action_manager_create_document_submenu_new(ThunarActionManager            *action_mgr);
 static void                    thunar_action_manager_new_files_created          (ThunarActionManager            *action_mgr,
                                                                                  GList                          *new_thunar_files);
-static gboolean                thunar_action_manager_action_undo                (ThunarActionManager            *action_mgr);
-static gboolean                thunar_action_manager_action_redo                (ThunarActionManager            *action_mgr);
 
 
 
@@ -309,8 +306,6 @@ static XfceGtkActionEntry thunar_action_manager_action_entries[] =
     { THUNAR_ACTION_MANAGER_ACTION_MOUNT,            NULL,                                                   "",                  XFCE_GTK_MENU_ITEM,       N_ ("_Mount"),                          N_ ("Mount the selected device"),                                                                NULL,                   G_CALLBACK (thunar_action_manager_action_open),                },
     { THUNAR_ACTION_MANAGER_ACTION_UNMOUNT,          NULL,                                                   "",                  XFCE_GTK_MENU_ITEM,       N_ ("_Unmount"),                        N_ ("Unmount the selected device"),                                                              NULL,                   G_CALLBACK (thunar_action_manager_action_unmount),             },
     { THUNAR_ACTION_MANAGER_ACTION_EJECT,            NULL,                                                   "",                  XFCE_GTK_MENU_ITEM,       N_ ("_Eject"),                          N_ ("Eject the selected device"),                                                                NULL,                   G_CALLBACK (thunar_action_manager_action_eject),               },
-    { THUNAR_ACTION_MANAGER_ACTION_UNDO,             "<Actions>/ThunarActionManager/undo",                   "<Primary>Z",        XFCE_GTK_IMAGE_MENU_ITEM, N_ ("_Undo"),                           N_ ("Undo the latest operation"),                                                               "edit-undo-symbolic",    G_CALLBACK (thunar_action_manager_action_undo),                },
-    { THUNAR_ACTION_MANAGER_ACTION_REDO,             "<Actions>/ThunarActionManager/redo",                   "<Primary><shift>Z", XFCE_GTK_IMAGE_MENU_ITEM, N_ ("_Redo"),                           N_ ("Redo the latest operation"),                                                               "edit-redo-symbolic",    G_CALLBACK (thunar_action_manager_action_redo),                },
     { THUNAR_ACTION_MANAGER_ACTION_EDIT_LAUNCHER,    NULL,                                                   "",                  XFCE_GTK_IMAGE_MENU_ITEM, N_ ("_Edit Launcher"),                  N_ ("Edit the selected action_mgr"),                                                             "gtk-edit",             G_CALLBACK (thunar_action_manager_action_edit_launcher),       },
 };
 
@@ -1855,16 +1850,6 @@ thunar_action_manager_append_menu_item (ThunarActionManager       *action_mgr,
         gtk_menu_item_set_label (GTK_MENU_ITEM (item), eject_label);
         return item;
 
-      case THUNAR_ACTION_MANAGER_ACTION_UNDO:
-        item = xfce_gtk_menu_item_new_from_action_entry (action_entry, G_OBJECT (action_mgr), GTK_MENU_SHELL (menu));
-        gtk_widget_set_sensitive (item, thunar_job_operation_can_undo ());
-        return item;
-
-      case THUNAR_ACTION_MANAGER_ACTION_REDO:
-        item = xfce_gtk_menu_item_new_from_action_entry (action_entry, G_OBJECT (action_mgr), GTK_MENU_SHELL (menu));
-        gtk_widget_set_sensitive (item, thunar_job_operation_can_redo ());
-        return item;
-
       default:
         return xfce_gtk_menu_item_new_from_action_entry (action_entry, G_OBJECT (action_mgr), GTK_MENU_SHELL (menu));
     }
@@ -3396,30 +3381,6 @@ thunar_action_manager_new_files_created (ThunarActionManager *action_mgr,
 
 
 
-static gboolean
-thunar_action_manager_action_undo (ThunarActionManager *action_mgr)
-{
-  _thunar_return_val_if_fail (THUNAR_IS_ACTION_MANAGER (action_mgr), FALSE);
-
-  thunar_job_operation_undo ();
-
-  return TRUE;
-}
-
-
-
-static gboolean
-thunar_action_manager_action_redo (ThunarActionManager *action_mgr)
-{
-  _thunar_return_val_if_fail (THUNAR_IS_ACTION_MANAGER (action_mgr), FALSE);
-
-  thunar_job_operation_redo ();
-
-  return TRUE;
-}
-
-
-
 XfceGtkActionEntry*
 thunar_action_manager_get_action_entries (void)
 {
diff --git a/thunar/thunar-action-manager.h b/thunar/thunar-action-manager.h
index 1c606df6..4416ee00 100644
--- a/thunar/thunar-action-manager.h
+++ b/thunar/thunar-action-manager.h
@@ -72,8 +72,6 @@ typedef enum
   THUNAR_ACTION_MANAGER_ACTION_MOUNT,
   THUNAR_ACTION_MANAGER_ACTION_UNMOUNT,
   THUNAR_ACTION_MANAGER_ACTION_EJECT,
-  THUNAR_ACTION_MANAGER_ACTION_UNDO,
-  THUNAR_ACTION_MANAGER_ACTION_REDO,
   THUNAR_ACTION_MANAGER_ACTION_EDIT_LAUNCHER,
 
   THUNAR_ACTION_MANAGER_N_ACTIONS
diff --git a/thunar/thunar-menu.c b/thunar/thunar-menu.c
index 494a9f14..9272b92b 100644
--- a/thunar/thunar-menu.c
+++ b/thunar/thunar-menu.c
@@ -252,15 +252,6 @@ thunar_menu_add_sections (ThunarMenu         *menu,
 
   _thunar_return_val_if_fail (THUNAR_IS_MENU (menu), FALSE);
 
-  if (menu_sections & THUNAR_MENU_SECTION_UNDO)
-    {
-      item_added = (thunar_action_manager_append_menu_item (menu->action_mgr, GTK_MENU_SHELL (menu), THUNAR_ACTION_MANAGER_ACTION_UNDO, force) != NULL);
-      item_added = (thunar_action_manager_append_menu_item (menu->action_mgr, GTK_MENU_SHELL (menu), THUNAR_ACTION_MANAGER_ACTION_REDO, force) != NULL);
-
-      if (item_added)
-        xfce_gtk_menu_append_separator (GTK_MENU_SHELL (menu));
-    }
-
   if (menu_sections & THUNAR_MENU_SECTION_CREATE_NEW_FILES)
     {
       item_added = FALSE;
diff --git a/thunar/thunar-menu.h b/thunar/thunar-menu.h
index e5b3c6e7..eab9a638 100644
--- a/thunar/thunar-menu.h
+++ b/thunar/thunar-menu.h
@@ -65,7 +65,6 @@ typedef enum
   THUNAR_MENU_SECTION_PROPERTIES          = 1 << 13,
   THUNAR_MENU_SECTION_MOUNTABLE           = 1 << 14,
   THUNAR_MENU_SECTION_REMOVE_FROM_RECENT  = 1 << 15,
-  THUNAR_MENU_SECTION_UNDO                = 1 << 16,
   THUNAR_MENU_SECTION_EDIT_LAUNCHER       = 1 << 17,
 
 } ThunarMenuSections;
diff --git a/thunar/thunar-window.c b/thunar/thunar-window.c
index a57b792e..050b7def 100644
--- a/thunar/thunar-window.c
+++ b/thunar/thunar-window.c
@@ -47,6 +47,7 @@
 #include <thunar/thunar-gtk-extensions.h>
 #include <thunar/thunar-history.h>
 #include <thunar/thunar-icon-view.h>
+#include <thunar/thunar-job-operation.h>
 #include <thunar/thunar-location-buttons.h>
 #include <thunar/thunar-location-entry.h>
 #include <thunar/thunar-marshal.h>
@@ -199,6 +200,10 @@ static gboolean  thunar_window_action_close_tab           (ThunarWindow
                                                            GtkWidget              *menu_item);
 static gboolean  thunar_window_action_close_window        (ThunarWindow           *window,
                                                            GtkWidget              *menu_item);
+static gboolean  thunar_window_action_undo                (ThunarWindow           *window,
+                                                           GtkWidget              *menu_item);
+static gboolean  thunar_window_action_redo                (ThunarWindow           *window,
+                                                           GtkWidget              *menu_item);
 static gboolean  thunar_window_action_preferences         (ThunarWindow           *window,
                                                            GtkWidget              *menu_item);
 static gboolean  thunar_window_action_reload              (ThunarWindow           *window,
@@ -485,6 +490,8 @@ static XfceGtkActionEntry thunar_window_action_entries[] =
     { THUNAR_WINDOW_ACTION_CLOSE_ALL_WINDOWS,              "<Actions>/ThunarWindow/close-all-windows",               "<Primary><Shift>w",    XFCE_GTK_IMAGE_MENU_ITEM, N_ ("Close _All Windows"),     N_ ("Close all Thunar windows"),                                                     NULL,                      G_CALLBACK (thunar_window_action_close_all_windows),  },
 
     { THUNAR_WINDOW_ACTION_EDIT_MENU,                      "<Actions>/ThunarWindow/edit-menu",                       "",                     XFCE_GTK_MENU_ITEM,       N_ ("_Edit"),                  NULL,                                                                                NULL,                      NULL,                                                 },
+    { THUNAR_WINDOW_ACTION_UNDO,                           "<Actions>/ThunarActionManager/undo",                     "<Primary>Z",           XFCE_GTK_IMAGE_MENU_ITEM, N_ ("_Undo"),                  N_ ("Undo the latest operation"),                                                    "edit-undo-symbolic",      G_CALLBACK (thunar_window_action_undo),               },
+    { THUNAR_WINDOW_ACTION_REDO,                           "<Actions>/ThunarActionManager/redo",                     "<Primary><shift>Z",    XFCE_GTK_IMAGE_MENU_ITEM, N_ ("_Redo"),                  N_ ("Redo the latest operation"),                                                    "edit-redo-symbolic",      G_CALLBACK (thunar_window_action_redo),               },
     { THUNAR_WINDOW_ACTION_PREFERENCES,                    "<Actions>/ThunarWindow/preferences",                     "",                     XFCE_GTK_IMAGE_MENU_ITEM, N_ ("Pr_eferences..."),        N_ ("Edit Thunars Preferences"),                                                     "preferences-system",      G_CALLBACK (thunar_window_action_preferences),        },
 
     { THUNAR_WINDOW_ACTION_VIEW_MENU,                      "<Actions>/ThunarWindow/view-menu",                       "",                     XFCE_GTK_MENU_ITEM,       N_ ("_View"),                  NULL,                                                                                NULL,                      NULL,                                                 },
@@ -1281,10 +1288,16 @@ thunar_window_update_edit_menu (ThunarWindow *window,
   _thunar_return_if_fail (THUNAR_IS_WINDOW (window));
 
   thunar_gtk_menu_clean (GTK_MENU (menu));
+
+  gtk_menu_item = xfce_gtk_menu_item_new_from_action_entry (get_action_entry (THUNAR_WINDOW_ACTION_UNDO), G_OBJECT (window), GTK_MENU_SHELL (menu));
+  gtk_widget_set_sensitive (gtk_menu_item, thunar_job_operation_can_undo ());
+  gtk_menu_item = xfce_gtk_menu_item_new_from_action_entry (get_action_entry (THUNAR_WINDOW_ACTION_REDO), G_OBJECT (window), GTK_MENU_SHELL (menu));
+  gtk_widget_set_sensitive (gtk_menu_item, thunar_job_operation_can_redo ());
+  xfce_gtk_menu_append_separator (GTK_MENU_SHELL (menu));
+
   thunar_menu_add_sections (THUNAR_MENU (menu), THUNAR_MENU_SECTION_CUT
                                               | THUNAR_MENU_SECTION_COPY_PASTE
-                                              | THUNAR_MENU_SECTION_TRASH_DELETE
-                                              | THUNAR_MENU_SECTION_UNDO);
+                                              | THUNAR_MENU_SECTION_TRASH_DELETE);
   if (window->view != NULL)
     {
       thunar_standard_view_append_menu_item (THUNAR_STANDARD_VIEW (window->view),
@@ -1322,6 +1335,7 @@ thunar_window_update_edit_menu (ThunarWindow *window,
           g_list_free (thunarx_menu_items);
         }
     }
+
   xfce_gtk_menu_item_new_from_action_entry (get_action_entry (THUNAR_WINDOW_ACTION_PREFERENCES), G_OBJECT (window), GTK_MENU_SHELL (menu));
   gtk_widget_show_all (GTK_WIDGET (menu));
 
@@ -3490,6 +3504,26 @@ thunar_window_action_close_window (ThunarWindow *window,
 
 
 
+static gboolean
+thunar_window_action_undo (ThunarWindow *window,
+                           GtkWidget    *menu_item)
+{
+  thunar_job_operation_undo ();
+  return TRUE; /* return value required in case of shortcut activation, in order to signal that the accel key got handled */
+}
+
+
+
+static gboolean
+thunar_window_action_redo (ThunarWindow *window,
+                           GtkWidget    *menu_item)
+{
+  thunar_job_operation_redo ();
+  return TRUE; /* return value required in case of shortcut activation, in order to signal that the accel key got handled */
+}
+
+
+
 static gboolean
 thunar_window_action_preferences (ThunarWindow *window,
                                   GtkWidget    *menu_item)
diff --git a/thunar/thunar-window.h b/thunar/thunar-window.h
index a4be46bd..c14c0839 100644
--- a/thunar/thunar-window.h
+++ b/thunar/thunar-window.h
@@ -49,6 +49,8 @@ typedef enum
   THUNAR_WINDOW_ACTION_CLOSE_WINDOW,
   THUNAR_WINDOW_ACTION_CLOSE_ALL_WINDOWS,
   THUNAR_WINDOW_ACTION_EDIT_MENU,
+  THUNAR_WINDOW_ACTION_UNDO,
+  THUNAR_WINDOW_ACTION_REDO,
   THUNAR_WINDOW_ACTION_PREFERENCES,
   THUNAR_WINDOW_ACTION_VIEW_MENU,
   THUNAR_WINDOW_ACTION_RELOAD,
