diff --git a/NEWS b/NEWS
index 9129b790..6d1c106d 100644
--- a/NEWS
+++ b/NEWS
@@ -1,3 +1,46 @@
+4.17.10 (2022-10-16)
+=======
+- Introduce separate class for 'thunar-job-operation-history'
+- Add Toolbar items for undo/redo (Issue #819)
+- Add notification on undo/redo (Issue #903)
+- Shorten wait time to show file transfer rate (Issue #888)
+- Add Image preview sidepane (Issue #357)
+- Prevent GLib-GIO-CRITICAL messages if 'file_>info' is not set
+- Add 'redo' feature and multi-level undo/redo (Issue #819)
+- Add 'undo' support for the latest link operation (Issue #819)
+- Add 'undo' support for the latest file-create operation (Issue 819)
+- Add 'undo' support for the latest trash operation (Issue #819)
+- Add 'undo' support for the latest rename operation (Issue #819)
+- Add 'undo' support for the latest move operation (Issue #819)
+- Add 'undo' support for the latest copy operation (Issue #819)
+- Update Dbus support for feature "Set Default Application" (issue #342)
+- hidden preference "misc-open-new-windows-in-split-view" (Issue #889)
+- Add title to message dialogs
+- Fix details view's name column's prelight when highlight is enabled
+- Add split-view button to Toolbar (Issue #831)
+- View accels not working after initial startup (Issue #803)
+- Remove "open with" prefix for apps in open with menu (Issue #863)
+- Prevent xfconf warning during startup for property "sort-order"
+- Thunar hangs with misc-directory-specific-settings (Issue #861)
+- Add DnD support for application/octet-stream (#278)
+- Add support for PrefersNonDefaultGPU in .desktop files (Issue #692)
+- Enable right click + drag for list view (Issue #851)
+- Inhibit some actions when a search is active (Issue #847)
+- Add posibillity to set custom color to specific files (Issue: #160)
+- Add support for alternative copy/paste/cut shortcuts (Issue #845)
+- Properly remember sort settings in all cases (Issue #836)
+- Create separate section for thumbnail settings in preferences dialog
+- Translation Updates:
+  Albanian, Arabic, Armenian, Armenian (Armenia), Asturian, Basque,
+  Bengali, Catalan, Chinese (China), Chinese (Hong Kong), Chinese
+  (Taiwan), Czech, Danish, Dutch, Eastern Armenian, English
+  (Australia), English (United Kingdom), Esperanto, Estonian, Finnish,
+  French, Galician, German, Greek, Hebrew, Hungarian, Icelandic,
+  Indonesian, Interlingue, Japanese, Kazakh, Korean, Latvian, Malay,
+  Norwegian Bokmål, Norwegian Nynorsk, Occitan (post 1500), Persian
+  (Iran), Polish, Portuguese, Romanian, Serbian, Slovak, Spanish,
+  Swedish, Telugu, Thai, Urdu, Urdu (Pakistan), Uyghur, Vietnamese
+
 4.17.9
 ======
 - Dont rely on xfconf value for menubar toggle (Issue #777)
diff --git a/configure.ac.in b/configure.ac.in
index 342152c9..4274e68e 100644
--- a/configure.ac.in
+++ b/configure.ac.in
@@ -26,10 +26,10 @@ m4_define([thunarx_verinfo], [0:0:0])
 m4_define([thunarx_version_api], [3])
 m4_define([thunar_version_major], [4])
 m4_define([thunar_version_minor], [17])
-m4_define([thunar_version_micro], [9])
+m4_define([thunar_version_micro], [10])
 m4_define([thunar_version_nano], [])
 m4_define([thunar_version_build], [@REVISION@])
-m4_define([thunar_version_tag], [git])
+m4_define([thunar_version_tag], [])
 m4_define([thunar_version], [thunar_version_major().thunar_version_minor().thunar_version_micro()ifelse(thunar_version_nano(), [], [], [.thunar_version_nano()])ifelse(thunar_version_tag(), [git], [thunar_version_tag()-thunar_version_build()], [thunar_version_tag()])])
 
 dnl *******************************************
