diff --git a/configure.ac.in b/configure.ac.in
index 2c0225fd..3e3c5b66 100644
--- a/configure.ac.in
+++ b/configure.ac.in
@@ -40,7 +40,7 @@ m4_define([thunar_debug_default], [ifelse(thunar_version_tag(), [git], [yes], [m
 dnl ***************************
 dnl *** Initialize autoconf ***
 dnl ***************************
-AC_COPYRIGHT([Copyright (c) 2004-2020 The Thunar development team. All rights reserved.])
+AC_COPYRIGHT([Copyright (c) 2004-2022 The Thunar development team. All rights reserved.])
 AC_INIT([thunar], [thunar_version], [https://gitlab.xfce.org/xfce/thunar], [thunar])
 AC_PREREQ([2.60])
 AC_CONFIG_MACRO_DIRS([m4])
diff --git a/thunar/thunar-application.c b/thunar/thunar-application.c
index 4961a241..cfecbd61 100644
--- a/thunar/thunar-application.c
+++ b/thunar/thunar-application.c
@@ -469,7 +469,7 @@ thunar_application_handle_local_options (GApplication *gapp,
   if (G_UNLIKELY (opt_version))
     {
       g_print ("%s %s (Xfce %s)\n\n", PACKAGE_NAME, PACKAGE_VERSION, xfce_version_string ());
-      g_print ("%s\n", "Copyright (c) 2004-2020");
+      g_print ("%s\n", "Copyright (c) 2004-2022");
       g_print ("\t%s\n\n", _("The Thunar development team. All rights reserved."));
       g_print ("%s\n\n", _("Written by Benedikt Meurer <benny@xfce.org>."));
       g_print (_("Please report bugs to <%s>."), PACKAGE_BUGREPORT);
diff --git a/thunar/thunar-dialogs.c b/thunar/thunar-dialogs.c
index 0bf1f5ca..46d940a0 100644
--- a/thunar/thunar-dialogs.c
+++ b/thunar/thunar-dialogs.c
@@ -413,8 +413,8 @@ thunar_dialogs_show_about (GtkWindow   *parent,
                          "copyright", "Copyright \302\251 2004-2011 Benedikt Meurer\n"
                                       "Copyright \302\251 2009-2011 Jannis Pohlmann\n"
                                       "Copyright \302\251 2009-2012 Nick Schermer\n"
-                                      "Copyright \302\251 2017-2020 Alexander Schwinn\n"
-                                      "Copyright \302\251 2017-2020 Andre Miranda",
+                                      "Copyright \302\251 2017-2022 Alexander Schwinn\n"
+                                      "Copyright \302\251 2017-2022 Andre Miranda",
                          "destroy-with-parent", TRUE,
                          "documenters", documenters,
                          "license", XFCE_LICENSE_GPL,
