diff --git a/thunar/thunar-action-manager.c b/thunar/thunar-action-manager.c
index 5c4b4cd7..4af610f1 100644
--- a/thunar/thunar-action-manager.c
+++ b/thunar/thunar-action-manager.c
@@ -674,7 +674,7 @@ thunar_action_manager_set_selected_files (ThunarComponent *component,
         }
       else
         {
-          if (action_mgr->files_are_all_executable && !thunar_file_is_executable (lp->data))
+          if (action_mgr->files_are_all_executable && !thunar_file_can_execute (lp->data))
             action_mgr->files_are_all_executable = FALSE;
           ++action_mgr->n_regulars_to_process;
         }
@@ -1181,7 +1181,7 @@ thunar_action_manager_poke_files_finish (ThunarBrowser *browser,
               files = g_list_prepend (files, lp->data);
 
               /* check if the file is executable */
-              executable = (executable && thunar_file_is_executable (lp->data));
+              executable = (executable && thunar_file_can_execute (lp->data));
             }
         }
 
diff --git a/thunar/thunar-dnd.c b/thunar/thunar-dnd.c
index b7581d60..fe0373a2 100644
--- a/thunar/thunar-dnd.c
+++ b/thunar/thunar-dnd.c
@@ -242,7 +242,7 @@ thunar_dnd_perform (GtkWidget    *widget,
           succeed = FALSE;
         }
     }
-  else if (thunar_file_is_executable (file))
+  else if (thunar_file_can_execute (file))
     {
       /* TODO any chance to determine the working dir here? */
       succeed = thunar_file_execute (file, NULL, widget, file_list, NULL, &error);
diff --git a/thunar/thunar-file.c b/thunar/thunar-file.c
index 58c4c59b..eb148972 100644
--- a/thunar/thunar-file.c
+++ b/thunar/thunar-file.c
@@ -1035,7 +1035,7 @@ thunar_file_info_reload (ThunarFile   *file,
     }
 
   /* check if this file is a desktop entry and we have the permission to execute it */
-  if (thunar_file_is_desktop_file (file) && thunar_file_is_executable (file))
+  if (thunar_file_is_desktop_file (file) && thunar_file_can_execute (file))
     {
       /* determine the custom icon and display name for .desktop files */
 
@@ -1629,7 +1629,7 @@ thunar_file_execute (ThunarFile  *file,
 
   if (thunar_file_is_desktop_file (file))
     {
-      is_secure = thunar_file_is_executable (file);
+      is_secure = thunar_file_can_execute (file);
 
       key_file = thunar_g_file_query_key_file (file->gfile, NULL, &err);
       if (key_file == NULL)
@@ -1859,7 +1859,7 @@ thunar_file_launch (ThunarFile  *file,
     }
 
   /* check if we should execute the file */
-  if (thunar_file_is_executable (file))
+  if (thunar_file_can_execute (file))
     return thunar_file_execute (file, NULL, parent, NULL, NULL, error);
 
   /* determine the default application to open the file */
@@ -2096,7 +2096,7 @@ thunar_file_accepts_drop (ThunarFile     *file,
             }
         }
     }
-  else if (thunar_file_is_executable (file))
+  else if (thunar_file_can_execute (file))
     {
       /* determine the possible actions */
       actions &= (GDK_ACTION_COPY | GDK_ACTION_MOVE | GDK_ACTION_LINK | GDK_ACTION_PRIVATE);
@@ -2978,7 +2978,7 @@ thunar_file_is_ancestor (const ThunarFile *file,
 
 
 /**
- * thunar_file_is_executable:
+ * thunar_file_can_execute:
  * @file : a #ThunarFile instance.
  *
  * Determines whether the owner of the current process is allowed
@@ -2989,7 +2989,7 @@ thunar_file_is_ancestor (const ThunarFile *file,
  * Return value: %TRUE if @file can be executed.
  **/
 gboolean
-thunar_file_is_executable (const ThunarFile *file)
+thunar_file_can_execute (const ThunarFile *file)
 {
   ThunarPreferences   *preferences;
   gboolean             exec_shell_scripts = FALSE;
diff --git a/thunar/thunar-file.h b/thunar/thunar-file.h
index 8955e30c..da862eec 100644
--- a/thunar/thunar-file.h
+++ b/thunar/thunar-file.h
@@ -209,7 +209,7 @@ gboolean          thunar_file_is_gfile_ancestor          (const ThunarFile
                                                           GFile                  *ancestor);
 gboolean          thunar_file_is_ancestor                (const ThunarFile       *file,
                                                           const ThunarFile       *ancestor);
-gboolean          thunar_file_is_executable              (const ThunarFile       *file);
+gboolean          thunar_file_can_execute                (const ThunarFile       *file);
 gboolean          thunar_file_is_writable                (const ThunarFile       *file);
 gboolean          thunar_file_is_hidden                  (const ThunarFile       *file);
 gboolean          thunar_file_is_home                    (const ThunarFile       *file);
diff --git a/thunar/thunar-properties-dialog.c b/thunar/thunar-properties-dialog.c
index 0b3469bb..7cccdaa1 100644
--- a/thunar/thunar-properties-dialog.c
+++ b/thunar/thunar-properties-dialog.c
@@ -1294,7 +1294,7 @@ thunar_properties_dialog_update_single (ThunarPropertiesDialog *dialog)
     }
 
   /* update the application chooser (shown only for non-executable regular files!) */
-  show_chooser = thunar_file_is_regular (file) && !thunar_file_is_executable (file);
+  show_chooser = thunar_file_is_regular (file) && !thunar_file_can_execute (file);
   gtk_widget_set_visible (dialog->openwith_chooser, show_chooser);
   if (show_chooser)
     thunar_chooser_button_set_file (THUNAR_CHOOSER_BUTTON (dialog->openwith_chooser), file);
diff --git a/thunar/thunar-standard-view.c b/thunar/thunar-standard-view.c
index c411a72f..1c892d6e 100644
--- a/thunar/thunar-standard-view.c
+++ b/thunar/thunar-standard-view.c
@@ -2198,7 +2198,7 @@ thunar_standard_view_get_drop_file (ThunarStandardView *standard_view,
       file = thunar_list_model_get_file (standard_view->model, &iter);
 
       /* we can only drop to directories and executable files */
-      if (!thunar_file_is_directory (file) && !thunar_file_is_executable (file))
+      if (!thunar_file_is_directory (file) && !thunar_file_can_execute (file))
         {
           /* drop to the folder instead */
           g_object_unref (G_OBJECT (file));
