diff --git a/thunar/thunar-column-editor.c b/thunar/thunar-column-editor.c
index e920f507..21fda5b6 100644
--- a/thunar/thunar-column-editor.c
+++ b/thunar/thunar-column-editor.c
@@ -361,12 +361,9 @@ thunar_column_editor_init (ThunarColumnEditor *column_editor)
   combo = gtk_combo_box_text_new ();
   enum_class = g_type_class_ref (THUNAR_TYPE_FOLDER_ITEM_COUNT);
 
-  gtk_combo_box_text_append_text (GTK_COMBO_BOX_TEXT (combo),
-                                  g_enum_get_value (enum_class, THUNAR_FOLDER_ITEM_COUNT_NEVER)->value_nick);
-  gtk_combo_box_text_append_text (GTK_COMBO_BOX_TEXT (combo),
-                                  g_enum_get_value (enum_class, THUNAR_FOLDER_ITEM_COUNT_ONLY_LOCAL)->value_nick);
-  gtk_combo_box_text_append_text (GTK_COMBO_BOX_TEXT (combo),
-                                  g_enum_get_value (enum_class, THUNAR_FOLDER_ITEM_COUNT_ALWAYS)->value_nick);
+  gtk_combo_box_text_append_text (GTK_COMBO_BOX_TEXT (combo), _(g_enum_get_value (enum_class, THUNAR_FOLDER_ITEM_COUNT_NEVER)->value_nick));
+  gtk_combo_box_text_append_text (GTK_COMBO_BOX_TEXT (combo), _(g_enum_get_value (enum_class, THUNAR_FOLDER_ITEM_COUNT_ONLY_LOCAL)->value_nick));
+  gtk_combo_box_text_append_text (GTK_COMBO_BOX_TEXT (combo), _(g_enum_get_value (enum_class, THUNAR_FOLDER_ITEM_COUNT_ALWAYS)->value_nick));
   g_type_class_unref (enum_class);
 
   g_object_bind_property_full (G_OBJECT (column_editor->preferences), "misc-folder-item-count",
diff --git a/thunar/thunar-job-operation.c b/thunar/thunar-job-operation.c
index e7ee0977..f5db65d4 100644
--- a/thunar/thunar-job-operation.c
+++ b/thunar/thunar-job-operation.c
@@ -299,7 +299,7 @@ thunar_job_operation_get_kind_nick (ThunarJobOperation *job_operation)
   enum_class = g_type_class_ref (THUNAR_TYPE_JOB_OPERATION_KIND);
   enum_value = g_enum_get_value (enum_class, job_operation->operation_kind);
 
-  return enum_value->value_nick;
+  return _(enum_value->value_nick);
 }
 
 
diff --git a/thunar/thunar-preferences-dialog.c b/thunar/thunar-preferences-dialog.c
index faaf16ef..e029b804 100644
--- a/thunar/thunar-preferences-dialog.c
+++ b/thunar/thunar-preferences-dialog.c
@@ -759,10 +759,8 @@ thunar_preferences_dialog_init (ThunarPreferencesDialog *dialog)
 
   combo = gtk_combo_box_text_new ();
   type = g_type_class_ref (THUNAR_TYPE_IMAGE_PREVIEW_MODE);
-  gtk_combo_box_text_append_text (GTK_COMBO_BOX_TEXT (combo),
-                                  g_enum_get_value (type, THUNAR_IMAGE_PREVIEW_MODE_STANDALONE)->value_nick);
-  gtk_combo_box_text_append_text (GTK_COMBO_BOX_TEXT (combo),
-                                  g_enum_get_value (type, THUNAR_IMAGE_PREVIEW_MODE_EMBEDDED)->value_nick);
+  gtk_combo_box_text_append_text (GTK_COMBO_BOX_TEXT (combo), _(g_enum_get_value (type, THUNAR_IMAGE_PREVIEW_MODE_STANDALONE)->value_nick));
+  gtk_combo_box_text_append_text (GTK_COMBO_BOX_TEXT (combo), _(g_enum_get_value (type, THUNAR_IMAGE_PREVIEW_MODE_EMBEDDED)->value_nick));
   g_type_class_unref (type);
   g_object_bind_property_full (G_OBJECT (dialog->preferences),
                                "misc-image-preview-mode",
@@ -1079,12 +1077,9 @@ thunar_preferences_dialog_init (ThunarPreferencesDialog *dialog)
 
   combo = gtk_combo_box_text_new ();
   type = g_type_class_ref (THUNAR_TYPE_USE_PARTIAL_MODE);
-  gtk_combo_box_text_append_text (GTK_COMBO_BOX_TEXT (combo),
-                                  g_enum_get_value (type, THUNAR_USE_PARTIAL_MODE_DISABLED)->value_nick);
-  gtk_combo_box_text_append_text (GTK_COMBO_BOX_TEXT (combo),
-                                  g_enum_get_value (type, THUNAR_USE_PARTIAL_MODE_REMOTE_ONLY)->value_nick);
-  gtk_combo_box_text_append_text (GTK_COMBO_BOX_TEXT (combo),
-                                  g_enum_get_value (type, THUNAR_USE_PARTIAL_MODE_ALWAYS)->value_nick);
+  gtk_combo_box_text_append_text (GTK_COMBO_BOX_TEXT (combo), _(g_enum_get_value (type, THUNAR_USE_PARTIAL_MODE_DISABLED)->value_nick));
+  gtk_combo_box_text_append_text (GTK_COMBO_BOX_TEXT (combo), _(g_enum_get_value (type, THUNAR_USE_PARTIAL_MODE_REMOTE_ONLY)->value_nick));
+  gtk_combo_box_text_append_text (GTK_COMBO_BOX_TEXT (combo), _(g_enum_get_value (type, THUNAR_USE_PARTIAL_MODE_ALWAYS)->value_nick));
   g_type_class_unref (type);
   g_object_bind_property_full (G_OBJECT (dialog->preferences),
                                "misc-transfer-use-partial",
@@ -1113,12 +1108,9 @@ thunar_preferences_dialog_init (ThunarPreferencesDialog *dialog)
 
   combo = gtk_combo_box_text_new ();
   type = g_type_class_ref (THUNAR_TYPE_VERIFY_FILE_MODE);
-  gtk_combo_box_text_append_text (GTK_COMBO_BOX_TEXT (combo),
-                                  g_enum_get_value (type, THUNAR_VERIFY_FILE_MODE_DISABLED)->value_nick);
-  gtk_combo_box_text_append_text (GTK_COMBO_BOX_TEXT (combo),
-                                  g_enum_get_value (type, THUNAR_VERIFY_FILE_MODE_REMOTE_ONLY)->value_nick);
-  gtk_combo_box_text_append_text (GTK_COMBO_BOX_TEXT (combo),
-                                  g_enum_get_value (type, THUNAR_VERIFY_FILE_MODE_ALWAYS)->value_nick);
+  gtk_combo_box_text_append_text (GTK_COMBO_BOX_TEXT (combo), _(g_enum_get_value (type, THUNAR_VERIFY_FILE_MODE_DISABLED)->value_nick));
+  gtk_combo_box_text_append_text (GTK_COMBO_BOX_TEXT (combo), _(g_enum_get_value (type, THUNAR_VERIFY_FILE_MODE_REMOTE_ONLY)->value_nick));
+  gtk_combo_box_text_append_text (GTK_COMBO_BOX_TEXT (combo), _(g_enum_get_value (type, THUNAR_VERIFY_FILE_MODE_ALWAYS)->value_nick));
   g_type_class_unref (type);
   g_object_bind_property_full (G_OBJECT (dialog->preferences),
                                "misc-transfer-verify-file",
