diff --git a/configure.ac.in b/configure.ac.in
index 439c47a7..4c9999d8 100644
--- a/configure.ac.in
+++ b/configure.ac.in
@@ -169,8 +169,8 @@ XDT_CHECK_PACKAGE([GLIB], [glib-2.0], [2.66.0])
 XDT_CHECK_PACKAGE([GIO], [gio-2.0], [2.66.0])
 XDT_CHECK_PACKAGE([GTHREAD], [gthread-2.0], [2.66.0])
 XDT_CHECK_PACKAGE([GMODULE], [gmodule-2.0], [2.66.0])
-XDT_CHECK_PACKAGE([GTK], [gtk+-3.0], [3.22.0])
-XDT_CHECK_PACKAGE([GDK_PIXBUF], [gdk-pixbuf-2.0], [2.14.0])
+XDT_CHECK_PACKAGE([GTK], [gtk+-3.0], [3.24.0])
+XDT_CHECK_PACKAGE([GDK_PIXBUF], [gdk-pixbuf-2.0], [2.40.0])
 XDT_CHECK_PACKAGE([LIBXFCE4UTIL], [libxfce4util-1.0], [4.17.2])
 XDT_CHECK_PACKAGE([LIBXFCE4UI], [libxfce4ui-2], [4.17.6])
 XDT_CHECK_PACKAGE([LIBXFCE4KBD_PRIVATE], [libxfce4kbd-private-3], [4.17.2])
@@ -194,7 +194,7 @@ dnl **********************************
 dnl *** Optional GIO UNIX features ***
 dnl **********************************
 XDT_CHECK_OPTIONAL_PACKAGE([GIO_UNIX], [gio-unix-2.0],
-                           [2.30.0], [gio-unix], [GIO UNIX features])
+                           [2.66.0], [gio-unix], [GIO UNIX features])
 
 dnl ***************************************************************
 dnl *** Optional support for GUDev (required for thunar-volman) ***
diff --git a/thunar/thunar-properties-dialog.c b/thunar/thunar-properties-dialog.c
index 7cccdaa1..aaa1720c 100644
--- a/thunar/thunar-properties-dialog.c
+++ b/thunar/thunar-properties-dialog.c
@@ -1342,7 +1342,6 @@ thunar_properties_dialog_update_single (ThunarPropertiesDialog *dialog)
       gtk_widget_hide (dialog->location_label);
     }
 
-#if GLIB_CHECK_VERSION (2, 65, 2)
   /* update the created time */
   date = thunar_file_get_date_string (file, THUNAR_FILE_DATE_CREATED, date_style, date_custom_style);
   if (G_LIKELY (date != NULL))
@@ -1352,7 +1351,6 @@ thunar_properties_dialog_update_single (ThunarPropertiesDialog *dialog)
       g_free (date);
     }
   else
-#endif
     {
       gtk_widget_hide (dialog->created_label);
     }
