diff --git a/thunar/thunar-preferences-dialog.c b/thunar/thunar-preferences-dialog.c
index e029b804..32ff2ee8 100644
--- a/thunar/thunar-preferences-dialog.c
+++ b/thunar/thunar-preferences-dialog.c
@@ -22,8 +22,8 @@
 #endif
 
 #include <libxfce4ui/libxfce4ui.h>
-#include <libxfce4kbd-private-3/libxfce4kbd-private/xfce-shortcuts-editor.h>
-#include <libxfce4kbd-private-3/libxfce4kbd-private/xfce-shortcuts-editor-dialog.h>
+#include <libxfce4kbd-private/xfce-shortcuts-editor.h>
+#include <libxfce4kbd-private/xfce-shortcuts-editor-dialog.h>
 
 #include <thunar/thunar-application.h>
 #include <thunar/thunar-compact-view.h>
