diff --git a/thunar/thunar-window.c b/thunar/thunar-window.c
index 36ac4195..e2beec1a 100644
--- a/thunar/thunar-window.c
+++ b/thunar/thunar-window.c
@@ -339,6 +339,7 @@ struct _ThunarWindow
   GtkWidget              *view_box;
   GtkWidget              *trash_infobar;
   GtkWidget              *trash_infobar_restore_button;
+  GtkWidget              *trash_infobar_empty_button;
 
   /* split view panes */
   GtkWidget              *paned_notebooks;
@@ -806,7 +807,7 @@ thunar_window_init (ThunarWindow *window)
   window->trash_infobar = gtk_info_bar_new();
   gtk_grid_attach (GTK_GRID (window->view_box), window->trash_infobar, 0, 0, 1, 1);
   window->trash_infobar_restore_button = gtk_info_bar_add_button (GTK_INFO_BAR (window->trash_infobar), "Restore Selected Items", RESTORE);
-  gtk_info_bar_add_button (GTK_INFO_BAR (window->trash_infobar), "Empty Trash", EMPTY);
+  window->trash_infobar_empty_button = gtk_info_bar_add_button (GTK_INFO_BAR (window->trash_infobar), "Empty Trash", EMPTY);
   g_signal_connect (window->trash_infobar,
                     "response",
                     G_CALLBACK (thunar_window_trash_infobar_clicked),
@@ -3767,6 +3768,9 @@ thunar_window_notify_loading (ThunarView   *view,
         {
           gdk_window_set_cursor (gtk_widget_get_window (GTK_WIDGET (window)), NULL);
         }
+
+      /* Set trash infobar's `empty trash` button sensitivity */
+      gtk_widget_set_sensitive (window->trash_infobar_empty_button, thunar_file_get_item_count (window->current_directory) > 0);
     }
 }
 
@@ -4462,4 +4466,4 @@ thunar_window_trash_selection_updated (ThunarWindow *window)
     gtk_widget_set_sensitive (window->trash_infobar_restore_button, TRUE);
   else
     gtk_widget_set_sensitive (window->trash_infobar_restore_button, FALSE);
-}
+}
\ No newline at end of file
