diff --git a/thunar/thunar-window.c b/thunar/thunar-window.c
index 861516c3..9e0ea203 100644
--- a/thunar/thunar-window.c
+++ b/thunar/thunar-window.c
@@ -293,6 +293,9 @@ static void      thunar_window_trash_infobar_clicked           (GtkInfoBar
                                                                 gint                    response_id,
                                                                 ThunarWindow           *window);
 static void      thunar_window_trash_selection_updated         (ThunarWindow           *window);
+static void      thunar_window_notebook_add_new_tab            (ThunarWindow           *window,
+                                                                ThunarFile             *directory,
+                                                                gboolean                force_switch_to_new_tab);
 
 
 
@@ -2255,8 +2258,10 @@ thunar_window_split_view_is_active (ThunarWindow *window)
 
 
 
-void thunar_window_notebook_open_new_tab (ThunarWindow *window,
-                                          ThunarFile   *directory)
+static void
+thunar_window_notebook_add_new_tab (ThunarWindow *window,
+                                    ThunarFile   *directory,
+                                    gboolean      force_switch_to_new_tab)
 {
   ThunarHistory *history = NULL;
   GtkWidget     *view;
@@ -2277,7 +2282,7 @@ void thunar_window_notebook_open_new_tab (ThunarWindow *window,
 
   /* switch to the new view */
   g_object_get (G_OBJECT (window->preferences), "misc-switch-to-new-tab", &switch_to_new_tab, NULL);
-  if (switch_to_new_tab == TRUE)
+  if (switch_to_new_tab == TRUE || force_switch_to_new_tab == TRUE)
     {
       page_num = gtk_notebook_page_num (GTK_NOTEBOOK (window->notebook_selected), view);
       gtk_notebook_set_current_page (GTK_NOTEBOOK (window->notebook_selected), page_num);
@@ -2289,6 +2294,14 @@ void thunar_window_notebook_open_new_tab (ThunarWindow *window,
 
 
 
+void thunar_window_notebook_open_new_tab (ThunarWindow *window,
+                                          ThunarFile   *directory)
+{
+  thunar_window_notebook_add_new_tab (window, directory, FALSE /* don't override `misc-switch-to-new-tab` preference */);
+}
+
+
+
 void
 thunar_window_update_directories (ThunarWindow *window,
                                   ThunarFile   *old_directory,
@@ -2625,7 +2638,7 @@ thunar_window_action_open_new_tab (ThunarWindow *window,
                                    GtkWidget    *menu_item)
 {
   /* open new tab with current directory as default */
-  thunar_window_notebook_open_new_tab (window, thunar_window_get_current_directory (window));
+  thunar_window_notebook_add_new_tab (window, thunar_window_get_current_directory (window), TRUE /* force tab switch */);
 }
 
 
