diff --git a/thunar/thunar-preferences-dialog.c b/thunar/thunar-preferences-dialog.c
index ead44256..8d3a35e1 100644
--- a/thunar/thunar-preferences-dialog.c
+++ b/thunar/thunar-preferences-dialog.c
@@ -292,30 +292,21 @@ thunar_preferences_dialog_init (ThunarPreferencesDialog *dialog)
   gtk_window_set_resizable (GTK_WINDOW (dialog), FALSE);
   gtk_window_set_title (GTK_WINDOW (dialog), _("File Manager Preferences"));
 
-#if LIBXFCE4UI_CHECK_VERSION (4, 15, 1)
   xfce_titled_dialog_create_action_area (XFCE_TITLED_DIALOG (dialog));
-#endif
 
   /* add the "Close" button */
   button = gtk_button_new_with_mnemonic (_("_Close"));
   image = gtk_image_new_from_icon_name ("window-close", GTK_ICON_SIZE_BUTTON);
   gtk_button_set_image (GTK_BUTTON (button), image);
-#if LIBXFCE4UI_CHECK_VERSION (4, 15, 1)
   xfce_titled_dialog_add_action_widget (XFCE_TITLED_DIALOG (dialog), button, GTK_RESPONSE_CLOSE);
-#else
-  gtk_dialog_add_action_widget (GTK_DIALOG (dialog), button, GTK_RESPONSE_CLOSE);
-#endif
+
   gtk_widget_show (button);
 
   /* add the "Help" button */
   button = gtk_button_new_with_mnemonic (_("_Help"));
   image = gtk_image_new_from_icon_name ("help-browser", GTK_ICON_SIZE_BUTTON);
   gtk_button_set_image (GTK_BUTTON (button), image);
-#if LIBXFCE4UI_CHECK_VERSION (4, 15, 1)
   xfce_titled_dialog_add_action_widget (XFCE_TITLED_DIALOG (dialog), button, GTK_RESPONSE_HELP);
-#else
-  gtk_dialog_add_action_widget (GTK_DIALOG (dialog), button, GTK_RESPONSE_HELP);
-#endif
   gtk_widget_show (button);
 
   notebook = gtk_notebook_new ();
diff --git a/thunar/thunar-progress-dialog.c b/thunar/thunar-progress-dialog.c
index 2fde19de..0577178c 100644
--- a/thunar/thunar-progress-dialog.c
+++ b/thunar/thunar-progress-dialog.c
@@ -272,11 +272,7 @@ thunar_progress_dialog_job_finished (ThunarProgressDialog *dialog,
   if (n_views == SCROLLVIEW_THRESHOLD-1)
     {
       /* reparent the content box */
-#if LIBXFCE4UI_CHECK_VERSION (4, 13, 2)
       xfce_widget_reparent (dialog->content_box, dialog->vbox);
-#else
-      gtk_widget_reparent (dialog->content_box, dialog->vbox);
-#endif
 
       /* destroy the scroll win */
       gtk_widget_destroy (dialog->scrollwin);
@@ -403,11 +399,8 @@ thunar_progress_dialog_add_job (ThunarProgressDialog *dialog,
       gtk_widget_show (viewport);
 
       /* move the content box into the viewport */
-#if LIBXFCE4UI_CHECK_VERSION (4, 13, 2)
       xfce_widget_reparent (dialog->content_box, viewport);
-#else
-      gtk_widget_reparent (dialog->content_box, viewport);
-#endif
+
     }
 
   g_signal_connect_swapped (view, "need-attention",
