diff --git a/thunar/thunar-icon-factory.c b/thunar/thunar-icon-factory.c
index a566d0ce..020a9b19 100644
--- a/thunar/thunar-icon-factory.c
+++ b/thunar/thunar-icon-factory.c
@@ -937,8 +937,7 @@ thunar_icon_factory_load_file_icon (ThunarIconFactory  *factory,
     }
 
   /* check if thumbnails are enabled and we can display a thumbnail for the item */
-  if (thunar_icon_factory_get_show_thumbnail (factory, file)
-      && (thunar_file_is_regular (file) || thunar_file_is_directory (file)) )
+  if (thunar_icon_factory_get_show_thumbnail (factory, file))
     {
       /* determine the preview icon first */
       gicon = thunar_file_get_preview_icon (file);
