diff --git a/NEWS b/NEWS
index cb08c556..2c9f1634 100644
--- a/NEWS
+++ b/NEWS
@@ -121,7 +121,6 @@ quickly
 - Improve usability of Computer:// in list view (Issue #297)
 - Select the correct bookmark after opening a new window (Issue #106)
 - Shortcut-view regression: Middle mouse click on folder did not work
-- Scroll to the correct height after changing to parent dir (Issue #531)
 - Make use of xfce icon "org.xfce.settings.default-applications"
 - Avoid excessively checking whether files are executables (Issue #180)
 - Add Option to keep focus while opening folder in new tab (Issue #364)
