diff --git a/thunar/thunar-application.c b/thunar/thunar-application.c
index f4877d3b..f2fa0b8e 100644
--- a/thunar/thunar-application.c
+++ b/thunar/thunar-application.c
@@ -134,6 +134,7 @@ static gboolean       thunar_application_dbus_register          (GApplication
 static void           thunar_application_load_css               (void);
 static void           thunar_application_accel_map_changed      (ThunarApplication      *application);
 static gboolean       thunar_application_accel_map_save         (gpointer                user_data);
+static gboolean       thunar_application_accel_map_load         (gpointer                user_data);
 static void           thunar_application_collect_and_launch     (ThunarApplication      *application,
                                                                  gpointer                parent,
                                                                  const gchar            *icon_name,
@@ -196,6 +197,7 @@ struct _ThunarApplication
 
   gboolean                        daemon;
 
+  guint                           accel_map_load_id;
   guint                           accel_map_save_id;
   GtkAccelMap                    *accel_map;
 
@@ -353,7 +355,6 @@ thunar_application_startup (GApplication *gapp)
 #ifdef HAVE_GUDEV
   static const gchar *subsystems[] = { "block", "input", "usb", NULL };
 #endif
-  gchar              *path;
 
   /* initialize the application */
   application->preferences = thunar_preferences_get ();
@@ -375,19 +376,8 @@ thunar_application_startup (GApplication *gapp)
   /* connect to the session manager */
   application->session_client = thunar_session_client_new (opt_sm_client_id);
 
-  /* check if we have a saved accel map */
-  path = xfce_resource_lookup (XFCE_RESOURCE_CONFIG, ACCEL_MAP_PATH);
-  if (G_LIKELY (path != NULL))
-    {
-      /* load the accel map */
-      gtk_accel_map_load (path);
-      g_free (path);
-    }
-
-  /* watch for changes */
-  application->accel_map = gtk_accel_map_get ();
-  g_signal_connect_swapped (G_OBJECT (application->accel_map), "changed",
-      G_CALLBACK (thunar_application_accel_map_changed), application);
+  /* schedule accel map load and update windows when finished, this way empty but active accelerators are preserved */
+  application->accel_map_load_id = gdk_threads_add_idle_full (G_PRIORITY_LOW, thunar_application_accel_map_load, application, NULL);
 
   thunar_application_load_css ();
 }
@@ -409,6 +399,9 @@ thunar_application_shutdown (GApplication *gapp)
       thunar_application_accel_map_save (application);
     }
 
+  if (G_UNLIKELY (application->accel_map_load_id != 0))
+    g_source_remove (application->accel_map_load_id);
+
   if (application->accel_map != NULL)
     g_object_unref (G_OBJECT (application->accel_map));
 
@@ -753,6 +746,33 @@ thunar_application_accel_map_save (gpointer user_data)
 
 
 
+static gboolean
+thunar_application_accel_map_load (gpointer user_data)
+{
+  ThunarApplication *application = user_data;
+  gchar             *path;
+
+  /* watch for changes */
+  application->accel_map = gtk_accel_map_get ();
+  g_signal_connect_swapped (G_OBJECT (application->accel_map), "changed",
+                            G_CALLBACK (thunar_application_accel_map_changed), application);
+
+  /* check if we have a saved accel map */
+  path = xfce_resource_lookup (XFCE_RESOURCE_CONFIG, ACCEL_MAP_PATH);
+  if (G_LIKELY (path != NULL))
+    {
+      /* load the accel map */
+      gtk_accel_map_load (path);
+      g_free (path);
+    }
+
+  application->accel_map_load_id = 0;
+
+  return FALSE;
+}
+
+
+
 static void
 thunar_application_accel_map_changed (ThunarApplication *application)
 {
