diff --git a/thunar/thunar-location-buttons.c b/thunar/thunar-location-buttons.c
index 3b91e029..ccfc908e 100644
--- a/thunar/thunar-location-buttons.c
+++ b/thunar/thunar-location-buttons.c
@@ -410,10 +410,10 @@ thunar_location_buttons_set_current_directory (ThunarNavigator *navigator,
   if (G_UNLIKELY (buttons->current_directory == current_directory))
     return;
 
-  /* check if we already have a button for that directory */
+  /* check if we already have a visible button for that directory */
   for (lp = buttons->list; lp != NULL; lp = lp->next)
     {
-      if (thunar_location_button_get_file (lp->data) == current_directory)
+      if (thunar_location_button_get_file (lp->data) == current_directory && gtk_widget_get_child_visible (GTK_WIDGET (lp->data)))
         {
           /* fake a "clicked" event for that button */
           gtk_button_clicked (GTK_BUTTON (lp->data));
