diff --git a/thunar/thunar-compact-view.c b/thunar/thunar-compact-view.c
index 5556ada5..70a93fda 100644
--- a/thunar/thunar-compact-view.c
+++ b/thunar/thunar-compact-view.c
@@ -73,6 +73,8 @@ thunar_compact_view_class_init (ThunarCompactViewClass *klass)
 static void
 thunar_compact_view_init (ThunarCompactView *compact_view)
 {
+  gboolean max_chars;
+
  /* initialize the icon view properties */
   exo_icon_view_set_margin (EXO_ICON_VIEW (gtk_bin_get_child (GTK_BIN (compact_view))), 3);
   exo_icon_view_set_layout_mode (EXO_ICON_VIEW (gtk_bin_get_child (GTK_BIN (compact_view))), EXO_ICON_VIEW_LAYOUT_COLS);
@@ -83,13 +85,20 @@ thunar_compact_view_init (ThunarCompactView *compact_view)
                 "ypad", 2u,
                 NULL);
 
-  /* setup the name renderer (wrap only very long names) */
+  g_object_get (G_OBJECT (THUNAR_STANDARD_VIEW (compact_view)->preferences), "misc-compact-view-max-chars", &max_chars, NULL);
+
+  /* setup the name renderer */
   g_object_set (G_OBJECT (THUNAR_STANDARD_VIEW (compact_view)->name_renderer),
-                "wrap-mode", PANGO_WRAP_WORD_CHAR,
-                "wrap-width", 1280,
                 "xalign", 0.0f,
                 "yalign", 0.5f,
                 NULL);
+
+  /* setup ellipsization */
+  if (max_chars > 0)
+    g_object_set (G_OBJECT (THUNAR_STANDARD_VIEW (compact_view)->name_renderer),
+                  "ellipsize", PANGO_ELLIPSIZE_MIDDLE,
+                  "width-chars", max_chars,
+                  NULL);
 }
 
 
diff --git a/thunar/thunar-preferences.c b/thunar/thunar-preferences.c
index 28d0d5f0..723e95c1 100644
--- a/thunar/thunar-preferences.c
+++ b/thunar/thunar-preferences.c
@@ -119,6 +119,7 @@ enum
   PROP_TREE_ICON_SIZE,
   PROP_MISC_SWITCH_TO_NEW_TAB,
   PROP_MISC_VERTICAL_SPLIT_PANE,
+  PROP_MISC_COMPACT_VIEW_MAX_CHARS,
   N_PROPERTIES,
 };
 
@@ -1077,6 +1078,19 @@ thunar_preferences_class_init (ThunarPreferencesClass *klass)
                           FALSE,
                           EXO_PARAM_READWRITE);
 
+  /**
+   * ThunarPreferences:compact-view-max-chars
+   *
+   * The ellipsization threshold for compact view. Negative values disable ellipsization.
+   **/
+  preferences_props[PROP_MISC_COMPACT_VIEW_MAX_CHARS] =
+      g_param_spec_int ("misc-compact-view-max-chars",
+                        "MiscCompactViewMaxChars",
+                        NULL,
+                        G_MININT, G_MAXINT,
+                        50,
+                        EXO_PARAM_READWRITE);
+
   /* install all properties */
   g_object_class_install_properties (gobject_class, N_PROPERTIES, preferences_props);
 }
