diff --git a/thunar/thunar-gtk-extensions.c b/thunar/thunar-gtk-extensions.c
index f2197c0f..ce752e08 100644
--- a/thunar/thunar-gtk-extensions.c
+++ b/thunar/thunar-gtk-extensions.c
@@ -218,6 +218,33 @@ thunar_gtk_menu_run_at_event (GtkMenu *menu, GdkEvent *event)
 
 
 
+/**
+ * thunar_gtk_menu_hide_accel_labels:
+ * @menu : a #GtkMenu instance
+ *
+ * Will hide the accel_labels of all menu items of this menu and its submenus
+ **/
+void
+thunar_gtk_menu_hide_accel_labels (GtkMenu *menu)
+{
+  GList     *children, *lp;
+  GtkWidget *submenu;
+
+  _thunar_return_if_fail (GTK_IS_MENU (menu));
+
+  children = gtk_container_get_children (GTK_CONTAINER (menu));
+  for (lp = children; lp != NULL; lp = lp->next)
+    {
+      xfce_gtk_menu_item_set_accel_label (lp->data, NULL);
+      submenu = gtk_menu_item_get_submenu (lp->data);
+      if (submenu != NULL)
+        thunar_gtk_menu_hide_accel_labels (GTK_MENU (submenu));
+    }
+  g_list_free (children);
+}
+
+
+
 /**
  * thunar_gtk_widget_set_tooltip:
  * @widget : a #GtkWidget for which to set the tooltip.
diff --git a/thunar/thunar-gtk-extensions.h b/thunar/thunar-gtk-extensions.h
index bb546cc6..9d00bf13 100644
--- a/thunar/thunar-gtk-extensions.h
+++ b/thunar/thunar-gtk-extensions.h
@@ -27,11 +27,13 @@ G_BEGIN_DECLS;
 
 void             thunar_gtk_label_set_a11y_relation           (GtkLabel           *label,
                                                                GtkWidget          *widget);
+
 void             thunar_gtk_menu_clean                        (GtkMenu            *menu);
 void             thunar_gtk_menu_run                          (GtkMenu            *menu);
-
 void             thunar_gtk_menu_run_at_event                 (GtkMenu            *menu,
                                                                GdkEvent           *event);
+void             thunar_gtk_menu_hide_accel_labels            (GtkMenu            *menu);
+
 void             thunar_gtk_widget_set_tooltip                (GtkWidget          *widget,
                                                                const gchar        *format,
                                                                ...) G_GNUC_PRINTF (2, 3);
diff --git a/thunar/thunar-location-buttons.c b/thunar/thunar-location-buttons.c
index c0b4a393..041787de 100644
--- a/thunar/thunar-location-buttons.c
+++ b/thunar/thunar-location-buttons.c
@@ -1235,7 +1235,7 @@ thunar_location_buttons_context_menu (ThunarLocationButton  *button,
                                             | THUNAR_MENU_SECTION_CUSTOM_ACTIONS
                                             | THUNAR_MENU_SECTION_PROPERTIES);
     }
-  thunar_menu_hide_accel_labels (context_menu);
+  thunar_gtk_menu_hide_accel_labels (GTK_MENU (context_menu));
   gtk_widget_show_all (GTK_WIDGET (context_menu));
   window = gtk_widget_get_toplevel (GTK_WIDGET (buttons));
   thunar_window_redirect_menu_tooltips_to_statusbar (THUNAR_WINDOW (window), GTK_MENU (context_menu));
diff --git a/thunar/thunar-menu.c b/thunar/thunar-menu.c
index 71149e33..107dbcaf 100644
--- a/thunar/thunar-menu.c
+++ b/thunar/thunar-menu.c
@@ -383,24 +383,3 @@ thunar_menu_get_launcher (ThunarMenu *menu)
   _thunar_return_val_if_fail (THUNAR_IS_MENU (menu), NULL);
   return GTK_WIDGET (menu->launcher);
 }
-
-
-
-/**
- * thunar_menu_hide_accel_labels:
- * @menu : a #ThunarMenu instance
- *
- * Will hide the accel_labels of all menu items of this menu
- **/
-void
-thunar_menu_hide_accel_labels (ThunarMenu *menu)
-{
-  GList *children, *lp;
-
-  _thunar_return_if_fail (THUNAR_IS_MENU (menu));
-
-  children = gtk_container_get_children (GTK_CONTAINER (menu));
-  for (lp = children; lp != NULL; lp = lp->next)
-    xfce_gtk_menu_item_set_accel_label (lp->data, NULL);
-  g_list_free (children);
-}
diff --git a/thunar/thunar-menu.h b/thunar/thunar-menu.h
index 44c25d5c..63eb0d84 100644
--- a/thunar/thunar-menu.h
+++ b/thunar/thunar-menu.h
@@ -75,7 +75,6 @@ GType      thunar_menu_get_type          (void) G_GNUC_CONST;
 gboolean   thunar_menu_add_sections      (ThunarMenu         *menu,
                                           ThunarMenuSections  menu_sections);
 GtkWidget* thunar_menu_get_launcher      (ThunarMenu         *menu);
-void       thunar_menu_hide_accel_labels (ThunarMenu         *menu);
 
 G_END_DECLS;
 
diff --git a/thunar/thunar-renamer-dialog.c b/thunar/thunar-renamer-dialog.c
index 6cf187ce..2df22dbb 100644
--- a/thunar/thunar-renamer-dialog.c
+++ b/thunar/thunar-renamer-dialog.c
@@ -922,7 +922,7 @@ thunar_renamer_dialog_context_menu (ThunarRenamerDialog *renamer_dialog)
   thunar_renamer_dialog_append_menu_item (renamer_dialog, GTK_MENU_SHELL (menu), THUNAR_RENAMER_ACTION_REMOVE_FILES);
   xfce_gtk_menu_append_separator (GTK_MENU_SHELL (menu));
   thunar_menu_add_sections (menu, THUNAR_MENU_SECTION_PROPERTIES);
-  thunar_menu_hide_accel_labels (menu);
+  thunar_gtk_menu_hide_accel_labels (GTK_MENU (menu));
   gtk_widget_show_all (GTK_WIDGET (menu));
 
   thunar_gtk_menu_run (GTK_MENU (menu));
diff --git a/thunar/thunar-shortcuts-view.c b/thunar/thunar-shortcuts-view.c
index eac9368f..0a923396 100644
--- a/thunar/thunar-shortcuts-view.c
+++ b/thunar/thunar-shortcuts-view.c
@@ -1302,7 +1302,7 @@ thunar_shortcuts_view_context_menu (ThunarShortcutsView *view,
                               (GDestroyNotify) gtk_tree_row_reference_free);
   }
 
-  thunar_menu_hide_accel_labels (context_menu);
+  thunar_gtk_menu_hide_accel_labels (GTK_MENU (context_menu));
   gtk_widget_show_all (GTK_WIDGET (context_menu));
   window = gtk_widget_get_toplevel (GTK_WIDGET (view));
   thunar_window_redirect_menu_tooltips_to_statusbar (THUNAR_WINDOW (window), GTK_MENU (context_menu));
diff --git a/thunar/thunar-standard-view.c b/thunar/thunar-standard-view.c
index f21b12c7..1251c083 100644
--- a/thunar/thunar-standard-view.c
+++ b/thunar/thunar-standard-view.c
@@ -3877,7 +3877,7 @@ thunar_standard_view_context_menu (ThunarStandardView *standard_view)
       thunar_menu_add_sections (context_menu, THUNAR_MENU_SECTION_ZOOM
                                             | THUNAR_MENU_SECTION_PROPERTIES);
     }
-  thunar_menu_hide_accel_labels (context_menu);
+  thunar_gtk_menu_hide_accel_labels (GTK_MENU (context_menu));
   gtk_widget_show_all (GTK_WIDGET (context_menu));
   thunar_window_redirect_menu_tooltips_to_statusbar (THUNAR_WINDOW (window), GTK_MENU (context_menu));
 
diff --git a/thunar/thunar-tree-view.c b/thunar/thunar-tree-view.c
index 185c45ae..74b9ef94 100644
--- a/thunar/thunar-tree-view.c
+++ b/thunar/thunar-tree-view.c
@@ -1283,7 +1283,7 @@ thunar_tree_view_context_menu (ThunarTreeView *view,
         thunar_menu_add_sections (context_menu, THUNAR_MENU_SECTION_PROPERTIES);
     }
 
-  thunar_menu_hide_accel_labels (context_menu);
+  thunar_gtk_menu_hide_accel_labels (GTK_MENU (context_menu));
   gtk_widget_show_all (GTK_WIDGET (context_menu));
   window = gtk_widget_get_toplevel (GTK_WIDGET (view));
   thunar_window_redirect_menu_tooltips_to_statusbar (THUNAR_WINDOW (window), GTK_MENU (context_menu));
