diff --git a/thunar/thunar-job-operation.c b/thunar/thunar-job-operation.c
index aaab0c5a..3e887813 100644
--- a/thunar/thunar-job-operation.c
+++ b/thunar/thunar-job-operation.c
@@ -21,6 +21,7 @@
 #include <thunar/thunar-enum-types.h>
 #include <thunar/thunar-io-jobs.h>
 #include <thunar/thunar-job-operation.h>
+#include <thunar/thunar-preferences.h>
 #include <thunar/thunar-private.h>
 
 /**
@@ -71,6 +72,7 @@ G_DEFINE_TYPE (ThunarJobOperation, thunar_job_operation, G_TYPE_OBJECT)
 
 /* List of job operations which were logged */
 static GList *job_operation_list = NULL;
+static gint   job_operation_list_max_size;
 
 /* List pointer to the operation which can be undone */
 static GList *lp_undo_job_operation = NULL;
@@ -81,11 +83,16 @@ static GList *lp_redo_job_operation = NULL;
 static void
 thunar_job_operation_class_init (ThunarJobOperationClass *klass)
 {
-  GObjectClass *gobject_class;
+  GObjectClass      *gobject_class;
+  ThunarPreferences *preferences;
 
   gobject_class = G_OBJECT_CLASS (klass);
   gobject_class->dispose = thunar_job_operation_dispose;
   gobject_class->finalize = thunar_job_operation_finalize;
+
+  preferences = thunar_preferences_get ();
+  g_object_get (G_OBJECT (preferences), "misc-undo-redo-history-size", &job_operation_list_max_size, NULL);
+  g_object_unref (preferences);
 }
 
 
@@ -239,6 +246,14 @@ thunar_job_operation_commit (ThunarJobOperation *job_operation)
   /* reset the undo pointer to latest operation and clear the redo pointer */
   lp_undo_job_operation = g_list_last (job_operation_list);
   lp_redo_job_operation = NULL;
+
+  /* Limit the size of the list */
+  if (job_operation_list_max_size != -1 && g_list_length (job_operation_list) > (guint)job_operation_list_max_size)
+    {
+      GList* first = g_list_first (job_operation_list);
+      job_operation_list = g_list_remove_link (job_operation_list, first);
+      g_list_free_full (first, g_object_unref);
+    }
 }
 
 
diff --git a/thunar/thunar-preferences.c b/thunar/thunar-preferences.c
index 2c9e9665..88535dfd 100644
--- a/thunar/thunar-preferences.c
+++ b/thunar/thunar-preferences.c
@@ -123,6 +123,7 @@ enum
   PROP_MISC_OPEN_NEW_WINDOWS_IN_SPLIT_VIEW,
   PROP_MISC_COMPACT_VIEW_MAX_CHARS,
   PROP_MISC_HIGHLIGHTING_ENABLED,
+  PROP_MISC_UNDO_REDO_HISTORY_SIZE,
   N_PROPERTIES,
 };
 
@@ -1132,6 +1133,20 @@ thunar_preferences_class_init (ThunarPreferencesClass *klass)
                             thunar_g_vfs_metadata_is_supported (),
                             EXO_PARAM_READWRITE);
 
+  /**
+   * ThunarPreferences:misc-undo-redo-history-size
+   *
+   * Maximum number of ThunarJobOperations which can be undone/redone
+   * -1 for unlimited
+   **/
+  preferences_props[PROP_MISC_UNDO_REDO_HISTORY_SIZE] =
+      g_param_spec_int ("misc-undo-redo-history-size",
+                        "MiscUndoRedoHistorySize",
+                        NULL,
+                        -1, G_MAXINT,
+                        10,
+                        EXO_PARAM_READWRITE);
+
   /* install all properties */
   g_object_class_install_properties (gobject_class, N_PROPERTIES, preferences_props);
 }
