diff --git a/docs/reference/thunar/thunar-docs.xml b/docs/reference/thunar/thunar-docs.xml
index 33cd1546..5fb34520 100644
--- a/docs/reference/thunar/thunar-docs.xml
+++ b/docs/reference/thunar/thunar-docs.xml
@@ -55,6 +55,7 @@
   <!ENTITY ThunarSessionClient SYSTEM "xml/thunar-session-client.xml">
   <!ENTITY ThunarShortcutsPane SYSTEM "xml/thunar-shortcuts-pane.xml">
   <!ENTITY ThunarJob SYSTEM "xml/thunar-job.xml">
+  <!ENTITY ThunarJobOperation SYSTEM "xml/thunar-job-operation.xml">
   <!ENTITY ThunarJobOperation SYSTEM "xml/thunar-job-operation-history.xml">
   <!ENTITY ThunarPangoExtensions SYSTEM "xml/thunar-pango-extensions.xml">
   <!ENTITY ThunarStatusbar SYSTEM "xml/thunar-statusbar.xml">
@@ -150,6 +151,7 @@
       <xi:include href="xml/thunar-enum-types.xml"/>
       <xi:include href="xml/thunar-preferences.xml"/>
       <xi:include href="xml/thunar-user.xml"/>
+      <xi:include href="xml/thunar-job-operation.xml"/>
       <xi:include href="xml/thunar-job-operation-history.xml"/>
     </chapter>
   </part>
diff --git a/thunar/Makefile.am b/thunar/Makefile.am
index bb6d3848..ed622ec5 100644
--- a/thunar/Makefile.am
+++ b/thunar/Makefile.am
@@ -123,8 +123,8 @@ thunar_SOURCES =							\
 	thunar-job.h							\
 	thunar-job-operation.c						\
 	thunar-job-operation.h						\
-	thunar-job-operation-history.c						\
-	thunar-job-operation-history.h						\
+	thunar-job-operation-history.c					\
+	thunar-job-operation-history.h					\
 	thunar-list-model.c						\
 	thunar-list-model.h						\
 	thunar-location-bar.c						\
diff --git a/thunar/thunar-application.h b/thunar/thunar-application.h
index fee97f8f..00b34ffd 100644
--- a/thunar/thunar-application.h
+++ b/thunar/thunar-application.h
@@ -23,7 +23,7 @@
 #ifndef __THUNAR_APPLICATION_H__
 #define __THUNAR_APPLICATION_H__
 
-#include <thunar/thunar-job-operation-history.h>
+#include <thunar/thunar-job-operation.h>
 #include <thunar/thunar-window.h>
 #include <thunar/thunar-thumbnail-cache.h>
 
diff --git a/thunar/thunar-job-operation-history.c b/thunar/thunar-job-operation-history.c
index a7f4d39d..0b8f7754 100644
--- a/thunar/thunar-job-operation-history.c
+++ b/thunar/thunar-job-operation-history.c
@@ -41,7 +41,7 @@ enum
 };
 
 
-static void thunar_job_operation_history_finalize     (GObject *object);
+static void thunar_job_operation_history_finalize     (GObject    *object);
 static void thunar_job_operation_history_get_property (GObject    *object,
                                                        guint       prop_id,
                                                        GValue     *value,
@@ -194,6 +194,8 @@ thunar_job_operation_history_get_default (void)
 void
 thunar_job_operation_history_commit (ThunarJobOperation *job_operation)
 {
+  GList* new_list = NULL;
+
   _thunar_return_if_fail (THUNAR_IS_JOB_OPERATION (job_operation));
 
   if (thunar_job_operation_get_kind (job_operation) == THUNAR_JOB_OPERATION_KIND_TRASH)
@@ -203,13 +205,12 @@ thunar_job_operation_history_commit (ThunarJobOperation *job_operation)
       thunar_job_operation_set_end_timestamp (job_operation, g_get_real_time () / (gint64) 1e6);
     }
 
-  /* When a new operation is added, drop all previous operations which where undone from the list */
+  /* When a new operation is added, drop all previous operations which were undone from the list */
   if (job_operation_history->lp_redo != NULL)
     {
-      GList* new_list = NULL;
       for (GList* lp = job_operation_history->job_operation_list;
-          lp != NULL && lp != job_operation_history->lp_redo;
-          lp = lp->next)
+           lp != NULL && lp != job_operation_history->lp_redo;
+           lp = lp->next)
         new_list = g_list_append (new_list, g_object_ref (lp->data));
       g_list_free_full (job_operation_history->job_operation_list, g_object_unref);
       job_operation_history->job_operation_list = new_list;
@@ -230,6 +231,7 @@ thunar_job_operation_history_commit (ThunarJobOperation *job_operation)
       g_list_free_full (first, g_object_unref);
     }
 
+  /* Notify all subscribers of our properties */
   g_object_notify (G_OBJECT (job_operation_history), "can-undo");
   g_object_notify (G_OBJECT (job_operation_history), "can-redo");
 }
@@ -347,6 +349,7 @@ thunar_job_operation_history_undo (void)
     if (err == NULL)
       thunar_notify_undo (operation_marker);
 
+    /* Notify all subscribers of our properties */
     g_object_notify (G_OBJECT (job_operation_history), "can-undo");
     g_object_notify (G_OBJECT (job_operation_history), "can-redo");
 }
@@ -424,6 +427,7 @@ thunar_job_operation_history_redo (void)
     if (err == NULL)
       thunar_notify_redo (operation_marker);
 
+    /* Notify all subscribers of our properties */
     g_object_notify (G_OBJECT (job_operation_history), "can-undo");
     g_object_notify (G_OBJECT (job_operation_history), "can-redo");
 }
diff --git a/thunar/thunar-job-operation.c b/thunar/thunar-job-operation.c
index bb3e68b6..9dcaa3a9 100644
--- a/thunar/thunar-job-operation.c
+++ b/thunar/thunar-job-operation.c
@@ -23,11 +23,11 @@
 
 /**
  * SECTION:thunar-job-operation
- * @Short_description: Stores a job operations (copy, move, trash, rename, etc.)
+ * @Short_description: Stores a job operation (copy, move, trash, rename, etc.)
  * @Title: ThunarJobOperation
  *
  * The #ThunarJobOperation class represents a single 'job operation', a file operation like copying, moving
- * trashing, renaming etc. and it's source/target locations.
+ * trashing, renaming etc. and its source/target locations.
  *
  */
 
@@ -48,7 +48,7 @@ struct _ThunarJobOperation
   GList                  *target_file_list;
 
   /* Files overwritten as a part of an operation */
-  GList                  *overwritten_files_;
+  GList                  *overwritten_files;
 
   /**
    * Optional timestampes (in seconds) which tell when the operation was started and ended.
@@ -79,7 +79,7 @@ thunar_job_operation_init (ThunarJobOperation *self)
   self->operation_kind = THUNAR_JOB_OPERATION_KIND_COPY;
   self->source_file_list = NULL;
   self->target_file_list = NULL;
-  self->overwritten_files_ = NULL;
+  self->overwritten_files = NULL;
 }
 
 
@@ -93,7 +93,7 @@ thunar_job_operation_finalize (GObject *object)
 
   g_list_free_full (op->source_file_list, g_object_unref);
   g_list_free_full (op->target_file_list, g_object_unref);
-  g_list_free_full (op->overwritten_files_, g_object_unref);
+  g_list_free_full (op->overwritten_files, g_object_unref);
 
   (*G_OBJECT_CLASS (thunar_job_operation_parent_class)->finalize) (object);
 }
@@ -174,7 +174,7 @@ thunar_job_operation_overwrite (ThunarJobOperation *job_operation,
 {
   _thunar_return_if_fail (THUNAR_IS_JOB_OPERATION (job_operation));
 
-  job_operation->overwritten_files_ = thunar_g_list_append_deep (job_operation->overwritten_files_, overwritten_file);
+  job_operation->overwritten_files = thunar_g_list_append_deep (job_operation->overwritten_files, overwritten_file);
 }
 
 
@@ -265,7 +265,7 @@ const GList*
 thunar_job_operation_get_overwritten_files (ThunarJobOperation *job_operation)
 {
   _thunar_return_val_if_fail (THUNAR_IS_JOB_OPERATION (job_operation), NULL);
-   return job_operation->overwritten_files_;
+   return job_operation->overwritten_files;
 }
 
 
diff --git a/thunar/thunar-notify.h b/thunar/thunar-notify.h
index 0a042ca6..e5092dfa 100644
--- a/thunar/thunar-notify.h
+++ b/thunar/thunar-notify.h
@@ -23,7 +23,7 @@
 
 #include <glib.h>
 #include <thunar/thunar-device.h>
-#include <thunar/thunar-job-operation-history.h>
+#include <thunar/thunar-job-operation.h>
 
 G_BEGIN_DECLS
 
