diff --git a/thunar/thunar-device-monitor.c b/thunar/thunar-device-monitor.c
index 0e827276..c79ce8c4 100644
--- a/thunar/thunar-device-monitor.c
+++ b/thunar/thunar-device-monitor.c
@@ -510,9 +510,10 @@ thunar_device_monitor_mount_added (GVolumeMonitor      *volume_monitor,
       if (G_UNLIKELY (location == NULL))
         return;
 
-      /* skip gphoto and mtp locations, since those also have a volume */
+      /* skip gphoto, mtp and cdda locations, since those also have a volume */
       if (g_file_has_uri_scheme (location, "gphoto2")
-          || g_file_has_uri_scheme (location, "mtp"))
+          || g_file_has_uri_scheme (location, "mtp")
+          || g_file_has_uri_scheme (location, "cdda"))
         {
           g_object_unref (location);
           return;
