diff --git a/thunar/thunar-io-jobs-util.c b/thunar/thunar-io-jobs-util.c
index e875f60d..827ff59d 100644
--- a/thunar/thunar-io-jobs-util.c
+++ b/thunar/thunar-io-jobs-util.c
@@ -67,7 +67,7 @@ thunar_io_jobs_util_next_duplicate_file (ThunarJob *job,
   const gchar *old_display_name;
   gchar       *display_name;
   gchar       *file_basename;
-  gchar       *dot = NULL;
+  gchar       *extension = NULL;
 
   _thunar_return_val_if_fail (THUNAR_IS_JOB (job), NULL);
   _thunar_return_val_if_fail (G_IS_FILE (file), NULL);
@@ -97,13 +97,13 @@ thunar_io_jobs_util_next_duplicate_file (ThunarJob *job,
     {
       /* get file extension if file is not a directory */
       if (g_file_info_get_file_type (info) != G_FILE_TYPE_DIRECTORY)
-        dot = thunar_util_str_get_extension (old_display_name);
+        extension = thunar_util_str_get_extension (old_display_name);
 
-      if (dot != NULL)
+      if (extension != NULL)
         {
-          file_basename = g_strndup (old_display_name, dot - old_display_name);
+          file_basename = g_strndup (old_display_name, extension - old_display_name);
           /* I18N: put " (copy #)" between basename and extension */
-          display_name = g_strdup_printf (_("%s (copy %u)%s"), file_basename, n, dot);
+          display_name = g_strdup_printf (_("%s (copy %u)%s"), file_basename, n, extension);
           g_free(file_basename);
         }
       else
@@ -173,7 +173,7 @@ thunar_io_jobs_util_next_renamed_file (ThunarJob *job,
   const gchar *old_display_name;
   gchar       *display_name;
   gchar       *file_basename;
-  gchar       *dot = NULL;
+  gchar       *extension = NULL;
 
   _thunar_return_val_if_fail (THUNAR_IS_JOB (job), NULL);
   _thunar_return_val_if_fail (G_IS_FILE (src_file), NULL);
@@ -203,13 +203,13 @@ thunar_io_jobs_util_next_renamed_file (ThunarJob *job,
   old_display_name = g_file_info_get_display_name (info);
   /* get file extension if file is not a directory */
   if (g_file_info_get_file_type (info) != G_FILE_TYPE_DIRECTORY)
-    dot = thunar_util_str_get_extension (old_display_name);
+    extension = thunar_util_str_get_extension (old_display_name);
 
-  if (dot != NULL)
+  if (extension != NULL)
     {
-      file_basename = g_strndup (old_display_name, dot - old_display_name);
+      file_basename = g_strndup (old_display_name, extension - old_display_name);
       /* I18N: put " (copy #)" between basename and extension */
-      display_name = g_strdup_printf (_("%s (copy %u)%s"), file_basename, n, dot);
+      display_name = g_strdup_printf (_("%s (copy %u)%s"), file_basename, n, extension);
       g_free(file_basename);
     }
   else
