diff --git a/thunar/thunar-util.c b/thunar/thunar-util.c
index 934d270f..3f0aa794 100644
--- a/thunar/thunar-util.c
+++ b/thunar/thunar-util.c
@@ -715,9 +715,9 @@ thunar_util_next_new_file_name (ThunarFile   *dir,
       if (!found_duplicate)
         break;
       g_free (new_name);
-      new_name = g_strdup_printf (_("%s %u%s"), file_name, ++count, extension ? extension : "");
+      new_name = g_strdup_printf (_("%.*s %u%s"), (int) file_name_size, file_name, ++count, extension ? extension : "");
     }
   g_object_unref (G_OBJECT (folder));
 
   return new_name;
-}
\ No newline at end of file
+}
