diff --git a/configure.ac.in b/configure.ac.in
index b9ea8e37..0ef56739 100644
--- a/configure.ac.in
+++ b/configure.ac.in
@@ -152,7 +152,7 @@ XDT_CHECK_PACKAGE([GTHREAD], [gthread-2.0], [2.56.0])
 XDT_CHECK_PACKAGE([GMODULE], [gmodule-2.0], [2.56.0])
 XDT_CHECK_PACKAGE([GTK], [gtk+-3.0], [3.22.0])
 XDT_CHECK_PACKAGE([GDK_PIXBUF], [gdk-pixbuf-2.0], [2.14.0])
-XDT_CHECK_PACKAGE([LIBXFCE4UTIL], [libxfce4util-1.0], [4.17.0])
+XDT_CHECK_PACKAGE([LIBXFCE4UTIL], [libxfce4util-1.0], [4.17.1])
 XDT_CHECK_PACKAGE([LIBXFCE4UI], [libxfce4ui-2], [4.17.0])
 XDT_CHECK_PACKAGE([LIBXFCE4KBD_PRIVATE], [libxfce4kbd-private-3], [4.12.0])
 XDT_CHECK_PACKAGE([XFCONF], [libxfconf-0], [4.12.0])
diff --git a/thunar/thunar-file.c b/thunar/thunar-file.c
index e3fe1ae6..8cb4d338 100644
--- a/thunar/thunar-file.c
+++ b/thunar/thunar-file.c
@@ -3617,7 +3617,6 @@ thunar_file_get_thumbnail_path (ThunarFile *file, ThunarThumbnailSize thumbnail_
           g_free (uri);
 
           filename = g_strconcat (g_checksum_get_string (checksum), ".png", NULL);
-          g_checksum_free (checksum);
 
           /* The thumbnail is in the format/location
            * $XDG_CACHE_HOME/thumbnails/(nromal|large)/MD5_Hash_Of_URI.png
@@ -3642,14 +3641,29 @@ thunar_file_get_thumbnail_path (ThunarFile *file, ThunarThumbnailSize thumbnail_
                                                        filename, NULL);
 
               if(!g_file_test(file->thumbnail_path, G_FILE_TEST_EXISTS))
-              {
-                /* Thumbnail doesn't exist in either spot */
-                g_free(file->thumbnail_path);
-                file->thumbnail_path = NULL;
-              }
+                {
+                  g_free(file->thumbnail_path);
+                  file->thumbnail_path = NULL;
+
+                  if (thunar_file_is_directory (file) == FALSE)
+                    {
+                      /* Thumbnail doesn't exist in either spot, look for shared repository */
+                      uri = thunar_file_dup_uri (file);
+                      file->thumbnail_path = xfce_create_shared_thumbnail_path (uri, thunar_thumbnail_size_get_nick (thumbnail_size));
+                      g_free (uri);
+
+                      if (!g_file_test (file->thumbnail_path, G_FILE_TEST_EXISTS))
+                        {
+                          /* Thumbnail doesn't exist */
+                          g_free (file->thumbnail_path);
+                          file->thumbnail_path = NULL;
+                        }
+                    }
+                }
             }
 
           g_free (filename);
+          g_checksum_free (checksum);
         }
     }
 
