diff --git a/thunar/thunar-file.c b/thunar/thunar-file.c
index 7b42b6a1..c855b6ac 100644
--- a/thunar/thunar-file.c
+++ b/thunar/thunar-file.c
@@ -2142,7 +2142,8 @@ thunar_file_get_date (const ThunarFile  *file,
       return g_file_info_get_attribute_int64 (file->recent_info ? file->recent_info : file->info, G_FILE_ATTRIBUTE_RECENT_MODIFIED);
 
     default:
-      _thunar_assert_not_reached ();
+      g_warn_if_reached ();
+      return 0;
     }
 
   return g_file_info_get_attribute_uint64 (file->info, attribute);
diff --git a/thunar/thunar-shortcuts-model.c b/thunar/thunar-shortcuts-model.c
index a3d6376b..12c91d60 100644
--- a/thunar/thunar-shortcuts-model.c
+++ b/thunar/thunar-shortcuts-model.c
@@ -1177,7 +1177,10 @@ thunar_shortcuts_model_get_hidden (ThunarShortcutsModel *model,
   else if (shortcut->location != NULL)
     uri = g_file_get_uri (shortcut->location);
   else
-    _thunar_assert_not_reached ();
+    {
+      g_warn_if_reached ();
+      return FALSE;
+    }
 
   if (uri == NULL)
     return FALSE;
@@ -2251,7 +2254,10 @@ thunar_shortcuts_model_set_hidden (ThunarShortcutsModel *model,
   else if (shortcut->location != NULL)
     uri = g_file_get_uri (shortcut->location);
   else
-    _thunar_assert_not_reached ();
+    {
+      g_warn_if_reached ();
+      return;
+    }
 
   /* prepare array */
   length = model->hidden_bookmarks != NULL ? g_strv_length (model->hidden_bookmarks) : 0;
