diff --git a/thunar/thunar-progress-dialog.c b/thunar/thunar-progress-dialog.c
index d48b7588..bc5411d5 100644
--- a/thunar/thunar-progress-dialog.c
+++ b/thunar/thunar-progress-dialog.c
@@ -96,7 +96,6 @@ thunar_progress_dialog_init (ThunarProgressDialog *dialog)
   gtk_window_set_title (GTK_WINDOW (dialog), _("File Operation Progress"));
   gtk_window_set_default_size (GTK_WINDOW (dialog), 450, 10);
   gtk_window_set_modal (GTK_WINDOW (dialog), FALSE);
-  gtk_window_set_transient_for (GTK_WINDOW (dialog), NULL);
   gtk_window_set_destroy_with_parent (GTK_WINDOW (dialog), FALSE);
   gtk_window_set_type_hint (GTK_WINDOW (dialog), GDK_WINDOW_TYPE_HINT_DIALOG);
 
