diff --git a/thunar/thunar-location-bar.c b/thunar/thunar-location-bar.c
index 81f9337b..50a97b51 100644
--- a/thunar/thunar-location-bar.c
+++ b/thunar/thunar-location-bar.c
@@ -288,15 +288,6 @@ thunar_location_bar_install_widget (ThunarLocationBar    *bar,
 }
 
 
-static gboolean
-thunar_location_bar_settings_changed_cb (gpointer user_data)
-{
-  ThunarLocationBar *bar = user_data;
-
-  thunar_location_bar_settings_changed (bar);
-  return FALSE;
-}
-
 
 static void
 thunar_location_bar_on_enry_edit_done (ThunarLocationEntry *entry,
@@ -304,9 +295,7 @@ thunar_location_bar_on_enry_edit_done (ThunarLocationEntry *entry,
 {
   g_signal_handlers_disconnect_by_func (entry, thunar_location_bar_on_enry_edit_done, bar);
 
-  g_object_ref (bar);
-  g_idle_add_full (G_PRIORITY_HIGH_IDLE, thunar_location_bar_settings_changed_cb, bar, g_object_unref);
-
+  thunar_location_bar_settings_changed (bar);
   g_signal_emit_by_name (bar, "entry-done");
 }
 
diff --git a/thunar/thunar-window.c b/thunar/thunar-window.c
index cc42dbdb..d464e338 100644
--- a/thunar/thunar-window.c
+++ b/thunar/thunar-window.c
@@ -2885,19 +2885,27 @@ thunar_window_start_open_location (ThunarWindow *window,
 {
   _thunar_return_if_fail (THUNAR_IS_WINDOW (window));
 
-  /* temporary show the location toolbar, even if it is normally hidden */
-  gtk_widget_show (window->location_toolbar);
-  thunar_location_bar_request_entry (THUNAR_LOCATION_BAR (window->location_bar), initial_text);
-
   /* setup a search if required */
   if (initial_text != NULL && thunar_util_is_a_search_query (initial_text) == TRUE)
     {
+      /* temporary show the location toolbar, even if it is normally hidden */
+      gtk_widget_show (window->location_toolbar);
+      thunar_location_bar_request_entry (THUNAR_LOCATION_BAR (window->location_bar), initial_text);
+
       thunar_window_update_search (window);
       window->is_searching = TRUE;
       thunar_launcher_set_searching (window->launcher, TRUE);
       if (THUNAR_IS_DETAILS_VIEW (window->view))
         thunar_details_view_set_location_column_visible (THUNAR_DETAILS_VIEW (window->view), TRUE);
     }
+  else /* location edit */
+    {
+      thunar_window_action_cancel_search (window);
+
+      /* temporary show the location toolbar, even if it is normally hidden */
+      gtk_widget_show (window->location_toolbar);
+      thunar_location_bar_request_entry (THUNAR_LOCATION_BAR (window->location_bar), initial_text);
+    }
 }
 
 
