diff --git a/thunar/thunar-window.c b/thunar/thunar-window.c
index d464e338..27cd22f7 100644
--- a/thunar/thunar-window.c
+++ b/thunar/thunar-window.c
@@ -1820,7 +1820,7 @@ thunar_window_unrealize (GtkWidget *widget)
   ThunarWindow *window = THUNAR_WINDOW (widget);
 
   /* disconnect from the clipboard manager */
-  g_signal_handlers_disconnect_by_func (G_OBJECT (window->clipboard), gtk_widget_queue_draw, widget);
+  g_signal_handlers_disconnect_by_func (G_OBJECT (window->clipboard), thunar_window_clipboard_manager_changed, widget);
 
   /* let the GtkWidget class unrealize the window */
   (*GTK_WIDGET_CLASS (thunar_window_parent_class)->unrealize) (widget);
