diff --git a/thunar/thunar-chooser-dialog.c b/thunar/thunar-chooser-dialog.c
index e8bda691..85afdbe5 100644
--- a/thunar/thunar-chooser-dialog.c
+++ b/thunar/thunar-chooser-dialog.c
@@ -791,6 +791,7 @@ thunar_chooser_dialog_action_forget (ThunarChooserDialog *dialog)
   const gchar      *name;
   GtkWidget        *message;
   GAppInfo         *app_info;
+  const gchar      *content_type;
   GError           *error = NULL;
   gint              response;
 
@@ -831,6 +832,11 @@ thunar_chooser_dialog_action_forget (ThunarChooserDialog *dialog)
       /* check if the user confirmed */
       if (G_LIKELY (response == GTK_RESPONSE_YES))
         {
+
+          /* Dont support this mime-type any more with that application */
+          content_type = thunar_file_get_content_type (dialog->file);
+          g_app_info_remove_supports_type (app_info, content_type, NULL);
+
           /* try to delete the application from the model */
           if (!thunar_chooser_model_remove (THUNAR_CHOOSER_MODEL (model), &iter, FALSE, &error))
             {
diff --git a/thunar/thunar-launcher.c b/thunar/thunar-launcher.c
index ff084a73..88a66519 100644
--- a/thunar/thunar-launcher.c
+++ b/thunar/thunar-launcher.c
@@ -873,7 +873,7 @@ thunar_launcher_open_files (ThunarLauncher *launcher,
       else
         {
           /* display a chooser dialog for the file and stop */
-          thunar_show_chooser_dialog (launcher->widget, lp->data, TRUE, FALSE);
+          thunar_show_chooser_dialog (launcher->widget, lp->data, TRUE, TRUE);
           break;
         }
     }
