diff --git a/thunar/thunar-window.c b/thunar/thunar-window.c
index be7baf43..83de10fe 100644
--- a/thunar/thunar-window.c
+++ b/thunar/thunar-window.c
@@ -2953,8 +2953,6 @@ static void
 thunar_window_resume_search (ThunarWindow *window,
                              const gchar  *initial_text)
 {
-  GType view_type;
-
   _thunar_return_if_fail (THUNAR_IS_WINDOW (window));
 
   /* when setting up the location entry do not resent the search query to the standard view, there is a search ongoing */
