diff --git a/thunar/thunar-transfer-job.c b/thunar/thunar-transfer-job.c
index 52c960a7..75ff0c11 100644
--- a/thunar/thunar-transfer-job.c
+++ b/thunar/thunar-transfer-job.c
@@ -557,7 +557,7 @@ ttj_copy_file (ThunarTransferJob *job,
 
   /**
    * MR !127 notes:
-   * (Discusssion: https://gitlab.xfce.org/xfce/thunar/-/merge_requests/127)
+   * (Discussion: https://gitlab.xfce.org/xfce/thunar/-/merge_requests/127)
    *
    * Proper solution for copying lost metadata requires
    * `g_file_build_attribute_list_for_copy`, which is
diff --git a/thunar/thunar-tree-model.c b/thunar/thunar-tree-model.c
index fcd49cd8..2d306c29 100644
--- a/thunar/thunar-tree-model.c
+++ b/thunar/thunar-tree-model.c
@@ -307,7 +307,7 @@ thunar_tree_model_init (ThunarTreeModel *model)
   /* allocate the "virtual root node" */
   model->root = g_node_new (NULL);
 
-  /* inititalize references to certain toplevel nodes */
+  /* initialize references to certain toplevel nodes */
   model->file_system = NULL;
   model->network = NULL;
 
diff --git a/thunar/thunar-tree-view.c b/thunar/thunar-tree-view.c
index 608b325a..185c45ae 100644
--- a/thunar/thunar-tree-view.c
+++ b/thunar/thunar-tree-view.c
@@ -1478,7 +1478,7 @@ thunar_tree_view_action_open (ThunarTreeView *view)
   device = thunar_tree_view_get_selected_device (view);
   file = thunar_tree_view_get_selected_file (view);
 
- if (device != NULL)
+  if (device != NULL)
     {
       if (thunar_device_is_mounted (device))
         thunar_tree_view_open_selection (view);
diff --git a/thunar/thunar-user.c b/thunar/thunar-user.c
index 3c29909d..1e3b3a2e 100644
--- a/thunar/thunar-user.c
+++ b/thunar/thunar-user.c
@@ -219,7 +219,7 @@ thunar_user_class_init (ThunarUserClass *klass)
 
   /* determine the current process' effective user id, we do
    * this only once to avoid the syscall overhead on every
-   * is_me() invokation.
+   * is_me() invocation.
    */
   thunar_user_effective_uid = geteuid ();
 
