diff --git a/thunar/thunar-window.c b/thunar/thunar-window.c
index 870eabf4..ec3d749c 100644
--- a/thunar/thunar-window.c
+++ b/thunar/thunar-window.c
@@ -2864,8 +2864,7 @@ thunar_window_menu_add_bookmarks (ThunarWindow *window,
   gchar          *parse_name;
   gchar          *accel_path;
   gchar          *tooltip;
-  const gchar    *name;
-  gchar          *remote_name;
+  gchar          *name;
   GtkIconTheme   *icon_theme;
   const gchar    *icon_name;
 
@@ -2874,6 +2873,7 @@ thunar_window_menu_add_bookmarks (ThunarWindow *window,
   for (lp = window->bookmarks; lp != NULL; lp = lp->next)
     {
       bookmark = lp->data;
+      name = g_strdup (bookmark->name);
       accel_path = thunar_window_bookmark_get_accel_path (bookmark->g_file);
       parse_name = g_file_get_parse_name (bookmark->g_file);
       tooltip = g_strdup_printf (_("Open the location \"%s\""), parse_name);
@@ -2881,34 +2881,33 @@ thunar_window_menu_add_bookmarks (ThunarWindow *window,
 
       if (g_file_has_uri_scheme (bookmark->g_file, "file"))
         {
-          /* try to open the file corresponding to the uri */
           thunar_file = thunar_file_get (bookmark->g_file, NULL);
-          if (G_LIKELY (thunar_file != NULL))
+
+          if (name == NULL)
             {
-              /* make sure the file refers to a directory */
-              if (G_UNLIKELY (thunar_file_is_directory (thunar_file)))
-                {
-                  name = bookmark->name;
-                  if (bookmark->name == NULL)
-                    name = thunar_file_get_display_name (thunar_file);
-
-                  icon_theme = gtk_icon_theme_get_for_screen (gtk_window_get_screen (GTK_WINDOW (window)));
-                  icon_name = thunar_file_get_icon_name (thunar_file, THUNAR_FILE_ICON_STATE_DEFAULT, icon_theme);
-                  xfce_gtk_image_menu_item_new_from_icon_name (name, tooltip, accel_path, G_CALLBACK (thunar_window_action_open_bookmark), G_OBJECT (bookmark->g_file), icon_name, view_menu);
-               }
+              if (thunar_file != NULL)
+                name = g_strdup (thunar_file_get_display_name (thunar_file));
+              else
+                name = g_file_get_basename (bookmark->g_file);
+            }
+
+          icon_theme = gtk_icon_theme_get_for_screen (gtk_window_get_screen (GTK_WINDOW (window)));
+          icon_name = thunar_file == NULL ? "folder" : thunar_file_get_icon_name (thunar_file, THUNAR_FILE_ICON_STATE_DEFAULT, icon_theme);
+
+          if (thunar_file != NULL)
             g_object_unref (thunar_file);
-          }
         }
       else
         {
-          if (bookmark->name == NULL)
-            remote_name = thunar_g_file_get_display_name_remote (bookmark->g_file);
-          else
-            remote_name = g_strdup (bookmark->name);
-          xfce_gtk_image_menu_item_new_from_icon_name (remote_name, tooltip, accel_path, G_CALLBACK (thunar_window_action_open_bookmark),  G_OBJECT (bookmark->g_file), "folder-remote", view_menu);
-          g_free (remote_name);
+          if (name == NULL)
+            name = thunar_g_file_get_display_name_remote (bookmark->g_file);
+
+          icon_name = "folder-remote";
         }
 
+      xfce_gtk_image_menu_item_new_from_icon_name (name, tooltip, accel_path, G_CALLBACK (thunar_window_action_open_bookmark), G_OBJECT (bookmark->g_file), icon_name, view_menu);
+
+      g_free (name);
       g_free (tooltip);
       g_free (accel_path);
     }
