diff --git a/thunar/thunar-window.c b/thunar/thunar-window.c
index a1fa8d0f..a1c17d93 100644
--- a/thunar/thunar-window.c
+++ b/thunar/thunar-window.c
@@ -5427,6 +5427,7 @@ thunar_window_reconnect_accelerators (ThunarWindow *window)
                                                window);
   thunar_launcher_append_accelerators (window->launcher, window->accel_group);
   thunar_statusbar_append_accelerators (THUNAR_STATUSBAR (window->statusbar), window->accel_group);
+  thunar_window_update_bookmarks (window);
 
   gtk_window_add_accel_group (GTK_WINDOW (window), window->accel_group);
 
