diff --git a/thunar/thunar-enum-types.h b/thunar/thunar-enum-types.h
index 4cfa5a31..6b68d7b2 100644
--- a/thunar/thunar-enum-types.h
+++ b/thunar/thunar-enum-types.h
@@ -417,6 +417,7 @@ typedef enum
     THUNAR_STATUS_BAR_INFO_SIZE_IN_BYTES  = 1 << 1,
     THUNAR_STATUS_BAR_INFO_FILETYPE       = 1 << 2,
     THUNAR_STATUS_BAR_INFO_DISPLAY_NAME   = 1 << 3,
+    THUNAR_STATUS_BAR_INFO_LAST_MODIFIED  = 1 << 4,
 } ThunarStatusBarInfo;
 
 guint    thunar_status_bar_info_toggle_bit    (guint               info,
diff --git a/thunar/thunar-list-model.c b/thunar/thunar-list-model.c
index 8b61025d..5341a9df 100644
--- a/thunar/thunar-list-model.c
+++ b/thunar/thunar-list-model.c
@@ -2812,23 +2812,27 @@ thunar_list_model_get_statusbar_text_for_files (ThunarListModel *store,
                                                 GList           *files,
                                                 gboolean         show_file_size_binary_format)
 {
-  guint64            size_summary     = 0;
-  gint               folder_count     = 0;
-  gint               non_folder_count = 0;
+  guint64            size_summary       = 0;
+  gint               folder_count       = 0;
+  gint               non_folder_count   = 0;
   GList             *lp;
-  GList             *text_list        = NULL;
-  gchar             *size_string      = NULL;
-  gchar             *temp_string      = NULL;
-  gchar             *folder_text      = NULL;
-  gchar             *non_folder_text  = NULL;
+  GList             *text_list          = NULL;
+  gchar             *size_string        = NULL;
+  gchar             *temp_string        = NULL;
+  gchar             *folder_text        = NULL;
+  gchar             *non_folder_text    = NULL;
   ThunarPreferences *preferences;
   guint              active;
-  gboolean           show_size, show_size_in_bytes;
+  guint64            last_modified_date = 0;
+  guint64            temp_last_modified_date;
+  ThunarFile        *last_modified_file = NULL;
+  gboolean           show_size, show_size_in_bytes, show_last_modified;
 
   preferences = thunar_preferences_get ();
   g_object_get (G_OBJECT (preferences), "misc-status-bar-active-info", &active, NULL);
   show_size = thunar_status_bar_info_check_active (active, THUNAR_STATUS_BAR_INFO_SIZE);
   show_size_in_bytes = thunar_status_bar_info_check_active (active, THUNAR_STATUS_BAR_INFO_SIZE_IN_BYTES);
+  show_last_modified = thunar_status_bar_info_check_active (active, THUNAR_STATUS_BAR_INFO_LAST_MODIFIED);
   g_object_unref (preferences);
 
   /* analyze files */
@@ -2844,6 +2848,12 @@ thunar_list_model_get_statusbar_text_for_files (ThunarListModel *store,
           if (thunar_file_is_regular (lp->data))
             size_summary += thunar_file_get_size (lp->data);
         }
+      temp_last_modified_date = thunar_file_get_date (lp->data, THUNAR_FILE_DATE_MODIFIED);
+      if (last_modified_date <= temp_last_modified_date)
+        {
+          last_modified_date = temp_last_modified_date;
+          last_modified_file = lp->data;
+        }
     }
 
   if (non_folder_count > 0)
@@ -2884,6 +2894,13 @@ thunar_list_model_get_statusbar_text_for_files (ThunarListModel *store,
   if (non_folder_text != NULL)
     text_list = g_list_append (text_list, non_folder_text);
 
+  if (show_last_modified && last_modified_file != NULL)
+    {
+      temp_string = thunar_file_get_date_string (last_modified_file, THUNAR_FILE_DATE_MODIFIED, store->date_style, store->date_custom_style);
+      text_list = g_list_append (text_list, g_strdup_printf (_("Last Modified: %s"), temp_string));
+      g_free (temp_string);
+    }
+
   temp_string = thunar_util_strjoin_list (text_list, " | ");
   g_list_free_full (text_list, g_free);
   return temp_string;
@@ -2932,7 +2949,7 @@ thunar_list_model_get_statusbar_text (ThunarListModel *store,
   gboolean           show_file_size_binary_format;
   GList             *relevant_files = NULL;
   guint              active;
-  gboolean           show_size, show_size_in_bytes, show_filetype, show_display_name;
+  gboolean           show_size, show_size_in_bytes, show_filetype, show_display_name, show_last_modified;
 
   _thunar_return_val_if_fail (THUNAR_IS_LIST_MODEL (store), NULL);
 
@@ -2942,6 +2959,7 @@ thunar_list_model_get_statusbar_text (ThunarListModel *store,
   show_size_in_bytes = thunar_status_bar_info_check_active (active, THUNAR_STATUS_BAR_INFO_SIZE_IN_BYTES);
   show_filetype = thunar_status_bar_info_check_active (active, THUNAR_STATUS_BAR_INFO_FILETYPE);
   show_display_name = thunar_status_bar_info_check_active (active, THUNAR_STATUS_BAR_INFO_DISPLAY_NAME);
+  show_last_modified = thunar_status_bar_info_check_active (active, THUNAR_STATUS_BAR_INFO_LAST_MODIFIED);
   show_file_size_binary_format = thunar_list_model_get_file_size_binary(store);
 
   if (selected_items == NULL) /* nothing selected */
@@ -3048,6 +3066,14 @@ thunar_list_model_get_statusbar_text (ThunarListModel *store,
               g_free (file_path);
             }
         }
+
+      if (show_last_modified)
+        {
+          gchar *date_string = thunar_file_get_date_string (file, THUNAR_FILE_DATE_MODIFIED, store->date_style, store->date_custom_style);
+          temp_string = g_strdup_printf (_("Last Modified: %s"), date_string);
+          text_list = g_list_append (text_list, temp_string);
+          g_free (date_string);
+        }
     }
   else /* more than one item selected */
     {
diff --git a/thunar/thunar-statusbar.c b/thunar/thunar-statusbar.c
index 9eb8aebd..d5e334dc 100644
--- a/thunar/thunar-statusbar.c
+++ b/thunar/thunar-statusbar.c
@@ -44,22 +44,23 @@ enum
 
 
 
-static void         thunar_statusbar_finalize                 (GObject              *object);
-static void         thunar_statusbar_set_property             (GObject              *object,
-                                                               guint                 prop_id,
-                                                               const GValue         *value,
-                                                               GParamSpec           *pspec);
-static void         thunar_statusbar_set_text                 (ThunarStatusbar      *statusbar,
-                                                               const gchar          *text);
-static GtkWidget*   thunar_statusbar_context_menu             (ThunarStatusbar      *statusbar);
-static gboolean     thunar_statusbar_button_press_event       (GtkWidget            *widget,
-                                                               GdkEventButton       *event,
-                                                               ThunarStatusbar      *statusbar);
-static gboolean     thunar_statusbar_action_show_size         (ThunarStatusbar      *statusbar);
-static gboolean     thunar_statusbar_action_show_size_bytes   (ThunarStatusbar      *statusbar);
-static gboolean     thunar_statusbar_action_show_filetype     (ThunarStatusbar      *statusbar);
-static gboolean     thunar_statusbar_action_show_display_name (ThunarStatusbar      *statusbar);
-static void         thunar_statusbar_update_all               (void);
+static void         thunar_statusbar_finalize                   (GObject              *object);
+static void         thunar_statusbar_set_property               (GObject              *object,
+                                                                 guint                 prop_id,
+                                                                 const GValue         *value,
+                                                                 GParamSpec           *pspec);
+static void         thunar_statusbar_set_text                   (ThunarStatusbar      *statusbar,
+                                                                 const gchar          *text);
+static GtkWidget*   thunar_statusbar_context_menu               (ThunarStatusbar      *statusbar);
+static gboolean     thunar_statusbar_button_press_event         (GtkWidget            *widget,
+                                                                 GdkEventButton       *event,
+                                                                 ThunarStatusbar      *statusbar);
+static gboolean     thunar_statusbar_action_show_size           (ThunarStatusbar      *statusbar);
+static gboolean     thunar_statusbar_action_show_size_bytes     (ThunarStatusbar      *statusbar);
+static gboolean     thunar_statusbar_action_show_filetype       (ThunarStatusbar      *statusbar);
+static gboolean     thunar_statusbar_action_show_display_name   (ThunarStatusbar      *statusbar);
+static gboolean     thunar_statusbar_action_show_last_modified  (ThunarStatusbar      *statusbar);
+static void         thunar_statusbar_update_all                 (void);
 
 
 struct _ThunarStatusbarClass
@@ -77,10 +78,11 @@ struct _ThunarStatusbar
 
 static XfceGtkActionEntry thunar_status_bar_action_entries[] =
     {
-        { THUNAR_STATUS_BAR_ACTION_TOGGLE_SIZE,           "<Actions>/ThunarStatusBar/toggle-size",            "", XFCE_GTK_CHECK_MENU_ITEM,       N_ ("Size"),          N_ ("Show size"),               NULL, G_CALLBACK (thunar_statusbar_action_show_size),            },
-        { THUNAR_STATUS_BAR_ACTION_TOGGLE_SIZE_IN_BYTES,  "<Actions>/ThunarStatusBar/toggle-size-in-bytes",   "", XFCE_GTK_CHECK_MENU_ITEM,       N_ ("Size in bytes"), N_ ("Show size in bytes"),      NULL, G_CALLBACK (thunar_statusbar_action_show_size_bytes),      },
-        { THUNAR_STATUS_BAR_ACTION_TOGGLE_FILETYPE,       "<Actions>/ThunarStatusBar/toggle-filetype",        "", XFCE_GTK_CHECK_MENU_ITEM,       N_ ("Filetype"),      N_ ("Show filetype"),           NULL, G_CALLBACK (thunar_statusbar_action_show_filetype),        },
-        { THUNAR_STATUS_BAR_ACTION_TOGGLE_DISPLAY_NAME,   "<Actions>/ThunarStatusBar/toggle-display-name",    "", XFCE_GTK_CHECK_MENU_ITEM,       N_ ("Display Name"),  N_ ("Show display name"),       NULL, G_CALLBACK (thunar_statusbar_action_show_display_name),    },
+        { THUNAR_STATUS_BAR_ACTION_TOGGLE_SIZE,           "<Actions>/ThunarStatusBar/toggle-size",             "", XFCE_GTK_CHECK_MENU_ITEM,       N_ ("Size"),          N_ ("Show size"),               NULL, G_CALLBACK (thunar_statusbar_action_show_size),            },
+        { THUNAR_STATUS_BAR_ACTION_TOGGLE_SIZE_IN_BYTES,  "<Actions>/ThunarStatusBar/toggle-size-in-bytes",    "", XFCE_GTK_CHECK_MENU_ITEM,       N_ ("Size in bytes"), N_ ("Show size in bytes"),      NULL, G_CALLBACK (thunar_statusbar_action_show_size_bytes),      },
+        { THUNAR_STATUS_BAR_ACTION_TOGGLE_FILETYPE,       "<Actions>/ThunarStatusBar/toggle-filetype",         "", XFCE_GTK_CHECK_MENU_ITEM,       N_ ("Filetype"),      N_ ("Show filetype"),           NULL, G_CALLBACK (thunar_statusbar_action_show_filetype),        },
+        { THUNAR_STATUS_BAR_ACTION_TOGGLE_DISPLAY_NAME,   "<Actions>/ThunarStatusBar/toggle-display-name",     "", XFCE_GTK_CHECK_MENU_ITEM,       N_ ("Display Name"),  N_ ("Show display name"),       NULL, G_CALLBACK (thunar_statusbar_action_show_display_name),    },
+        { THUNAR_STATUS_BAR_ACTION_TOGGLE_LAST_MODIFIED,  "<Actions>/ThunarStatusBar/toggle-last-modified",    "", XFCE_GTK_CHECK_MENU_ITEM,       N_ ("Last Modified"), N_ ("Show last modified"),      NULL, G_CALLBACK (thunar_statusbar_action_show_last_modified),   },
     };
 
 #define get_action_entry(id) xfce_gtk_get_action_entry_by_id (thunar_status_bar_action_entries, G_N_ELEMENTS (thunar_status_bar_action_entries) ,id)
@@ -234,13 +236,14 @@ thunar_statusbar_context_menu (ThunarStatusbar *statusbar)
   GtkWidget *context_menu = gtk_menu_new();
   GtkWidget *widget;
   guint      active;
-  gboolean   show_size, show_size_in_bytes, show_filetype, show_display_name;
+  gboolean   show_size, show_size_in_bytes, show_filetype, show_display_name, show_last_modified;
 
   g_object_get (G_OBJECT (statusbar->preferences), "misc-status-bar-active-info", &active, NULL);
   show_size = active & THUNAR_STATUS_BAR_INFO_SIZE;
   show_size_in_bytes = active & THUNAR_STATUS_BAR_INFO_SIZE_IN_BYTES;
   show_filetype = active & THUNAR_STATUS_BAR_INFO_FILETYPE;
   show_display_name = active & THUNAR_STATUS_BAR_INFO_DISPLAY_NAME;
+  show_last_modified = active & THUNAR_STATUS_BAR_INFO_LAST_MODIFIED;
 
   xfce_gtk_toggle_menu_item_new_from_action_entry (get_action_entry (THUNAR_STATUS_BAR_ACTION_TOGGLE_SIZE), G_OBJECT (statusbar), show_size, GTK_MENU_SHELL (context_menu));
   widget = xfce_gtk_toggle_menu_item_new_from_action_entry (get_action_entry (THUNAR_STATUS_BAR_ACTION_TOGGLE_SIZE_IN_BYTES), G_OBJECT (statusbar), show_size_in_bytes, GTK_MENU_SHELL (context_menu));
@@ -248,6 +251,7 @@ thunar_statusbar_context_menu (ThunarStatusbar *statusbar)
     gtk_widget_set_sensitive (widget, FALSE);
   xfce_gtk_toggle_menu_item_new_from_action_entry (get_action_entry (THUNAR_STATUS_BAR_ACTION_TOGGLE_FILETYPE), G_OBJECT (statusbar), show_filetype, GTK_MENU_SHELL (context_menu));
   xfce_gtk_toggle_menu_item_new_from_action_entry (get_action_entry (THUNAR_STATUS_BAR_ACTION_TOGGLE_DISPLAY_NAME), G_OBJECT (statusbar), show_display_name, GTK_MENU_SHELL (context_menu));
+  xfce_gtk_toggle_menu_item_new_from_action_entry (get_action_entry (THUNAR_STATUS_BAR_ACTION_TOGGLE_LAST_MODIFIED), G_OBJECT (statusbar), show_last_modified, GTK_MENU_SHELL (context_menu));
 
   gtk_widget_show_all (GTK_WIDGET (context_menu));
 
@@ -316,6 +320,21 @@ thunar_statusbar_action_show_display_name (ThunarStatusbar *statusbar)
 
 
 
+static gboolean
+thunar_statusbar_action_show_last_modified (ThunarStatusbar *statusbar)
+{
+  guint active;
+
+  g_object_get (G_OBJECT (statusbar->preferences), "misc-status-bar-active-info", &active, NULL);
+  g_object_set (G_OBJECT (statusbar->preferences), "misc-status-bar-active-info", thunar_status_bar_info_toggle_bit (active, THUNAR_STATUS_BAR_INFO_LAST_MODIFIED), NULL);
+  thunar_statusbar_update_all ();
+
+  /* required in case of shortcut activation, in order to signal that the accel key got handled */
+  return TRUE;
+}
+
+
+
 static void
 thunar_statusbar_update_all (void)
 {
diff --git a/thunar/thunar-statusbar.h b/thunar/thunar-statusbar.h
index 415425cb..36e7dbf6 100644
--- a/thunar/thunar-statusbar.h
+++ b/thunar/thunar-statusbar.h
@@ -44,6 +44,7 @@ typedef enum
     THUNAR_STATUS_BAR_ACTION_TOGGLE_SIZE_IN_BYTES,
     THUNAR_STATUS_BAR_ACTION_TOGGLE_FILETYPE,
     THUNAR_STATUS_BAR_ACTION_TOGGLE_DISPLAY_NAME,
+    THUNAR_STATUS_BAR_ACTION_TOGGLE_LAST_MODIFIED,
 
     THUNAR_STATUS_BAR_N_ACTIONS
 } ThunarStatusBarAction;
