diff --git a/plugins/thunar-apr/thunar-apr-desktop-page.c b/plugins/thunar-apr/thunar-apr-desktop-page.c
index 9b820f4d..4c099509 100644
--- a/plugins/thunar-apr/thunar-apr-desktop-page.c
+++ b/plugins/thunar-apr/thunar-apr-desktop-page.c
@@ -36,8 +36,6 @@
 #include <unistd.h>
 #endif
 
-#include <exo/exo.h>
-
 #include <libxfce4ui/libxfce4ui.h>
 #include <libxfce4util/libxfce4util.h>
 #include <thunar-apr/thunar-apr-desktop-page.h>
@@ -539,7 +537,7 @@ thunar_apr_desktop_page_file_changed (ThunarAprAbstractPage *abstract_page,
 
       /* update the "Description" entry */
       value = g_key_file_get_locale_string (key_file, G_KEY_FILE_DESKTOP_GROUP, "GenericName", NULL, NULL);
-      if (!exo_str_is_equal (value, desktop_page->description_text))
+      if (g_strcmp0 (value, desktop_page->description_text) != 0)
         {
           /* update the entry */
           gtk_entry_set_text (GTK_ENTRY (desktop_page->description_entry), (value != NULL) ? value : "");
@@ -555,7 +553,7 @@ thunar_apr_desktop_page_file_changed (ThunarAprAbstractPage *abstract_page,
 
       /* update the "Comment" entry */
       value = g_key_file_get_locale_string (key_file, G_KEY_FILE_DESKTOP_GROUP, "Comment", NULL, NULL);
-      if (!exo_str_is_equal (value, desktop_page->comment_text))
+      if (g_strcmp0 (value, desktop_page->comment_text) != 0)
         {
           /* update the entry */
           gtk_entry_set_text (GTK_ENTRY (desktop_page->comment_entry), (value != NULL) ? value : "");
@@ -574,7 +572,7 @@ thunar_apr_desktop_page_file_changed (ThunarAprAbstractPage *abstract_page,
         {
           /* update the "Command" entry */
           value = g_key_file_get_string (key_file, G_KEY_FILE_DESKTOP_GROUP, "Exec", NULL);
-          if (!exo_str_is_equal (value, desktop_page->command_text))
+          if (g_strcmp0 (value, desktop_page->command_text) != 0)
             {
               /* update the entry */
               gtk_entry_set_text (GTK_ENTRY (desktop_page->command_entry), (value != NULL) ? value : "");
@@ -590,7 +588,7 @@ thunar_apr_desktop_page_file_changed (ThunarAprAbstractPage *abstract_page,
 
           /* update the "Path" entry */
           value = g_key_file_get_string (key_file, G_KEY_FILE_DESKTOP_GROUP, "Path", NULL);
-          if (!exo_str_is_equal (value, desktop_page->path_text))
+          if (g_strcmp0 (value, desktop_page->path_text) != 0)
             {
               /* update the entry */
               gtk_entry_set_text (GTK_ENTRY (desktop_page->path_entry), (value != NULL) ? value : "");
@@ -629,7 +627,7 @@ thunar_apr_desktop_page_file_changed (ThunarAprAbstractPage *abstract_page,
         {
           /* update the "URL" entry */
           value = g_key_file_get_string (key_file, G_KEY_FILE_DESKTOP_GROUP, "URL", NULL);
-          if (!exo_str_is_equal (value, desktop_page->url_text))
+          if (g_strcmp0 (value, desktop_page->url_text) != 0)
             {
               /* update the entry */
               gtk_entry_set_text (GTK_ENTRY (desktop_page->url_entry), (value != NULL) ? value : "");
diff --git a/plugins/thunar-sbr/thunar-sbr-date-renamer.c b/plugins/thunar-sbr/thunar-sbr-date-renamer.c
index 4c9bf95d..0d2bf11f 100644
--- a/plugins/thunar-sbr/thunar-sbr-date-renamer.c
+++ b/plugins/thunar-sbr/thunar-sbr-date-renamer.c
@@ -34,10 +34,12 @@
 #include <stdio.h>
 #endif
 
-#include <exo/exo.h>
-
 #include <thunar-sbr/thunar-sbr-date-renamer.h>
 
+#include <libxfce4util/libxfce4util.h>
+
+
+
 #ifdef HAVE_EXIF
 #include <libexif/exif-data.h>
 #endif
@@ -673,7 +675,7 @@ thunar_sbr_date_renamer_set_format (ThunarSbrDateRenamer *date_renamer,
   g_return_if_fail (THUNAR_SBR_IS_DATE_RENAMER (date_renamer));
 
   /* check if we have a new format */
-  if (G_LIKELY (!exo_str_is_equal (date_renamer->format, format)))
+  if (G_LIKELY (g_strcmp0 (date_renamer->format, format) != 0))
     {
       /* apply the new format */
       g_free (date_renamer->format);
diff --git a/plugins/thunar-sbr/thunar-sbr-insert-renamer.c b/plugins/thunar-sbr/thunar-sbr-insert-renamer.c
index cf038f8e..65f731f7 100644
--- a/plugins/thunar-sbr/thunar-sbr-insert-renamer.c
+++ b/plugins/thunar-sbr/thunar-sbr-insert-renamer.c
@@ -22,10 +22,10 @@
 #include <config.h>
 #endif
 
-#include <exo/exo.h>
-
 #include <thunar-sbr/thunar-sbr-insert-renamer.h>
 
+#include <libxfce4util/libxfce4util.h>
+
 
 
 /* Property identifiers */
@@ -558,7 +558,7 @@ thunar_sbr_insert_renamer_set_text (ThunarSbrInsertRenamer *insert_renamer,
   g_return_if_fail (THUNAR_SBR_IS_INSERT_RENAMER (insert_renamer));
 
   /* check if we have a new text */
-  if (G_LIKELY (!exo_str_is_equal (insert_renamer->text, text)))
+  if (G_LIKELY (g_strcmp0 (insert_renamer->text, text) != 0))
     {
       /* apply the new text */
       g_free (insert_renamer->text);
diff --git a/plugins/thunar-sbr/thunar-sbr-number-renamer.c b/plugins/thunar-sbr/thunar-sbr-number-renamer.c
index d5c0a41f..0845f6d4 100644
--- a/plugins/thunar-sbr/thunar-sbr-number-renamer.c
+++ b/plugins/thunar-sbr/thunar-sbr-number-renamer.c
@@ -32,10 +32,10 @@
 #include <string.h>
 #endif
 
-#include <exo/exo.h>
-
 #include <thunar-sbr/thunar-sbr-number-renamer.h>
 
+#include <libxfce4util/libxfce4util.h>
+
 
 
 /* Property identifiers */
@@ -598,7 +598,7 @@ thunar_sbr_number_renamer_set_start (ThunarSbrNumberRenamer *number_renamer,
   g_return_if_fail (THUNAR_SBR_IS_NUMBER_RENAMER (number_renamer));
 
   /* check if we have a new start */
-  if (!exo_str_is_equal (number_renamer->start, start))
+  if (g_strcmp0 (number_renamer->start, start) != 0)
     {
       /* apply the new start */
       g_free (number_renamer->start);
@@ -645,7 +645,7 @@ thunar_sbr_number_renamer_set_text (ThunarSbrNumberRenamer *number_renamer,
   g_return_if_fail (THUNAR_SBR_IS_NUMBER_RENAMER (number_renamer));
 
   /* check if we have a new text */
-  if (G_LIKELY (!exo_str_is_equal (number_renamer->text, text)))
+  if (G_LIKELY (g_strcmp0 (number_renamer->text, text) != 0))
     {
       /* apply the new text */
       g_free (number_renamer->text);
diff --git a/plugins/thunar-sbr/thunar-sbr-replace-renamer.c b/plugins/thunar-sbr/thunar-sbr-replace-renamer.c
index 91ea25cd..6bb33dc6 100644
--- a/plugins/thunar-sbr/thunar-sbr-replace-renamer.c
+++ b/plugins/thunar-sbr/thunar-sbr-replace-renamer.c
@@ -33,10 +33,10 @@
 #include <pcre.h>
 #endif
 
-#include <exo/exo.h>
-
 #include <thunar-sbr/thunar-sbr-replace-renamer.h>
 
+#include <libxfce4util/libxfce4util.h>
+
 
 
 /* Property identifiers */
@@ -742,7 +742,7 @@ thunar_sbr_replace_renamer_set_pattern (ThunarSbrReplaceRenamer *replace_renamer
   g_return_if_fail (g_utf8_validate (pattern, -1, NULL));
 
   /* check if we have a new pattern */
-  if (!exo_str_is_equal (replace_renamer->pattern, pattern))
+  if (g_strcmp0 (replace_renamer->pattern, pattern) != 0)
     {
       /* apply the new value */
       g_free (replace_renamer->pattern);
@@ -851,7 +851,7 @@ thunar_sbr_replace_renamer_set_replacement (ThunarSbrReplaceRenamer *replace_ren
   g_return_if_fail (g_utf8_validate (replacement, -1, NULL));
 
   /* check if we have a new replacement */
-  if (!exo_str_is_equal (replace_renamer->replacement, replacement))
+  if (g_strcmp0 (replace_renamer->replacement, replacement) != 0)
     {
       /* apply the setting */
       g_free (replace_renamer->replacement);
diff --git a/thunar/thunar-application.c b/thunar/thunar-application.c
index 578047c3..b071b9fe 100644
--- a/thunar/thunar-application.c
+++ b/thunar/thunar-application.c
@@ -1523,7 +1523,7 @@ thunar_application_bulk_rename (ThunarApplication *application,
   for (n = 0; filenames[n] != NULL; ++n)
     {
       /* check if the filename is an absolute path or looks like an URI */
-      if (g_path_is_absolute (filenames[n]) || exo_str_looks_like_an_uri (filenames[n]))
+      if (g_path_is_absolute (filenames[n]) || g_uri_is_valid (filenames[n], G_URI_FLAGS_NONE, NULL))
         {
           /* determine the file for the filename directly */
           file = thunar_file_get_for_uri (filenames[n], error);
@@ -1751,7 +1751,7 @@ thunar_application_process_filenames (ThunarApplication               *applicati
   for (n = 0; filenames[n] != NULL; ++n)
     {
       /* check if the filename is an absolute path or looks like an URI */
-      if (g_path_is_absolute (filenames[n]) || exo_str_looks_like_an_uri (filenames[n]))
+      if (g_path_is_absolute (filenames[n]) || g_uri_is_valid (filenames[n], G_URI_FLAGS_NONE, NULL))
         {
           /* determine the file for the filename directly */
           file = thunar_file_get_for_uri (filenames[n], &derror);
diff --git a/thunar/thunar-dialogs.c b/thunar/thunar-dialogs.c
index cbc73715..88fece86 100644
--- a/thunar/thunar-dialogs.c
+++ b/thunar/thunar-dialogs.c
@@ -335,7 +335,7 @@ thunar_dialogs_show_rename_file (gpointer    parent,
       text = xfce_filename_input_get_text (filename_input);
 
       /* check if we have a new name here */
-      if (G_LIKELY (!exo_str_is_equal (filename, text)))
+      if (G_LIKELY (g_strcmp0 (filename, text)) != 0)
         {
           /* try to rename the file */
           job = thunar_io_jobs_rename_file (file, text);
@@ -1043,7 +1043,7 @@ thunar_dialogs_show_insecure_program (gpointer     parent,
                                        "this program, click Cancel."),
                                        thunar_file_get_display_name (file));
   g_string_append (secondary, "\n\n");
-  if (exo_str_looks_like_an_uri (command))
+  if (g_uri_is_valid (command, G_URI_FLAGS_NONE, NULL))
     g_string_append_printf (secondary, G_KEY_FILE_DESKTOP_KEY_URL"=%s", command);
   else
     g_string_append_printf (secondary, G_KEY_FILE_DESKTOP_KEY_EXEC"=%s", command);
diff --git a/thunar/thunar-file.c b/thunar/thunar-file.c
index 4772841c..c57167ee 100644
--- a/thunar/thunar-file.c
+++ b/thunar/thunar-file.c
@@ -1632,7 +1632,7 @@ thunar_file_execute (ThunarFile  *file,
         }
 
       type = g_key_file_get_string (key_file, G_KEY_FILE_DESKTOP_GROUP, G_KEY_FILE_DESKTOP_KEY_TYPE, NULL);
-      if (G_LIKELY (exo_str_is_equal (type, "Application")))
+      if (G_LIKELY (g_strcmp0 (type, "Application") == 0))
         {
           exec = g_key_file_get_string (key_file, G_KEY_FILE_DESKTOP_GROUP, G_KEY_FILE_DESKTOP_KEY_EXEC, NULL);
           if (G_LIKELY (exec != NULL))
@@ -1669,7 +1669,7 @@ thunar_file_execute (ThunarFile  *file,
                            _("No Exec field specified"));
             }
         }
-      else if (exo_str_is_equal (type, "Link"))
+      else if (g_strcmp0 (type, "Link") == 0)
         {
           url = g_key_file_get_string (key_file, G_KEY_FILE_DESKTOP_GROUP, G_KEY_FILE_DESKTOP_KEY_URL, NULL);
           if (G_LIKELY (url != NULL))
@@ -4425,7 +4425,7 @@ thunar_file_same_filesystem (const ThunarFile *file_a,
                                                       G_FILE_ATTRIBUTE_ID_FILESYSTEM);
 
   /* compare the filesystem IDs */
-  return exo_str_is_equal (filesystem_id_a, filesystem_id_b);
+  return (g_strcmp0 (filesystem_id_a, filesystem_id_b) == 0);
 }
 
 
diff --git a/thunar/thunar-gobject-extensions.c b/thunar/thunar-gobject-extensions.c
index ebe0482d..c22646ab 100644
--- a/thunar/thunar-gobject-extensions.c
+++ b/thunar/thunar-gobject-extensions.c
@@ -33,10 +33,10 @@
 #include <string.h>
 #endif
 
-#include <exo/exo.h>
-
 #include <thunar/thunar-gobject-extensions.h>
 
+#include <gio/gio.h>
+
 
 
 static void transform_string_to_boolean (const GValue *src, GValue *dst);
@@ -68,7 +68,7 @@ transform_string_to_enum (const GValue *src,
   for (n = 0; n < klass->n_values; ++n)
     {
       value = klass->values[n].value;
-      if (exo_str_is_equal (klass->values[n].value_name, g_value_get_string (src)))
+      if (g_strcmp0 (klass->values[n].value_name, g_value_get_string (src)) == 0)
         break;
     }
   g_type_class_unref (klass);
diff --git a/thunar/thunar-icon-factory.c b/thunar/thunar-icon-factory.c
index 7299ebaa..25287476 100644
--- a/thunar/thunar-icon-factory.c
+++ b/thunar/thunar-icon-factory.c
@@ -679,7 +679,7 @@ thunar_icon_key_equal (gconstpointer a,
     return FALSE;
 
   /* do a full string comparison on the names */
-  return exo_str_is_equal (a_key->name, b_key->name);
+  return (g_strcmp0 (a_key->name, b_key->name) == 0);
 }
 
 
diff --git a/thunar/thunar-path-entry.c b/thunar/thunar-path-entry.c
index ad7de598..3004ab1a 100644
--- a/thunar/thunar-path-entry.c
+++ b/thunar/thunar-path-entry.c
@@ -590,7 +590,7 @@ thunar_path_entry_changed (GtkEditable *editable)
           thunar_window_action_cancel_search (THUNAR_WINDOW (window));
         }
       /* location/folder-path code */
-      if (G_UNLIKELY (exo_str_looks_like_an_uri (text)))
+      if (G_UNLIKELY (g_uri_is_valid (text, G_URI_FLAGS_NONE, NULL)))
         {
           /* try to parse the URI text */
           escaped_text = g_uri_escape_string (text, G_URI_RESERVED_CHARS_ALLOWED_IN_PATH, TRUE);
diff --git a/thunar/thunar-progress-view.c b/thunar/thunar-progress-view.c
index d1823fcf..94cc7e71 100644
--- a/thunar/thunar-progress-view.c
+++ b/thunar/thunar-progress-view.c
@@ -726,7 +726,7 @@ thunar_progress_view_set_icon_name (ThunarProgressView *view,
 {
   _thunar_return_if_fail (THUNAR_IS_PROGRESS_VIEW (view));
 
-  if (exo_str_is_equal (view->icon_name, icon_name))
+  if (g_strcmp0 (view->icon_name, icon_name) == 0)
     return;
 
   g_free (view->icon_name);
@@ -743,7 +743,7 @@ thunar_progress_view_set_title (ThunarProgressView *view,
 {
   _thunar_return_if_fail (THUNAR_IS_PROGRESS_VIEW (view));
 
-  if (exo_str_is_equal (view->title, title))
+  if (g_strcmp0 (view->title, title) == 0)
     return;
 
   g_free (view->title);
diff --git a/thunar/thunar-renamer-dialog.c b/thunar/thunar-renamer-dialog.c
index c52745f9..6cf187ce 100644
--- a/thunar/thunar-renamer-dialog.c
+++ b/thunar/thunar-renamer-dialog.c
@@ -524,7 +524,7 @@ thunar_renamer_dialog_init (ThunarRenamerDialog *renamer_dialog)
       active_str = xfce_rc_read_entry_untranslated (rc, "LastActiveMode", "");
       for (active = 0, n = 0; n < klass->n_values; ++n)
         {
-          if (exo_str_is_equal (active_str, klass->values[n].value_name))
+          if (g_strcmp0 (active_str, klass->values[n].value_name) == 0)
             active = n;
           gtk_combo_box_text_append_text (GTK_COMBO_BOX_TEXT (mcombo), _(klass->values[n].value_nick));
         }
@@ -553,7 +553,7 @@ thunar_renamer_dialog_init (ThunarRenamerDialog *renamer_dialog)
           gtk_notebook_append_page (GTK_NOTEBOOK (notebook), lp->data, NULL);
 
           /* check if this page should be active by default */
-          if (exo_str_is_equal (G_OBJECT_TYPE_NAME (lp->data), active_str))
+          if (g_strcmp0 (G_OBJECT_TYPE_NAME (lp->data), active_str) == 0)
             active = g_list_position (renamers, lp);
 
           /* try to load the settings for the renamer */
diff --git a/thunar/thunar-renamer-model.c b/thunar/thunar-renamer-model.c
index 4fbd1cdb..c5f7e743 100644
--- a/thunar/thunar-renamer-model.c
+++ b/thunar/thunar-renamer-model.c
@@ -874,7 +874,7 @@ thunar_renamer_model_process_item (ThunarRenamerModel     *renamer_model,
     }
 
   /* check if the new name is equal to the old one */
-  if (exo_str_is_equal (name, display_name))
+  if (g_strcmp0 (name, display_name) == 0)
     {
       /* just return NULL then */
       g_free (name);
@@ -921,7 +921,7 @@ THUNAR_THREADS_ENTER
 
           /* determine the new name for the item */
           name = thunar_renamer_model_process_item (renamer_model, item, idx);
-          if (!exo_str_is_equal (item->name, name))
+          if (g_strcmp0 (item->name, name) != 0)
             {
               /* apply new name */
               g_free (item->name);
diff --git a/thunar/thunar-window.c b/thunar/thunar-window.c
index 550f4938..690866c9 100644
--- a/thunar/thunar-window.c
+++ b/thunar/thunar-window.c
@@ -904,9 +904,9 @@ thunar_window_init (ThunarWindow *window)
   g_signal_connect_object (G_OBJECT (window->preferences), "notify::misc-change-window-icon", G_CALLBACK (thunar_window_update_window_icon), window, G_CONNECT_SWAPPED);
 
   /* determine the selected side pane */
-  if (exo_str_is_equal (last_side_pane, g_type_name (THUNAR_TYPE_SHORTCUTS_PANE)))
+  if (g_strcmp0 (last_side_pane, g_type_name (THUNAR_TYPE_SHORTCUTS_PANE)) == 0)
     type = THUNAR_TYPE_SHORTCUTS_PANE;
-  else if (exo_str_is_equal (last_side_pane, g_type_name (THUNAR_TYPE_TREE_PANE)))
+  else if (g_strcmp0 (last_side_pane, g_type_name (THUNAR_TYPE_TREE_PANE)) == 0)
     type = THUNAR_TYPE_TREE_PANE;
   else
     type = G_TYPE_NONE;
@@ -1255,9 +1255,9 @@ thunar_window_update_view_menu (ThunarWindow *window,
   gtk_menu_set_accel_group (GTK_MENU (sub_items), window->accel_group);
   g_object_get (window->preferences, "last-location-bar", &last_location_bar, NULL);
   xfce_gtk_toggle_menu_item_new_from_action_entry (get_action_entry (THUNAR_WINDOW_ACTION_VIEW_LOCATION_SELECTOR_PATHBAR), G_OBJECT (window),
-                                                   exo_str_is_equal (last_location_bar, g_type_name (THUNAR_TYPE_LOCATION_ENTRY)), GTK_MENU_SHELL (sub_items));
+                                                   (g_strcmp0 (last_location_bar, g_type_name (THUNAR_TYPE_LOCATION_ENTRY)) == 0), GTK_MENU_SHELL (sub_items));
   xfce_gtk_toggle_menu_item_new_from_action_entry (get_action_entry (THUNAR_WINDOW_ACTION_VIEW_LOCATION_SELECTOR_TOOLBAR), G_OBJECT (window),
-                                                   exo_str_is_equal (last_location_bar, g_type_name (THUNAR_TYPE_LOCATION_BUTTONS)), GTK_MENU_SHELL (sub_items));
+                                                   (g_strcmp0 (last_location_bar, g_type_name (THUNAR_TYPE_LOCATION_BUTTONS)) == 0), GTK_MENU_SHELL (sub_items));
   g_free (last_location_bar);
   gtk_menu_item_set_submenu (GTK_MENU_ITEM (item), GTK_WIDGET (sub_items));
   item = xfce_gtk_menu_item_new_from_action_entry (get_action_entry (THUNAR_WINDOW_ACTION_VIEW_SIDE_PANE_MENU), G_OBJECT (window), GTK_MENU_SHELL (menu));
@@ -1265,9 +1265,9 @@ thunar_window_update_view_menu (ThunarWindow *window,
   gtk_menu_set_accel_group (GTK_MENU (sub_items), window->accel_group);
   g_object_get (window->preferences, "last-side-pane", &last_side_pane, NULL);
   xfce_gtk_toggle_menu_item_new_from_action_entry (get_action_entry (THUNAR_WINDOW_ACTION_VIEW_SIDE_PANE_SHORTCUTS), G_OBJECT (window),
-                                                   exo_str_is_equal (last_side_pane, g_type_name (THUNAR_TYPE_SHORTCUTS_PANE)), GTK_MENU_SHELL (sub_items));
+                                                   (g_strcmp0 (last_side_pane, g_type_name (THUNAR_TYPE_SHORTCUTS_PANE)) == 0), GTK_MENU_SHELL (sub_items));
   xfce_gtk_toggle_menu_item_new_from_action_entry (get_action_entry (THUNAR_WINDOW_ACTION_VIEW_SIDE_PANE_TREE), G_OBJECT (window),
-                                                   exo_str_is_equal (last_side_pane, g_type_name (THUNAR_TYPE_TREE_PANE)), GTK_MENU_SHELL (sub_items));
+                                                   (g_strcmp0 (last_side_pane, g_type_name (THUNAR_TYPE_TREE_PANE)) == 0), GTK_MENU_SHELL (sub_items));
   g_free (last_side_pane);
   gtk_menu_item_set_submenu (GTK_MENU_ITEM (item), GTK_WIDGET (sub_items));
   xfce_gtk_toggle_menu_item_new_from_action_entry (get_action_entry (THUNAR_WINDOW_ACTION_VIEW_STATUSBAR), G_OBJECT (window),
@@ -2760,7 +2760,7 @@ thunar_window_update_location_bar_visible (ThunarWindow *window)
 
   g_object_get (window->preferences, "last-location-bar", &last_location_bar, NULL);
 
-  if (exo_str_is_equal (last_location_bar, g_type_name (G_TYPE_NONE)))
+  if (g_strcmp0 (last_location_bar, g_type_name (G_TYPE_NONE)) == 0)
     {
       gtk_widget_hide (window->location_toolbar);
       if (window->view != NULL)
@@ -3451,7 +3451,7 @@ thunar_window_action_pathbar_changed (ThunarWindow *window)
   _thunar_return_val_if_fail (THUNAR_IS_WINDOW (window), FALSE);
 
   g_object_get (window->preferences, "last-location-bar", &last_location_bar, NULL);
-  pathbar_checked = exo_str_is_equal (last_location_bar, g_type_name (THUNAR_TYPE_LOCATION_ENTRY));
+  pathbar_checked = (g_strcmp0 (last_location_bar, g_type_name (THUNAR_TYPE_LOCATION_ENTRY)) == 0);
   g_free (last_location_bar);
 
   if (pathbar_checked)
@@ -3474,7 +3474,7 @@ thunar_window_action_toolbar_changed (ThunarWindow *window)
   _thunar_return_val_if_fail (THUNAR_IS_WINDOW (window), FALSE);
 
   g_object_get (window->preferences, "last-location-bar", &last_location_bar, NULL);
-  toolbar_checked = exo_str_is_equal (last_location_bar, g_type_name (THUNAR_TYPE_LOCATION_BUTTONS));
+  toolbar_checked = (g_strcmp0 (last_location_bar, g_type_name (THUNAR_TYPE_LOCATION_BUTTONS)) == 0);
   g_free (last_location_bar);
 
   if (toolbar_checked)
@@ -3498,7 +3498,7 @@ thunar_window_action_shortcuts_changed (ThunarWindow *window)
   _thunar_return_val_if_fail (THUNAR_IS_WINDOW (window), FALSE);
 
   g_object_get (window->preferences, "last-side-pane", &last_side_pane, NULL);
-  shortcuts_checked = exo_str_is_equal (last_side_pane, g_type_name (THUNAR_TYPE_SHORTCUTS_PANE));
+  shortcuts_checked = (g_strcmp0 (last_side_pane, g_type_name (THUNAR_TYPE_SHORTCUTS_PANE)) == 0);
   g_free (last_side_pane);
 
   if (shortcuts_checked)
@@ -3524,7 +3524,7 @@ thunar_window_action_tree_changed (ThunarWindow *window)
   _thunar_return_val_if_fail (THUNAR_IS_WINDOW (window), FALSE);
 
   g_object_get (window->preferences, "last-side-pane", &last_side_pane, NULL);
-  tree_view_checked = exo_str_is_equal (last_side_pane, g_type_name (THUNAR_TYPE_TREE_PANE));
+  tree_view_checked = (g_strcmp0 (last_side_pane, g_type_name (THUNAR_TYPE_TREE_PANE)) == 0);
   g_free (last_side_pane);
 
   if (tree_view_checked)
@@ -4976,7 +4976,7 @@ thunar_window_set_directories (ThunarWindow   *window,
   for (n = 0; uris[n] != NULL; n++)
     {
       /* check if the string looks like an uri */
-      if (!exo_str_looks_like_an_uri (uris[n]))
+      if (!g_uri_is_valid (uris[n], G_URI_FLAGS_NONE, NULL))
         continue;
 
       /* get the file for the uri */
