diff --git a/NEWS b/NEWS
index 53c42408..aad5e699 100644
--- a/NEWS
+++ b/NEWS
@@ -1,3 +1,40 @@
+4.17.12 (2022-12-01)
+=======
+- Follow Symlinks in 'thunar_file_can_execute' (#944)
+- renamed 'thunar_file_is_executable' to 'thunar_file_can_execute'
+- Regression: Trust all .desktop files inside XDG_DATA_DIRS (Issue #944)
+- Check if trusted_button exists before using it (Issue #944)
+- Show meaningful error for invalid remote bookmarks (Isssue #941)
+- Fix for launcher trusted/executable checkboxes (Issue #944)
+- thunar-apr: Properly free errors (Issue #944)
+- Update copyright to 2022
+- Only show hamburger button when menubar is disabled (Issue #934)
+- Only show hamburger button when menubar is disabled (Issue #934)
+- Properly round percentages in 'usage' text
+- Dbus methods not selecting files/folders (Issue #929)
+- Fix argument type for xfce_shortcuts_editor_new (Issue #928)
+- Reposition Highlight tab in Properties dialog
+- Image preview visible flag desynced between instances
+- Properly include pcre2 CFLAGS and LDFLAGS (Issue #848)
+- Fix capitalization in some strings
+- Set style to reasonable value when setting wallpaper (Issue #907)
+- Fix blurriness in image preview pane when UI scale > 1
+- Fix blurriness in gesture icons when UI scale > 1
+- Fix blurriness in drag icons when UI scale > 1
+- Fix blurriness in path bar location image when UI scale > 1
+- Fix blurriness in ThunarImage when UI scaling > 1
+- Fix blurriness in emblem chooser when UI scaling > 1
+- Fix blurriness in icon renderers at UI scale > 1
+- Translation Updates:
+  Albanian, Arabic, Armenian (Armenia), Basque, Bulgarian, Catalan,
+  Chinese (China), Chinese (Taiwan), Czech, Danish, Dutch, Eastern
+  Armenian, English (Australia), English (United Kingdom), Estonian,
+  Finnish, French, Galician, German, Greek, Hebrew, Hungarian,
+  Icelandic, Indonesian, Interlingue, Italian, Japanese, Kazakh,
+  Korean, Malay, Norwegian Bokmål, Norwegian Nynorsk, Occitan (post
+  1500), Persian (Iran), Polish, Portuguese, Portuguese (Brazil),
+  Romanian, Serbian, Slovak, Spanish, Swedish, Thai
+
 4.17.11 (2022-11-01)
 =======
 - Fix typo for setting "misc-image-preview-mode"
diff --git a/configure.ac.in b/configure.ac.in
index 3e3c5b66..884590a9 100644
--- a/configure.ac.in
+++ b/configure.ac.in
@@ -26,10 +26,10 @@ m4_define([thunarx_verinfo], [0:0:0])
 m4_define([thunarx_version_api], [3])
 m4_define([thunar_version_major], [4])
 m4_define([thunar_version_minor], [17])
-m4_define([thunar_version_micro], [11])
+m4_define([thunar_version_micro], [12])
 m4_define([thunar_version_nano], [])
 m4_define([thunar_version_build], [@REVISION@])
-m4_define([thunar_version_tag], [git])
+m4_define([thunar_version_tag], [])
 m4_define([thunar_version], [thunar_version_major().thunar_version_minor().thunar_version_micro()ifelse(thunar_version_nano(), [], [], [.thunar_version_nano()])ifelse(thunar_version_tag(), [git], [thunar_version_tag()-thunar_version_build()], [thunar_version_tag()])])
 
 dnl *******************************************
